<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Home</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">jazzPanda_workflowr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="xenium-human-breast-cancer-analysis.html">Xenium breast cancer data</a>
    </li>
    <li>
      <a href="xenium-human-lung-cancer-analysis.html">Xenium human lung data</a>
    </li>
    <li>
      <a href="xenium-mouse-brain-analysis.html">Xenium mouse brain data</a>
    </li>
    <li>
      <a href="cosmx-human-healthy-liver.html">Cosmx healthy liver data</a>
    </li>
    <li>
      <a href="cosmx-human-liver-cancer.html">Cosmx liver cancer data</a>
    </li>
    <li>
      <a href="merscope-human-breast-cancer.html">Merscope breast cancer data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Home</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-01-21
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 1
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong>
<code>jazzPanda_workflowr/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsghpersonalmelodyphipsonlabjazzPandaworkflowrtree3dea452c78d5948194281937875b5fd327947949targetblank3dea452a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/tree/3dea452c78d5948194281937875b5fd327947949" target="_blank">3dea452</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsghpersonalmelodyphipsonlabjazzPandaworkflowrtree3dea452c78d5948194281937875b5fd327947949targetblank3dea452a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/tree/3dea452c78d5948194281937875b5fd327947949" target="_blank">3dea452</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .env
    Ignored:    .ipynb_checkpoints/

Untracked files:
    Untracked:  .env.example

Unstaged changes:
    Modified:   .gitignore
    Modified:   analysis/cosmx-human-healthy-liver.Rmd
    Modified:   analysis/cosmx-human-liver-cancer.Rmd
    Modified:   analysis/index.Rmd
    Modified:   analysis/merscope-human-breast-cancer.Rmd
    Modified:   analysis/run_workflow_cosmx_healthy_liver.sh
    Modified:   analysis/run_workflow_cosmx_liver_cancer.sh
    Modified:   analysis/run_workflow_merscope_human_breast_cancer.sh
    Modified:   analysis/run_workflow_xenium_human_breast_cancer.sh
    Modified:   analysis/run_workflow_xenium_human_lung_cancer.sh
    Modified:   analysis/run_workflow_xenium_mouse_brain.sh
    Modified:   analysis/slurm_out/workflowr_cosmx_healthy_liver.err
    Modified:   analysis/slurm_out/workflowr_cosmx_healthy_liver.out
    Modified:   analysis/slurm_out/workflowr_cosmx_liver_cancer.err
    Modified:   analysis/slurm_out/workflowr_cosmx_liver_cancer.out
    Modified:   analysis/slurm_out/workflowr_merscope_human_breast_cancer.err
    Modified:   analysis/slurm_out/workflowr_merscope_human_breast_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_human_breast_cancer.err
    Modified:   analysis/slurm_out/workflowr_xenium_human_breast_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_human_lung_cancer.err
    Modified:   analysis/slurm_out/workflowr_xenium_human_lung_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_mouse_brain.err
    Modified:   analysis/slurm_out/workflowr_xenium_mouse_brain.out
    Modified:   analysis/xenium-human-breast-cancer-analysis.Rmd
    Modified:   analysis/xenium-human-lung-cancer-analysis.Rmd
    Modified:   analysis/xenium-mouse-brain-analysis.Rmd
    Modified:   code/utils.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/index.Rmd</code>) and HTML
(<code>docs/index.html</code>) files. If you’ve configured a remote Git
repository (see <code>?wflow_git_remote</code>), click on the hyperlinks
in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/3dea452c78d5948194281937875b5fd327947949/analysis/index.Rmd" target="_blank">3dea452</a>
</td>
<td>
Melody
</td>
<td>
2025-10-10
</td>
<td>
Add script
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/3dea452c78d5948194281937875b5fd327947949/docs/index.html" target="_blank">3dea452</a>
</td>
<td>
Melody
</td>
<td>
2025-10-10
</td>
<td>
Add script
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This website presents an end-to-end analysis workflow for detecting
marker genes using the <a
href="https://github.com/phipsonlab/jazzPanda">jazzPanda R package</a>.
Additional analysis scripts and extended explorations used in the
jazzPanda paper are available in the <a
href="https://github.com/phipsonlab/jazzPanda_paper">analysis
repository</a>. The underlying datasets, including raw vendor data and
lightly processed data objects, are publicly available via <a
href="https://zenodo.org/uploads/18149456">Zenodo</a>.</p>
<p>This website hosts the analysis outputs for detecting marker genes
with the .</p>
<p>jazzPanda is designed for <strong>spatial transcriptomics</strong>
data and integrates spatial context into marker detection, providing
more robust identification of genes that define clusters of cells across
different tissue contexts. The analyses presented here demonstrate the
application of jazzPanda to a variety of publicly available datasets
from multiple spatial technologies.</p>
<div id="raw-data-access" class="section level2">
<h2>Raw data access</h2>
<div id="x-xenium" class="section level3">
<h3>10x Xenium</h3>
<ul>
<li>Human lung cancer sample
<ul>
<li>Source: <a
href="https://www.10xgenomics.com/resources/datasets/xenium-human-lung-preview-data-1-standard"
class="uri">https://www.10xgenomics.com/resources/datasets/xenium-human-lung-preview-data-1-standard</a></li>
</ul></li>
<li>Mouse brain samples
<ul>
<li>The three samples used in the analysis are replicate 1, replicate 2
and replicate 3.</li>
<li>Source: <a
href="https://www.10xgenomics.com/resources/datasets/fresh-frozen-mouse-brain-replicates-1-standard"
class="uri">https://www.10xgenomics.com/resources/datasets/fresh-frozen-mouse-brain-replicates-1-standard</a></li>
</ul></li>
<li>Human HER2+ breast cancer samples
<ul>
<li>The two samples used in the analysis are sample 1 (rep1) and sample
2.</li>
<li>Source: <a
href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast"
class="uri">https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast</a></li>
<li>Paper: <a
href="https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1"
class="uri">https://www.biorxiv.org/content/10.1101/2022.10.06.510405v1</a></li>
</ul></li>
<li>Gene panel
<ul>
<li>Source: <a
href="https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/panel-design/pre-designed-xenium-gene-expression-panels"
class="uri">https://www.10xgenomics.com/support/in-situ-gene-expression/documentation/steps/panel-design/pre-designed-xenium-gene-expression-panels</a></li>
<li>Human lung cancer sample: Xenium Human Lung Gene Expression
Panel</li>
<li>Mouse brain samples: Xenium Mouse Brain Gene Expression Panel</li>
<li>Human HER2+ breast cancer samples: Xenium Human Breast Gene
Expression Panel</li>
</ul></li>
</ul>
</div>
<div id="nanostring-cosmx" class="section level3">
<h3>Nanostring Cosmx</h3>
<ul>
<li>Human liver healthy and cancer samples</li>
<li>Source: <a
href="https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/human-liver-rna-ffpe-dataset/"
class="uri">https://nanostring.com/products/cosmx-spatial-molecular-imager/ffpe-dataset/human-liver-rna-ffpe-dataset/</a></li>
</ul>
</div>
<div id="merscope" class="section level3">
<h3>Merscope</h3>
<ul>
<li>Human breast cancer sample</li>
<li>Source: <a
href="https://info.vizgen.com/ffpe-showcase?submissionGuid=c9a25730-3fe5-444b-bef8-1a74d51ddefb"
class="uri">https://info.vizgen.com/ffpe-showcase?submissionGuid=c9a25730-3fe5-444b-bef8-1a74d51ddefb</a></li>
</ul>
</div>
</div>
<div id="analysis" class="section level2">
<h2>Analysis</h2>
<ul>
<li><a href="xenium-human-lung-cancer-analysis.html">Xenium human lung
cancer sample</a></li>
<li><a href="xenium-mouse-brain-analysis.html">Xenium mouse brain
samples</a></li>
<li><a href="xenium-human-breast-cancer-analysis.html">Xenium human
HER2+ breast cancer samples</a></li>
<li><a href="cosmx-human-liver-cancer.html">Cosmx human liver cancer
sample</a></li>
<li><a href="cosmx-human-healthy-liver.html">Cosmx human normal liver
sample</a></li>
<li><a href="merscope-human-breast-cancer.html">Merscope breast cancer
sample</a></li>
</ul>
</div>
<div id="license" class="section level2">
<h2>License</h2>
<p>The code in this analysis is covered by the <a
href="https://choosealicense.com/licenses/mit/" title="MIT License">MIT
license</a> and the written content is covered by a <a
href="https://choosealicense.com/licenses/cc-by-4.0/"
title="CC-BY License">Creative Commons CC-BY</a> license.</p>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
