<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Melody Jin" />

<meta name="date" content="2025-09-29" />

<title>Merscope human breast cancer sample</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">jazzPanda_workflowr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="xenium-human-breast-cancer-analysis.html">Xenium breast cancer data</a>
    </li>
    <li>
      <a href="xenium-human-lung-cancer-analysis.html">Xenium human lung data</a>
    </li>
    <li>
      <a href="xenium-mouse-brain-analysis.html">Xenium mouse brain data</a>
    </li>
    <li>
      <a href="cosmx-human-healthy-liver.html">Cosmx healthy liver data</a>
    </li>
    <li>
      <a href="cosmx-human-liver-cancer.html">Cosmx liver cancer data</a>
    </li>
    <li>
      <a href="merscope-human-breast-cancer.html">Merscope breast cancer data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Merscope human breast cancer sample</h1>
<h4 class="author">Melody Jin</h4>
<h4 class="date">2025-09-29</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2026-02-18
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>jazzPanda_workflowr/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed12345code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(12345)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed12345code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(12345)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsghpersonalmelodyphipsonlabjazzPandaworkflowrtreeb68a25dca5d7cde9ee87094399c7ac7ad961d40etargetblankb68a25da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/tree/b68a25dca5d7cde9ee87094399c7ac7ad961d40e" target="_blank">b68a25d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsghpersonalmelodyphipsonlabjazzPandaworkflowrtreeb68a25dca5d7cde9ee87094399c7ac7ad961d40etargetblankb68a25da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/tree/b68a25dca5d7cde9ee87094399c7ac7ad961d40e" target="_blank">b68a25d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    .env
    Ignored:    .ipynb_checkpoints/

Untracked files:
    Untracked:  analysis/figure/
    Untracked:  analysis/slurm_out/workflowr_analysis.err
    Untracked:  analysis/slurm_out/workflowr_analysis.out
    Untracked:  analysis/slurm_out/workflowr_cosmx_hhliver.err
    Untracked:  analysis/slurm_out/workflowr_cosmx_hhliver.out

Unstaged changes:
    Modified:   analysis/run_workflow_cosmx_healthy_liver.sh
    Modified:   analysis/run_workflow_cosmx_liver_cancer.sh
    Modified:   analysis/run_workflow_merscope_human_breast_cancer.sh
    Modified:   analysis/run_workflow_xenium_human_breast_cancer.sh
    Modified:   analysis/run_workflow_xenium_human_lung_cancer.sh
    Modified:   analysis/run_workflow_xenium_mouse_brain.sh
    Modified:   analysis/slurm_out/workflowr_cosmx_liver_cancer.err
    Modified:   analysis/slurm_out/workflowr_cosmx_liver_cancer.out
    Modified:   analysis/slurm_out/workflowr_merscope_human_breast_cancer.err
    Modified:   analysis/slurm_out/workflowr_merscope_human_breast_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_human_breast_cancer.err
    Modified:   analysis/slurm_out/workflowr_xenium_human_breast_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_human_lung_cancer.err
    Modified:   analysis/slurm_out/workflowr_xenium_human_lung_cancer.out
    Modified:   analysis/slurm_out/workflowr_xenium_mouse_brain.err
    Modified:   analysis/slurm_out/workflowr_xenium_mouse_brain.out

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/merscope-human-breast-cancer.Rmd</code>) and HTML
(<code>docs/merscope-human-breast-cancer.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/analysis/merscope-human-breast-cancer.Rmd" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
<td>
update links
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/merscope-human-breast-cancer.html" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
<td>
update links
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/analysis/merscope-human-breast-cancer.Rmd" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
<td>
add scripts
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/merscope-human-breast-cancer.html" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
<td>
add scripts
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(jazzPanda)
library(SpatialExperiment)
library(Seurat)
library(ggplot2)
library(data.table)
library(dplyr)
library(tidyr)
library(glmnet)
library(caret)
library(ComplexUpset)
library(ggrepel)
library(gridExtra)
library(patchwork)
library(RColorBrewer)
library(limma)
library(here)
library(data.table)
library(xtable)
library(corrplot)
source(here(&quot;code/utils.R&quot;))
library(Banksy)
library(harmony)</code></pre>
<div id="load-data" class="section level1">
<h1>Load data</h1>
<div id="load-cell-and-transcript" class="section level2">
<h2>load cell and transcript</h2>
<pre class="r"><code>library(SpatialExperimentIO)
data_nm  &lt;- &quot;merscope_hbreast&quot;
se =  readMerscopeSXE(dirName = rdata$mhb, 
                      countMatPattern = &quot;cell_by_gene.csv&quot;, 
                      metaDataPattern = &quot;cell_metadata.csv&quot;)
x_avg &lt;- (se@colData$min_x + se@colData$max_x) / 2
y_avg &lt;- (se@colData$min_y + se@colData$max_y) / 2

# create a matrix
coords &lt;- cbind(x = x_avg, y = y_avg)

# assign to spatialCoordiantes
SpatialExperiment::spatialCoords(se) &lt;- coords

blank_genes &lt;- grep(&quot;^Blank&quot;, rownames(se), value = TRUE)


blank_idx &lt;- grep(&quot;^Blank&quot;, rownames(se))
se &lt;- se[-blank_idx, ]


lib_size &lt;- Matrix::colSums(assay(se, &quot;counts&quot;))
colData(se)$lib_size &lt;- lib_size

# Filter cells with library size between 30 and 2500
se &lt;- se[, lib_size &gt; 30 &amp; lib_size &lt; 2500]
saveRDS(se, here(gdata,paste0(data_nm, &quot;_se.Rds&quot;)))</code></pre>
<pre class="r"><code>data_nm  &lt;- &quot;merscope_hbreast&quot;
se = readRDS(here(gdata,paste0(data_nm, &quot;_se.Rds&quot;)))

cat(&quot;Loading transcripts\n&quot;)</code></pre>
<pre><code>Loading transcripts</code></pre>
<pre class="r"><code>transcript_df&lt;-fread(file.path(rdata$mhb,
                            &quot;merscope_hbc_detected_transcripts.csv&quot;))
transcript_df$x &lt;- transcript_df$global_x
transcript_df$y &lt;- transcript_df$global_y
transcript_df$gene = make.names(transcript_df$gene)
transcript_df$feature_name = transcript_df$gene
cat(&quot;Transcripts loaded \n&quot;)</code></pre>
<pre><code>Transcripts loaded </code></pre>
<pre class="r"><code>real_genes &lt;- row.names(se)
nc_coords &lt;- transcript_df[!(transcript_df$gene %in% real_genes), ]

nc_coords$feature_name &lt;- nc_coords$gene
nc_coords$feature_name &lt;-factor(nc_coords$feature_name)</code></pre>
</div>
</div>
<div id="single-cell-summary" class="section level1">
<h1>Single cell summary</h1>
<p>These summary plots characterise per-cell detection counts and
sparsity. We can calculate per-cell and per-gene summary statistics from
the counts matrix:</p>
<ul>
<li><p>Total detections per cell: sums all transcript counts in each
cell.</p></li>
<li><p>Proportion of zeroes per cell: fraction of genes not detected in
each cell.</p></li>
<li><p>Detected genes per cell: number of non-zero genes per
cell.</p></li>
<li><p>Average expression per gene: mean expression across cells,
ignoring zeros.</p></li>
</ul>
<p>Each distribution is visualised with histograms and density curves to
assess data quality and sparsity.</p>
<pre class="r"><code>cm=assay(se, &quot;counts&quot;)
td_r1 &lt;- colSums(cm)
summary(td_r1)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
     31     163     368     523     734    2499 </code></pre>
<pre class="r"><code>pz_r1 &lt;-colMeans(cm==0)
summary(pz_r1)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  0.330   0.632   0.744   0.735   0.844   0.984 </code></pre>
<pre class="r"><code>numgene_r1 &lt;- colSums(cm!=0)
summary(numgene_r1)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      8      78     128     133     184     335 </code></pre>
<pre class="r"><code># Build the entire summary as one string
output &lt;- paste0(
  &quot;\n================= Summary Statistics =================\n\n&quot;,
  &quot;--- MERSCOPE human breast cancer sample---\n&quot;,
  make_sc_summary(td_r1, &quot;Total detections per cell:&quot;),
  make_sc_summary(pz_r1, &quot;Proportion of zeroes per cell:&quot;),
  make_sc_summary(numgene_r1, &quot;Detected genes per cell:&quot;),
  &quot;\n========================================================\n&quot;
)

cat(output)</code></pre>
<pre><code>
================= Summary Statistics =================

--- MERSCOPE human breast cancer sample---
Total detections per cell:
  Min: 31.0000 |  1Q: 163.0000 |  Median: 368.0000 |  Mean: 522.5778 |  3Q: 734.0000 |  Max: 2499.0000
Proportion of zeroes per cell:
  Min: 0.3300 |  1Q: 0.6320 |  Median: 0.7440 |  Mean: 0.7346 |  3Q: 0.8440 |  Max: 0.9840
Detected genes per cell:
  Min: 8.0000 |  1Q: 78.0000 |  Median: 128.0000 |  Mean: 132.6814 |  3Q: 184.0000 |  Max: 335.0000

========================================================</code></pre>
<pre class="r"><code>td_df &lt;-as.data.frame(cbind(as.vector(td_r1),
                              rep(&quot;hbreast&quot;, length(td_r1))))
colnames(td_df) &lt;- c(&quot;td&quot;,&quot;sample&quot;)
td_df$td= as.numeric(td_df$td)


p1&lt;-ggplot(data = td_df, aes(x = td, color=sample)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 100, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, linewidth = 2) +
  labs(title = &quot;Distribution of total detections per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5), 
        strip.text = element_text(size=12)) 

pz &lt;- as.data.frame(cbind(as.vector(pz_r1),rep(&quot;hbreast&quot;, length(pz_r1))))
colnames(pz) &lt;- c(&quot;prop_ze&quot;,&quot;sample&quot;)
pz$prop_ze&lt;- as.numeric(pz$prop_ze)

p2&lt;-ggplot(data = pz, aes(x = prop_ze)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 0.05, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, linewidth = 2) +
  labs(title = &quot;Distribution of proportion of zeroes per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))


numgens &lt;- as.data.frame(cbind(as.vector(numgene_r1),rep(&quot;hbreast&quot;, length(numgene_r1))))
colnames(numgens) &lt;- c(&quot;numgen&quot;,&quot;sample&quot;)
numgens$numgen&lt;- as.numeric(numgens$numgen)

p3&lt;-ggplot(data = numgens, aes(x = numgen)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 10, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, linewidth = 2) +
  labs(title = &quot;Distribution of detected genes per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))</code></pre>
<pre class="r"><code>cm_new1&lt;-as.matrix(cm)
cm_new1[cm_new1==0] &lt;- NA
#cm_new1 &lt;- as.data.frame(cm_new1)
avg2_vals &lt;- rowMeans(cm_new1,na.rm = TRUE)
summary(avg2_vals)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.02    1.22    1.61    2.47    2.64   24.25 </code></pre>
<pre class="r"><code>avg_exp &lt;- as.data.frame(cbind(&quot;avg&quot;=avg2_vals, 
                             &quot;sample&quot;=rep(&quot;hliver_cancer&quot;, nrow(cm_new1))))

avg_exp$avg&lt;-as.numeric(avg_exp$avg)


p4&lt;-ggplot(data = avg_exp, aes(x = avg, color=sample)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 1, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;orange&quot;, linewidth = 1) +
  labs(title = &quot;Distribution of average gene expression per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +

  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))


layout_design &lt;- (p1|p2)/(p3|p4)

layout_design</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/ave-expression-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-ave-expression-1">
Past versions of ave-expression-1.png
</button>
</p>
<div id="fig-ave-expression-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/ave-expression-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/ave-expression-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The plots show that most cells have relatively low total detection
counts, with a long right tail corresponding to transcriptionally rich
cells. The proportion of zero entries per cell peaks around 70–80% and
the number of detected genes per cell displays a broad unimodal
distribution. The average expression per cell is right-skewed, with a
small subset of cells showing high mean transcript abundance.</p>
</div>
<div id="negative-control-coordinates" class="section level1">
<h1>Negative control coordinates</h1>
<p>This section inspects negative control probes (“blank genes”) to
assess nonspecific hybridization and technical background. The first
plot visualizes their spatial distribution across the tissue, and the
second quantifies per-probe detection counts.</p>
<p>The barplot reveals variation in detection counts among blank probes,
but overall counts remain low relative to endogenous transcripts.</p>
<pre class="r"><code>real_genes &lt;- row.names(se)
nc_coords &lt;- transcript_df[!(transcript_df$gene %in% real_genes), ]

cat(&quot;#negative controls&quot;, length(unique((nc_coords$gene))), &quot;\n&quot;)</code></pre>
<pre><code>#negative controls 50 </code></pre>
<pre class="r"><code>nc_tb &lt;- as.data.frame(sort(table(nc_coords$feature_name), 
                           decreasing = TRUE))
colnames(nc_tb) &lt;- c(&quot;name&quot;,&quot;value_count&quot;)
nc_coords$feature_name &lt;- nc_coords$gene
nc_coords$feature_name &lt;-factor(nc_coords$feature_name, levels = nc_tb$name)

ggplot(nc_tb, aes(x = name, y = value_count)) +
    geom_bar(stat = &quot;identity&quot;) + 
    #geom_text(aes(label = value_count),  vjust = -0.5, size =1.5) + 
    scale_y_continuous(expand = c(0,1), limits = c(0, 155000))+
    labs(title = &quot; &quot;, x = &quot; &quot;, y = &quot;Number of total detections&quot;) +
    theme_bw()+
    defined_theme+
    theme(axis.text.x= element_text(size=10,vjust=0.5,hjust = 1, angle = 90),
          axis.text.y=element_text(size=10))</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/negprobes-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-negprobes-1">
Past versions of negprobes-1.png
</button>
</p>
<div id="fig-negprobes-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/negprobes-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/negprobes-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="spatial-patterns-of-blank-genes" class="section level2">
<h2>Spatial patterns of blank genes</h2>
<p>The spatial heatmap shows that blank probe detections are uniformly
distributed with no spatial clustering, indicating minimal spatial
artefacts or region-specific nonspecific binding</p>
<pre class="r"><code>ggplot(nc_coords, aes(x=x, y=y))+
    geom_hex(bins=auto_hex_bin(n=nrow(nc_coords)))+
    theme_bw()+
    scale_fill_gradient(low=&quot;white&quot;, high=&quot;maroon4&quot;) + 
    ggtitle(&quot;Blank genes detections&quot;)+
    defined_theme+
    theme(legend.position = &quot;top&quot;,
          plot.title = element_text(size=rel(1.3)),
          aspect.ratio = 10/12)</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/blankgenes-vis-xy-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-blankgenes-vis-xy-1">
Past versions of blankgenes-vis-xy-1.png
</button>
</p>
<div id="fig-blankgenes-vis-xy-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/blankgenes-vis-xy-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/blankgenes-vis-xy-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="banksy-clustering" class="section level1">
<h1>Banksy clustering</h1>
<p>We performed spatially informed clustering using Banksy, which
integrates gene expression profiles with local spatial context through
adaptive geometric features (k = 15, λ = 0.2). Clustering was performed
at a resolution of 0.5, providing 10 clusters in total.</p>
<pre class="r"><code>seu &lt;- as.Seurat(se, data = NULL)
seu &lt;- NormalizeData(seu, normalization.method = &quot;LogNormalize&quot;)
seu &lt;- FindVariableFeatures(seu, nfeatures = nrow(seu))

# copy log-normalized data back to se
aname &lt;- &quot;logcounts&quot;
logcounts_mat &lt;- GetAssayData(seu, slot = &quot;data&quot;)[, colnames(se)]
assay(se, aname, withDimnames = FALSE) &lt;- logcounts_mat

lambda &lt;- 0.2
k_geom &lt;- 15
use_agf &lt;- TRUE

se &lt;- computeBanksy(se, assay_name = aname,
                    compute_agf = TRUE, k_geom = k_geom)

se &lt;- runBanksyPCA(se, use_agf = use_agf, lambda = lambda, seed = 1000)
se &lt;- runBanksyUMAP(se, use_agf = use_agf, lambda = lambda, seed = 1000)

cat(&quot;Clustering starts\n&quot;)
se &lt;- clusterBanksy(se, use_agf = use_agf, lambda = lambda,
                    resolution = c(0.5, 0.8), seed = 1000)

se &lt;- connectClusters(se)


cluster_info &lt;- data.frame(
    x = spatialCoords(se)[, 1],
    y = spatialCoords(se)[, 2],
    cell_id = colnames(se), 
    cluster = paste0(&quot;c&quot;, colData(se)$clust_M1_lam0.2_k50_res0.5),  
    sample =&quot;sample01&quot;
)

cluster_info$cluster = factor(cluster_info$cluster,
                               levels = paste0(&quot;c&quot;,sort(unique(colData(se)$clust_M1_lam0.2_k50_res0.5))))

saveRDS(cluster_info, here(gdata,paste0(data_nm, &quot;_clusters.Rds&quot;)))
saveRDS(se, here(gdata,paste0(data_nm, &quot;_se.Rds&quot;)))
saveRDS(seu, here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))</code></pre>
<pre class="r"><code># load generated data  (see code above for how they were created)
cluster_info = readRDS(here(gdata,paste0(data_nm, &quot;_clusters.Rds&quot;)))
cluster_names = paste0(&quot;c&quot;, 1:10)


seu = readRDS(here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))
seu &lt;- subset(seu, cells = cluster_info$cell_id)

Idents(seu)=cluster_info$cluster[match(colnames(seu), cluster_info$cell_id)]
Idents(seu) = factor(Idents(seu), levels = cluster_names)
seu$sample = cluster_info$sample[match(colnames(seu), cluster_info$cell_id)]


fig_ratio = cluster_info %&gt;%
  group_by(sample) %&gt;%
  summarise(
    x_range = diff(range(x, na.rm = TRUE)),
    y_range = diff(range(y, na.rm = TRUE)),
    ratio   = y_range / x_range,
    .groups = &quot;drop&quot;
  ) %&gt;% summarise(max_ratio = max(ratio)) %&gt;%  pull(max_ratio)</code></pre>
<pre class="r"><code>ggplot(data = cluster_info,
        aes(x = x, y = y, color=cluster))+
        geom_point(size=0.0001)+
        facet_wrap(~sample)+
        theme_classic()+
        guides(color=guide_legend(title=&quot;&quot;, nrow = 2,
        override.aes=list(alpha=1, size=6)))+
        defined_theme+
        theme(aspect.ratio = fig_ratio,
              legend.position = &quot;bottom&quot;,
              strip.text = element_blank())</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/cluster-xy-whole-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cluster-xy-whole-1">
Past versions of cluster-xy-whole-1.png
</button>
</p>
<div id="fig-cluster-xy-whole-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/cluster-xy-whole-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/cluster-xy-whole-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The following figure shows spatial maps of the ten identified Banksy
clusters. Each panel displays the spatial localization of a cluster
across the tissue section. Clusters c1–c5 form broad contiguous domains
that align with morphologically distinct regions, while c6–c10 appear
more localized, marking smaller or transitional cell populations. The
clear spatial separation between major clusters indicates that Banksy
effectively captures structured tissue organization.</p>
<pre class="r"><code>ggplot(data = cluster_info,
            aes(x = x, y = y))+
    geom_hex(bins = 300)+
    facet_wrap(~cluster, nrow=4)+
    theme_classic()+
    scale_fill_gradient(low=&quot;grey50&quot;, high=&quot;red&quot;) + 
    guides(color=guide_legend(title=&quot;&quot;,override.aes=list(alpha=1, size=8)))+
        defined_theme+theme(legend.position = &quot;right&quot;, 
                            strip.text = element_text(size=12))</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/cluster-xy-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cluster-xy-1">
Past versions of cluster-xy-1.png
</button>
</p>
<div id="fig-cluster-xy-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/cluster-xy-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/cluster-xy-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This code summarizes the number of cells assigned to each Banksy
cluster at resolution 0.5. The bar chart shows that clusters c1–c4
contain the majority of cells, indicating dominant tissue compartments,
while clusters c5–c10 represent progressively smaller populations. The
presence of several low-frequency clusters suggests distinct but less
abundant spatially localized cell states.</p>
<pre class="r"><code>sum_df &lt;- as.data.frame(table(cluster_info$cluster))
colnames(sum_df) = c(&quot;cellType&quot;,&quot;n&quot;)
ggplot(sum_df, aes(x = cellType, y = n, fill=n)) +
    geom_bar(stat = &quot;identity&quot;, position=&quot;stack&quot;, fill=&quot;black&quot;) +
    geom_text(aes(label = n),size=3, 
              position = position_dodge(width = 0.9), vjust = -0.5) + 
   scale_y_continuous(expand = c(0,1), limits=c(0,220000))+
    labs(title = &quot; &quot;, x = &quot; &quot;, y = &quot;number of cells&quot;, fill=&quot;cell type&quot;) +
    theme_bw()+
    theme(legend.position = &quot;none&quot;,
          axis.text= element_text(size=15,vjust=0.5,hjust = 0.5),
          strip.text = element_text(size = rel(1)),
          axis.line=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title=element_text(size=12),
          panel.background=element_blank(),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),
          plot.background=element_blank())</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/ct-prop-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-ct-prop-1">
Past versions of ct-prop-1.png
</button>
</p>
<div id="fig-ct-prop-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/ct-prop-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/ct-prop-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="detect-marker-genes" class="section level1">
<h1>Detect marker genes</h1>
<div id="jazzpanda" class="section level2">
<h2>jazzPanda</h2>
<p>We applied jazzPanda to this dataset using both the linear modelling
and correlation-based approaches to detect spatially enriched marker
genes across clusters.</p>
<div id="linear-modelling-approach" class="section level3">
<h3>Linear modelling approach</h3>
<p>We first used the linear modelling approach to identify spatially
enriched marker genes. The dataset was converted into spatial gene and
cluster vectors using a 50 × 50 square binning strategy. Transcript
detections for the negative cotnrol Blank genes were included as the
background control. For each gene, a linear model was fitted to estimate
its spatial association with relevant cell type patterns.</p>
<pre class="r"><code>grid_length =50
hbreast_vector_lst&lt;-get_vectors(x=list(&quot;sample01&quot; = transcript_df), 
                                sample_names = &quot;sample01&quot;,
                                cluster_info = cluster_info,
                                bin_type=&quot;square&quot;,
                                bin_param=c(grid_length,grid_length),
                                test_genes = row.names(se),
                                n_cores = 5)

### linear modelling approach 
nc_names &lt;- unique(nc_coords$feature_name)
nc_coords$sample &lt;- &quot;sample01&quot;
kpt_cols &lt;- c(&quot;x&quot;,&quot;y&quot;,&quot;feature_name&quot;,&quot;sample&quot;,&quot;barcode_id&quot;)
nc_coords_mapped = as.data.frame(nc_coords)[,kpt_cols]


nc_vectors &lt;- create_genesets(x=list(&quot;sample01&quot; = nc_coords), 
                              sample_names = &quot;sample01&quot;,
                              name_lst=list(blanks=nc_names),
                              bin_type=&quot;square&quot;,
                              bin_param=c(grid_length, grid_length),
                              cluster_info = NULL)
saveRDS(nc_vectors, &quot;nc_vectors.Rds&quot;)
set.seed(589)
cat(&quot;Running lasso_markers \n&quot;)

jazzPanda_res_lst &lt;- lasso_markers(gene_mt=hbreast_vector_lst$gene_mt,
                                   cluster_mt = hbreast_vector_lst$cluster_mt,
                                   sample_names=c(&quot;sample01&quot;),
                                   keep_positive=TRUE,
                                   background=nc_vectors,
                                   n_fold = 10)



saveRDS(hbreast_vector_lst, here(gdata,paste0(data_nm, &quot;_sq50_vector_lst.Rds&quot;)))
saveRDS(jazzPanda_res_lst, here(gdata,paste0(data_nm, &quot;_jazzPanda_res_lst.Rds&quot;)))</code></pre>
<p>We then used <code>get_top_mg</code> to extract unique marker genes
for each cluster based on a coefficient cutoff of 0.1. This function
returns the top spatially enriched genes that are most specific to each
cluster. In contrast, if shared marker genes across clusters are of
interest, all significant genes can be retrieved using
<code>get_full_mg</code>.</p>
<pre class="r"><code># load from saved objects
jazzPanda_res_lst = readRDS(here(gdata,paste0(data_nm, &quot;_jazzPanda_res_lst.Rds&quot;)))
sv_lst = readRDS(here(gdata,paste0(data_nm, &quot;_sq50_vector_lst.Rds&quot;)))
nbins = 2500
jazzPanda_res = get_top_mg(jazzPanda_res_lst, coef_cutoff=0.1) </code></pre>
<div id="cluster-to-cluster-correlation" class="section level4">
<h4>Cluster to cluster correlation</h4>
<p>The heatmap shows several strongly correlated clusters, notably
c1–c3–c4, suggesting they represent transcriptionally related epithelial
compartments. In contrast, clusters such as c2 and c7–c9 display weak or
negative correlations with others, indicating distinct expression
profiles.</p>
<pre class="r"><code>cluster_mtt = sv_lst$cluster_mt
cluster_mtt = cluster_mtt[,cluster_names]
cor_M_cluster = cor(cluster_mtt,cluster_mtt,method = &quot;pearson&quot;)
col &lt;- colorRampPalette(c(&quot;#4477AA&quot;, &quot;#77AADD&quot;, &quot;#FFFFFF&quot;,&quot;#EE9988&quot;, &quot;#BB4444&quot;))

corrplot(
  cor_M_cluster,
  method      = &quot;color&quot;,
  col         = col(200),
  diag        = TRUE,
  addCoef.col = &quot;black&quot;,
  type        = &quot;upper&quot;,
  tl.col      = &quot;black&quot;,
  tl.cex      = 1,
  number.cex  = 0.8,
  tl.srt      = 0,
  mar         = c(0, 0, 3, 0),   
  sig.level   = 0.05,
  insig       = &quot;blank&quot;,
  title       = &quot;cluster–cluster correlation&quot;, 
  cex.main    = 1.2
)</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/cc-corr-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cc-corr-1">
Past versions of cc-corr-1.png
</button>
</p>
<div id="fig-cc-corr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/cc-corr-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/cc-corr-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>We can plot the top 10 spatial marker genes per cluster identified by
the linear modelling approach.</p>
<pre class="r"><code>glm_mg = jazzPanda_res[jazzPanda_res$top_cluster != &quot;NoSig&quot;, ]
glm_mg$top_cluster = factor(glm_mg$top_cluster, levels = cluster_names)
top_mg_glm &lt;- glm_mg %&gt;%
  group_by(top_cluster) %&gt;%
  arrange(desc(glm_coef), .by_group = TRUE) %&gt;% 
  slice_max(order_by = glm_coef, n = 10) %&gt;%
  select(top_cluster, gene, glm_coef)

p1&lt;-DotPlot(seu, features = top_mg_glm$gene) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (jazzPanda-glm)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top10_glm-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top10_glm-1">
Past versions of top10_glm-1.png
</button>
</p>
<div id="fig-top10_glm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top10_glm-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top10_glm-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="top3-marker-genes-by-jazzpanda_glm" class="section level4">
<h4>Top3 marker genes by jazzPanda_glm</h4>
<p>We visualized the top three marker genes identified by the linear
modelling approach for each cluster. For each selected gene, its spatial
transcript locations were plotted alongside the corresponding cluster
map. Across clusters, the spatial expression patterns of marker genes
closely match the distribution of their associated cell types</p>
<pre class="r"><code>for (cl in cluster_names){
    inters=jazzPanda_res[jazzPanda_res$top_cluster==cl,&quot;gene&quot;]
    rounded_val=signif(as.numeric(jazzPanda_res[inters,&quot;glm_coef&quot;]),
                          digits = 3)
    inters_df = as.data.frame(cbind(gene=inters, value=rounded_val))
    inters_df$value = as.numeric(inters_df$value)
    inters_df=inters_df[order(inters_df$value, decreasing = TRUE),]
    inters_df$text= paste(inters_df$gene,inters_df$value,sep=&quot;: &quot;)
        
    if (length(inters &gt; 0)){
        inters_df = inters_df[1:min(3, nrow(inters_df)),]
        inters = inters_df$gene
        iters_sp1= transcript_df$feature_name %in% inters
        vis_r1 =transcript_df[iters_sp1,
                        c(&quot;x&quot;,&quot;y&quot;,&quot;feature_name&quot;)]
        vis_r1$value = inters_df[match(vis_r1$feature_name,inters_df$gene),
                        &quot;value&quot;]
        vis_r1=vis_r1[order(vis_r1$value,decreasing = TRUE),]
        vis_r1$text_label= paste(vis_r1$feature_name,
                            vis_r1$value,sep=&quot;: &quot;)
        vis_r1$text_label=factor(vis_r1$text_label, levels=inters_df$text)
    
        vis_r1 = vis_r1[order(vis_r1$text_label),]

        genes_plt &lt;- plot_top3_genes(vis_df=vis_r1, fig_ratio=fig_ratio)

        cluster_data =  cluster_info[cluster_info$cluster==cl, ]
        cl_pt&lt;- ggplot(data =cluster_data,
                    aes(x = x, y = y, color=sample))+ 
                    geom_point(size=0.01)+
                    facet_wrap(~cluster, nrow=1,strip.position = &quot;left&quot;)+
                    scale_color_manual(values = c(&quot;black&quot;))+
                    defined_theme + 
            theme(legend.position = &quot;none&quot;, 
                              strip.background = element_blank(), 
                              aspect.ratio = fig_ratio,
                              legend.key.width = unit(15, &quot;cm&quot;),
                              strip.text = element_text(size = rel(1.3)))
        layout_design &lt;- (wrap_plots(cl_pt, ncol = 1) | genes_plt) +
            plot_layout(widths = c(1, 3), heights = c(1))
        
        print(layout_design)
    }
  }</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-1.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-1">
Past versions of top3-gene-vis-xy-glm-1.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-2.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-2">
Past versions of top3-gene-vis-xy-glm-2.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-2.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-2.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-3.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-3">
Past versions of top3-gene-vis-xy-glm-3.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-3.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-3.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-4.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-4">
Past versions of top3-gene-vis-xy-glm-4.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-4.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-4.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-5.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-5">
Past versions of top3-gene-vis-xy-glm-5.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-5" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-5.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-5.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-6.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-6">
Past versions of top3-gene-vis-xy-glm-6.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-6" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-6.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-6.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-7.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-7">
Past versions of top3-gene-vis-xy-glm-7.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-7" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-7.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-7.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-8.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-8">
Past versions of top3-gene-vis-xy-glm-8.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-8" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-8.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-8.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-9.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-glm-9">
Past versions of top3-gene-vis-xy-glm-9.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-glm-9" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-9.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-glm-9.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="cluster-vector-versus-marker-gene-vector"
class="section level4">
<h4>Cluster vector versus marker gene vector</h4>
<p>We visualized the top three marker genes (linear modelling approach)
for each cluster by plotting their spatial gene vectors against the
corresponding cluster vector. Each scatter plot shows how strongly each
marker gene’s spatial pattern aligns with the overall spatial signal of
its cluster, allowing assessment of both marker specificity and
coherence with spatial domains.</p>
<p>Most top marker genes show strong positive correlations with their
cluster vectors, confirming spatial alignment between gene expression
and cluster structure. For example, MYC, TAPBP, and CD40 (c1) display
clear linear trends, reflecting high spatial coherence within the tumor
region. Similarly, COL1A1 and ACTA2 (c2) exhibit broad, continuous
gradients consistent with fibroblast-rich zones. Immune-related markers
such as TRAC, CD3E, and CCL5 (c7) show distinct yet compact
distributions, characteristic of localized immune niches.</p>
<pre class="r"><code>plot_lst=list()
for (cl in cluster_names){
    # ct_nm = anno_df[anno_df$cluster==cl, &quot;anno_name&quot;]
    inters_df=jazzPanda_res[jazzPanda_res$top_cluster==cl,]
    
    if (nrow(inters_df) &gt;0){
        inters_df=inters_df[order(inters_df$glm_coef, decreasing = TRUE),]
        
        inters_df = inters_df[1:min(3, nrow(inters_df)),]
        mk_genes = inters_df$gene
        dff = as.data.frame(cbind(sv_lst$cluster_mt[,cl],
                              sv_lst$gene_mt[,mk_genes]))
        colnames(dff) = c(&quot;cluster&quot;, mk_genes)
        dff$vector_id = c(1:nbins)
        long_df &lt;- dff %&gt;% 
        tidyr::pivot_longer(cols = -c(cluster, vector_id), names_to = &quot;gene&quot;, 
               values_to = &quot;vector_count&quot;)
        long_df$gene = factor(long_df$gene, levels=mk_genes)
        p&lt;-ggplot(long_df, aes(x = cluster, y = vector_count, color=gene )) +
        geom_point(size=0.01) +
        facet_wrap(~gene, scales=&quot;free_y&quot;, nrow=1) +
        labs(x = paste(cl,&quot; cluster vector &quot;, sep=&quot;&quot;), 
         y = &quot;Top3 marker gene vector&quot;) +
        theme_minimal()+
        scale_y_continuous(expand = c(0.01,0.01))+ 
        scale_x_continuous(expand =  c(0.01,0.01))+ 
        theme(panel.grid = element_blank(),legend.position = &quot;none&quot;,
        strip.text = element_text(size = 12),
        axis.line=element_blank(),
        axis.text=element_text(size = 11),
        axis.ticks=element_line(color=&quot;black&quot;),
        axis.title=element_text(size = 12),
        panel.border  =element_rect(colour = &quot;black&quot;, 
                                    fill=NA, linewidth=0.5)
        )
        if (length(mk_genes) == 2) {
            p &lt;- (p | plot_spacer()) + 
            plot_layout(widths = c(2, 1))
        } else if (length(mk_genes) == 1) {
            p &lt;- (p | plot_spacer() | plot_spacer()) + 
            plot_layout(widths = c(1, 1, 1))
        } 
        
        plot_lst[[cl]] = p
        
    }
    
}
combined_plot &lt;- wrap_plots(plot_lst, ncol = 2)

combined_plot </code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/vvplot-glm-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-vvplot-glm-1">
Past versions of vvplot-glm-1.png
</button>
</p>
<div id="fig-vvplot-glm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/vvplot-glm-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="marker-gene-summary-per-cluster" class="section level4">
<h4>Marker gene summary per cluster</h4>
<p>c1: High MYC, EPCAM, ERBB2, and FOXM1 indicate proliferative tumor
epithelium with active oncogenic and stress-response signaling.</p>
<p>c2: Strong FN1, COL1A1, ACTA2, and TGFBR2 expression indicates
matrix-remodeling and fibroblast activation typical of CAFs.</p>
<p>c3: CDH1, HLA-DRB1, and CIITA indicate epithelial cells with
antigen-presentation and partial EMT features.</p>
<p>c4: No clear markers detected, suggesting a mixed or boundary region
with low transcriptional contrast.</p>
<p>c5: C1QC, LYZ, FCGR3A, and MRC1 indicate macrophages with an M2-like
immunosuppressive phenotype.</p>
<p>c6: SPP1, VEGFA, and SOD2 indicate hypoxia-responsive,
pro-angiogenic, and ECM-remodeling activity.</p>
<p>c7: TRAC, CD3E, CD8A, and CCL5 indicate T cells with mixed cytotoxic
and regulatory signatures.</p>
<p>c8: COL4A1, PECAM1, PDGFRB, and VWF indicate vascular endothelial and
perivascular stromal identity.</p>
<p>c9: XBP1, MZB1, IRF4, and CD79A indicate antibody-secreting plasma
cells.</p>
<p>c10: CD207, FCER1A, and CD1C indicate antigen-presenting dendritic
cells involved in immune surveillance.</p>
<pre class="r"><code>## linear modelling 
output_data &lt;- do.call(rbind, lapply(cluster_names, function(cl) {
  subset &lt;- jazzPanda_res[jazzPanda_res$top_cluster == cl, ]
  sorted_subset &lt;- subset[order(-subset$glm_coef), ]
  if (length(sorted_subset) == 0){
    return(data.frame(Cluster = cl, Genes = &quot;  &quot;))
  }
  if (cl != &quot;NoSig&quot;){
    gene_list &lt;- paste(sorted_subset$gene, &quot;(&quot;, round(sorted_subset$glm_coef, 2), &quot;)&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }else{
    gene_list &lt;- paste(sorted_subset$gene, &quot;&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }
  
  return(data.frame(Cluster = cl, Genes = gene_list))
}))

knitr::kable(
  output_data,
  caption = &quot;Detected marker genes for each cluster by linear modelling approach in jazzPanda, in decreasing value of glm coefficients&quot;
)</code></pre>
<table>
<caption>Detected marker genes for each cluster by linear modelling
approach in jazzPanda, in decreasing value of glm coefficients</caption>
<colgroup>
<col width="0%" />
<col width="99%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="left">Genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">c1</td>
<td align="left">MYC(45.66), TAPBP(29), CD40(23.64), EPCAM(18.03),
MMP7(15.08), CEBPB(14.42), FZD7(12.88), CDKN1B(12.34), DKK1(12.06),
LRP6(11.86), PTK2(11.39), STAT6(9.03), SOX9(8.76), FOXM1(8.69),
IFNGR1(8.68), ERBB2(8.43), TP53(8.28), NFE2L2(8.16), JUNB(8.13),
EPHB3(7.81), BRD4(7.57), CDK4(7.42), MCM2(7.32), MYBL2(7.29),
FGFR2(7.28), CREBBP(7.08), PCNA(6.69), ERBB3(6.01), EPHB4(5.79),
NFKB1(5.65), CXCL10(5.3), KRAS(5.25), NFKBIA(5.22), SMO(4.98),
STAT3(4.89), SRPRB(4.73), IFNGR2(4.69), DIABLO(4.66), SMARCA4(3.95),
SHARPIN(3.86), ICOSLG(3.78), LGR5(3.59), TCF7L2(3.55), SMAD2(3.4),
CEACAM1(3.12), ATR(2.73), RAF1(2.66), BAX(2.55), SRC(2.33), MLH1(2.26),
CCNB1(2.26), IRS1(2.25), LRP5(2.22), CDK2(2.14), IRF3(2.1), PRKCA(2.1),
SOX2(2.05), IKBKB(2.04), BRAF(2.01), NF1(1.95), CX3CL1(1.82),
MSH3(1.79), APC(1.78), CHEK2(1.51), TBX3(1.47), KIT(1.41), AMOTL2(1.31),
CASP8(1.25), RET(1.22), DNMT1(1.07), SLC13A3(1.05), CCNE1(0.96),
IFNAR2(0.9), AKT2(0.81), RGMB(0.76), WNT3(0.56), ARAF(0.53),
PROX1(0.51), ATM(0.45), ARG1(0.36), CSF2(0.35), PKIB(0.33), CD160(0.25),
FGF2(0.24)</td>
</tr>
<tr class="even">
<td align="left">c2</td>
<td align="left">FN1(79.49), COL1A1(51.85), ACTA2(10.65), COL11A1(9.79),
COL5A1(8.87), LRP1(5.77), CXCL12(3.78), BST2(3.45), DUSP1(2.93),
FGFR1(2.55), TNC(2.32), CSF1(2.21), CCND1(2.1), PDGFRA(2.03),
CDKN1A(1.97), TGFBR2(1.89), LOX(1.17), TEAD1(1.01), IFITM1(0.98),
SMOC2(0.96), PDPN(0.81), COL6A3(0.76), MFAP5(0.71), FGF1(0.6),
VEGFC(0.6), CMKLR1(0.57), MMP2(0.47), ELN(0.38), NCAM1(0.32),
TWIST1(0.2)</td>
</tr>
<tr class="odd">
<td align="left">c3</td>
<td align="left">CDH1(26.98), HLA.DRB1(10.33), HLA.DPB1(9.52),
CIITA(8.38), LAMB3(7.17), TNFSF10(4.36), KITLG(3.66), IDO1(3.17),
TGFB2(2.9), ICAM2(2.2), SNAI1(1.57), MET(1.28), PTEN(0.93), CD177(0.36),
TMEM59(0.32)</td>
</tr>
<tr class="even">
<td align="left">c4</td>
<td align="left">()</td>
</tr>
<tr class="odd">
<td align="left">c5</td>
<td align="left">C1QC(24.02), LYZ(23.33), FCGR3A(19.04), CD14(15.38),
CYBB(11.97), ITGB2(7.74), CSF1R(7.7), FOS(7.64), MRC1(5.67),
FCGR2A(4.79), CCL2(4.2), CCR1(4.03), SERPINA1(3.65), CD4(3.53),
MAFB(3.28), LGALS9(3.25), CD163(3.23), PDK4(3.14), FBLN1(2.77),
MSR1(2.65), HAVCR2(2.26), CD209(1.99), HLA.DPA1(1.8), TLR2(1.79),
CD86(1.45), CCL8(1.39), SIGLEC1(1.32), TREM2(1.21), TLR1(1.2),
CSF3R(0.92), VSIR(0.91), CSF2RA(0.78), LGR6(0.67), CCR2(0.51),
CLEC4C(0.42), CD80(0.37), CX3CR1(0.36), IRF5(0.33), CXCL2(0.25)</td>
</tr>
<tr class="even">
<td align="left">c6</td>
<td align="left">SPP1(70.89), VEGFA(43.68), SOD2(22.66), MMP9(7.88),
SERPINE1(6.39), TGFBI(6.28), JUN(5.84), ICAM1(4.3), MMP12(3.68),
IL4I1(2.88), CXCL8(2.46), PDK1(2.35), LDHA(2.12), ATF3(1.96),
SOCS3(1.42), ITGAX(0.93), CXCL16(0.9), PLOD2(0.84), IL6(0.75),
MARCO(0.65), PTGS2(0.63), S100A9(0.44), CCL3(0.38), MUC1(0.38),
CA9(0.33), CXCL5(0.24), CD22(0.18), CCL7(0.11)</td>
</tr>
<tr class="odd">
<td align="left">c7</td>
<td align="left">TRAC(7.18), CCL5(4.66), CD3E(4.34), CD2(3.74),
ZAP70(3.18), PTPRC(3.16), MMP1(3.15), CD8A(2.87), IL2RB(2.58),
CXCL9(2.29), CCR7(2.11), CD5(1.97), GPX3(1.79), CD3D(1.68), GZMA(1.55),
ITGA4(1.43), LAMP3(1.35), ITK(1.32), TIGIT(1.29), GNLY(1.21),
CCR4(1.18), STAT4(1.13), TRBC1(1.13), GATA3(1.1), KLRK1(1.08),
CTSW(1.04), STAT5A(0.98), FOXP3(0.96), STING1(0.96), GZMK(0.89),
CD28(0.89), NKG7(0.89), SELL(0.86), CXCR3(0.79), CD27(0.72), CCR5(0.64),
KLRG1(0.64), CXCR6(0.63), CD3G(0.63), EOMES(0.6), KLF2(0.57),
PDCD1LG2(0.53), TBX21(0.5), TRAT1(0.45), CD40LG(0.44), SELPLG(0.43),
ICOS(0.37), NLRC5(0.36), GZMH(0.35), PDCD1(0.33), LAG3(0.32),
PRF1(0.31), CCL4(0.28), KLRB1(0.26), FASLG(0.26), CXCL11(0.23),
CD226(0.22), CXCR5(0.18)</td>
</tr>
<tr class="even">
<td align="left">c8</td>
<td align="left">COL4A1(32.39), PLVAP(19.59), PDGFRB(16.87),
PECAM1(15.37), VWF(12.14), MMP11(10.92), PGF(6.9), INSR(6.72),
CDH5(6.28), NRP1(5.3), TGFB3(5.15), ENG(5.11), KDR(4.66), NOTCH1(4.56),
PDGFB(4.08), CD248(4.01), WWTR1(3.89), ANGPT2(3.89), MMRN2(3.67),
TGFB1(3.48), CAV1(3.3), NDUFA4L2(3.2), BMP1(2.94), ZEB1(2.78),
ACKR3(2.69), SNAI2(2.64), ITGA1(2.4), FLT4(2.39), CLEC14A(2.08),
CLDN5(1.71), PREX2(1.12), FLT1(1.02), TNFRSF4(0.96), NOS3(0.86),
IL3RA(0.63), ANGPT1(0.4), PAX5(0.29), DUSP6(0.28), HGF(0.28),
LEF1(0.23), TEK(0.2), CD68(0.15)</td>
</tr>
<tr class="odd">
<td align="left">c9</td>
<td align="left">XBP1(53.96), MZB1(10.63), IRF4(7.47), POU2AF1(6.65),
CD79A(6.41), FCRL5(5.16), DERL3(3.59), ICAM3(2.4), TNFRSF13C(1.38),
CD79B(1.1), MS4A1(0.96), GZMB(0.73), CTSG(0.52), CD19(0.45),
STAP1(0.37), BLK(0.35), XCR1(0.18)</td>
</tr>
<tr class="even">
<td align="left">c10</td>
<td align="left">CD207(47.96), FCER1A(7.37), CD1C(3.31), CD1E(2.29),
CCL22(2.1), CD1B(0.85)</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="correlation-approach" class="section level3">
<h3>Correlation approach</h3>
<p>We next applied the correlation-based method using the same 50 × 50
binning strategy. This approach highlights genes whose spatial patterns
are highly correlated with specific cell type domains.</p>
<pre class="r"><code>grid_length = 50
all_genes_names = row.names(se)
cat(&quot;Running compute_permp \n&quot;)
seed_number&lt;-589
set.seed(seed_number)
perm_p &lt;- compute_permp(x=list(&quot;sample01&quot; = transcript_df),
                            cluster_info=cluster_info, 
                            perm.size=5000,
                            bin_type=&quot;square&quot;,
                            bin_param=c(grid_length,grid_length),
                            test_genes=all_genes_names,
                            correlation_method = &quot;spearman&quot;, 
                            n_cores=5, 
                            correction_method=&quot;BH&quot;)

saveRDS(perm_p, here(gdata,paste0(data_nm, &quot;_perm_lst.Rds&quot;)))</code></pre>
<p>We can get the correlation between every gene and cluster using the
<code>get_cor</code> function, which quantifies how strongly each gene’s
spatial expression pattern aligns with the spatial distribution of each
cluster.</p>
<p>To assess statistical significance, permutation testing results are
retrieved from the saved perm_lst object. The corresponding raw p-values
can be obtained with <code>get_perm_p</code>, and the
multiple-testing–adjusted p-values with <code>get_perm_adjp</code>.
These adjusted p-values help identify genes that are significantly
associated with specific spatial domains after controlling for random
spatial correlations.</p>
<pre class="r"><code># load from saved objects
perm_lst = readRDS(here(gdata,paste0(data_nm, &quot;_perm_lst.Rds&quot;)))
perm_res = get_perm_adjp(perm_lst)
obs_corr = get_cor(perm_lst)</code></pre>
<p>The follwoing figure shows the top 10 marker genes per cluster
detected by the correlation-based approach. Distinct expression peaks
are evident across clusters: c1 and c2 display strong, localized
expression of structural and signaling genes; c5 shows broad enrichment
of macrophage-associated markers; c7 exhibits a concentrated signal for
T-cell–related genes (TRAC, CD3E, GZMA); and c8–c9 show tight,
cluster-specific expression consistent with vascular and plasma cell
domains. In contrast, c3, c4, and c6 display weaker or more diffuse
profiles, suggesting less pronounced cluster-specific transcription.</p>
<pre class="r"><code>top_mg_corr &lt;- data.frame(cluster = character(),
                          gene    = character(),
                          corr    = numeric(),
                          stringsAsFactors = FALSE)
for (cl in cluster_names) {
  obs_cutoff &lt;- quantile(obs_corr[, cl], 0.75, na.rm = TRUE)
  perm_cl &lt;- intersect(
    rownames(perm_res[perm_res[, cl] &lt; 0.05, ]),
    rownames(obs_corr[obs_corr[, cl] &gt; obs_cutoff, ])
  )
  
  if (length(perm_cl) &gt; 0) {
    selected_rows &lt;- obs_corr[perm_cl, , drop = FALSE]
    ord &lt;- order(selected_rows[, cl], decreasing = TRUE,
                 na.last = NA)
    top_genes &lt;- head(rownames(selected_rows)[ord], 5)
    top_vals  &lt;- head(selected_rows[ord, cl], 5)
    
    top_mg_corr &lt;- rbind(
      top_mg_corr,
      data.frame(cluster = rep(cl, length(top_genes)),
                 gene    = top_genes,
                 corr    = top_vals,
                 stringsAsFactors = FALSE)
    )
  }
}


p1&lt;-DotPlot(seu, features = unique(top_mg_corr$gene)) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (jazzPanda-correlation)&quot;, 
         x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top10_corr-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top10_corr-1">
Past versions of top10_corr-1.png
</button>
</p>
<div id="fig-top10_corr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top10_corr-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="top3-marker-genes-by-jazzpanda_correlation"
class="section level4">
<h4>Top3 marker genes by jazzPanda_correlation</h4>
<p>For each cluster, the top three genes with the highest significant
correlations were visualized alongside the spatial cluster map.</p>
<p>The spatial expression patterns of these marker genes closely align
with the distribution of their associated clusters, confirming that the
correlation approach captures coherent spatial co-localization between
genes and cell-type domains.</p>
<pre class="r"><code>for (cl in cluster_names){
    #ct_nm = anno_df[anno_df$cluster==cl, &quot;anno_name&quot;]
    obs_cutoff = quantile(obs_corr[, cl], 0.75)
    perm_cl&lt;-intersect(row.names(perm_res[perm_res[,cl]&lt;0.05,]),
                     row.names(obs_corr[obs_corr[, cl]&gt;obs_cutoff,]))
    inters&lt;-perm_cl
    if (length(inters &gt; 0)){
        rounded_val&lt;-signif(as.numeric(obs_corr[inters,cl]), digits = 3)
        inters_df = as.data.frame(cbind(gene=inters, value=rounded_val))
        inters_df$value = as.numeric(inters_df$value)
        inters_df&lt;-inters_df[order(inters_df$value, decreasing = TRUE),]
        inters_df = inters_df[1:min(3, nrow(inters_df)),]
        inters_df$text= paste(inters_df$gene,inters_df$value,sep=&quot;: &quot;)
        inters = inters_df$gene
        iters_sp1= transcript_df$feature_name %in% inters
        vis_r1 =transcript_df[iters_sp1,
                        c(&quot;x&quot;,&quot;y&quot;,&quot;feature_name&quot;)]
        vis_r1$value = inters_df[match(vis_r1$feature_name,inters_df$gene),
                        &quot;value&quot;]
        vis_r1=vis_r1[order(vis_r1$value,decreasing = TRUE),]
        vis_r1$text_label= paste(vis_r1$feature_name,
                            vis_r1$value,sep=&quot;: &quot;)
        vis_r1$text_label=factor(vis_r1$text_label, levels=inters_df$text)
    
        vis_r1 = vis_r1[order(vis_r1$text_label),]
        genes_plt &lt;- plot_top3_genes(vis_df=vis_r1, fig_ratio=fig_ratio)
        
        cluster_data =  cluster_info[cluster_info$cluster==cl, ]
        cl_pt&lt;- ggplot(data =cluster_data,
                    aes(x = x, y = y, color=sample))+ 
                    #geom_hex(bins = 100)+
                    geom_point(size=0.01)+
                    facet_wrap(~cluster, nrow=1,strip.position = &quot;left&quot;)+
                    scale_color_manual(values = c(&quot;black&quot;))+
                    defined_theme + theme(legend.position = &quot;none&quot;, 
                              strip.background = element_blank(), 
                              aspect.ratio = fig_ratio,
                              legend.key.width = unit(15, &quot;cm&quot;),
                              strip.text = element_text(size = rel(1.3)))
        layout_design &lt;- (wrap_plots(cl_pt, ncol = 1) | genes_plt) +
            plot_layout(widths = c(1, 3), heights = c(1))
        
        print(layout_design)
    }
  }</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-1.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-1">
Past versions of top3-gene-vis-xy-corr-1.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-1.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-2.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-2">
Past versions of top3-gene-vis-xy-corr-2.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-2.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-2.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-3.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-3">
Past versions of top3-gene-vis-xy-corr-3.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-3.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-3.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-4.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-4">
Past versions of top3-gene-vis-xy-corr-4.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-4.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-4.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-5.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-5">
Past versions of top3-gene-vis-xy-corr-5.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-5" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-5.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-5.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-6.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-6">
Past versions of top3-gene-vis-xy-corr-6.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-6" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-6.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-6.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-7.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-7">
Past versions of top3-gene-vis-xy-corr-7.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-7" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-7.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-7.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-8.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-8">
Past versions of top3-gene-vis-xy-corr-8.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-8" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-8.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-8.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-9.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-9">
Past versions of top3-gene-vis-xy-corr-9.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-9" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-9.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-9.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-10.png" width="1440" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-top3-gene-vis-xy-corr-10">
Past versions of top3-gene-vis-xy-corr-10.png
</button>
</p>
<div id="fig-top3-gene-vis-xy-corr-10" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/5b183701b021b0098a5bae0112964e534b4443b6/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-10.png" target="_blank">5b18370</a>
</td>
<td>
Melody
</td>
<td>
2026-02-17
</td>
</tr>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/top3-gene-vis-xy-corr-10.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="cluster-vector-versus-marker-gene-vector-1"
class="section level4">
<h4>Cluster vector versus marker gene vector</h4>
<p>Clusters c1–c3 exhibit very tight correlations, reflecting highly
coherent spatial domains with clear gene–cluster alignment. c5 and c7
also show strong, compact relationships. c8 and c9 display moderately
dispersed but still positive trends, suggesting more spatially
heterogeneous yet well-defined vascular and plasma cell domains. c6 and
c10 show slightly broader scatter, consistent with more diffuse or
gradient-like spatial expression.</p>
<pre class="r"><code>plot_lst=list()
for (cl in cluster_names){
    #ct_nm = anno_df[anno_df$cluster==cl, &quot;anno_name&quot;]
    obs_cutoff = quantile(obs_corr[, cl], 0.75)
    perm_cl&lt;-intersect(row.names(perm_res[perm_res[,cl]&lt;0.05,]),
                     row.names(obs_corr[obs_corr[, cl]&gt;obs_cutoff,]))
    inters&lt;-perm_cl
    if (length(inters) &gt;0){
        rounded_val&lt;-signif(as.numeric(obs_corr[inters,cl]), digits = 3)
        inters_df = as.data.frame(cbind(gene=inters, value=rounded_val))
        inters_df$value = as.numeric(inters_df$value)
        inters_df&lt;-inters_df[order(inters_df$value, decreasing = TRUE),]

        inters_df = inters_df[1:min(3, nrow(inters_df)),]
        mk_genes = inters_df$gene
        
        
        dff = as.data.frame(cbind(sv_lst$cluster_mt[,cl],
                          sv_lst$gene_mt[,mk_genes]))
        colnames(dff) = c(&quot;cluster&quot;, mk_genes)
        dff$vector_id = c(1:nbins)
        long_df &lt;- dff %&gt;% 
        tidyr::pivot_longer(cols = -c(cluster, vector_id), names_to = &quot;gene&quot;, 
           values_to = &quot;vector_count&quot;)
        long_df$gene = factor(long_df$gene, levels=mk_genes)
        p&lt;-ggplot(long_df, aes(x = cluster, y = vector_count, color=gene )) +
        geom_point(size=0.01) +
        facet_wrap(~gene, scales=&quot;free_y&quot;, nrow=1) +
        labs(x = paste(cl, &quot; cluster vector &quot;, sep=&quot;&quot;), 
        y = &quot;Top3 marker gene vector&quot;) +
        theme_minimal()+
        scale_y_continuous(expand = c(0.01,0.01))+ 
        scale_x_continuous(expand =  c(0.01,0.01))+ 
        theme(panel.grid = element_blank(),
          legend.position = &quot;none&quot;,
        strip.text = element_text(size = 12),
        axis.line=element_blank(),
        axis.text=element_text(size = 11),
        axis.ticks=element_line(color=&quot;black&quot;),
        axis.title=element_text(size = 12),
        panel.border  =element_rect(colour = &quot;black&quot;, 
                                fill=NA, linewidth=0.5)
        )
        if (length(mk_genes) == 2) {
            p &lt;- (p | plot_spacer()) + 
            plot_layout(widths = c(2, 1))
        } else if (length(mk_genes) == 1) {
            p &lt;- (p | plot_spacer() | plot_spacer()) + 
            plot_layout(widths = c(1, 1, 1))
        } 
        
        plot_lst[[cl]] = p
    }
}


combined_plot &lt;- wrap_plots(plot_lst, ncol = 2)

combined_plot </code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/vvplot-corr-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-vvplot-corr-1">
Past versions of vvplot-corr-1.png
</button>
</p>
<div id="fig-vvplot-corr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/vvplot-corr-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="marker-gene-summary-per-cluster-1" class="section level4">
<h4>Marker gene summary per cluster</h4>
<p>In the correlation-based approach, genes were selected based on two
criteria: (1) significant permutation p-values and (2) observed
correlation values above the 75th percentile for each cluster. The
resulting table lists, for each cluster, the genes with the strongest
spatial correlation to that cluster’s distribution, ordered by
decreasing correlation coefficient.</p>
<p>It is important to inspect and adjust this cutoff as needed — in some
cases, the 75th percentile threshold may still fall below 0.1, which
could include weakly correlated genes. A higher cutoff can be applied to
ensure that only strongly spatially associated genes are retained.</p>
<p>c1: High correlations for MYC, ERBB2, TP53, and EPCAM indicate strong
epithelial and proliferative programs consistent with tumor
epithelium.</p>
<p>c2: COL11A1, COL1A1, ACTA2, and TGFBR2 show robust ECM and fibroblast
activation signatures typical of stromal remodeling.</p>
<p>c3: CDH1, ERBB2, TAPBP, and PTK2 indicate epithelial differentiation
with immune-interacting and signaling activity.</p>
<p>c4: Overlap with c1–c3 markers (PTK2, RELA, SMARCA4) suggests an
epithelial-like transitional state rather than a distinct
compartment.</p>
<p>c5: CD14, C1QC, FCGR3A, and CSF1R highlight macrophage and monocyte
identity with strong myeloid polarization.</p>
<p>c6: SPP1, CXCL8, and VEGFA indicate hypoxia-driven and angiogenic
activity linked to stress or invasive niches.</p>
<p>c7: CD2, TRAC, CD3E, and GZMA mark T cells with active cytotoxic and
regulatory programs.</p>
<p>c8: PLVAP, VWF, PECAM1, and ANGPT2 define endothelial and
perivascular zones with strong angiogenic signatures.</p>
<p>c9: CD79A, MZB1, and POU2AF1 indicate antibody-secreting B-lineage
cells.</p>
<p>c10: CD207, FCER1A, and CD1C mark dendritic cells with
antigen-presentation and immune surveillance functions.</p>
<pre class="r"><code># correlation appraoch
# Combine into a new dataframe for LaTeX output
output_data &lt;- do.call(rbind, lapply(cluster_names, function(cl) {
  obs_cutoff = quantile(obs_corr[, cl], 0.75)
  perm_cl=intersect(row.names(perm_res[perm_res[,cl]&lt;0.05,]),
                   row.names(obs_corr[obs_corr[, cl]&gt;obs_cutoff,]))

  if (length(perm_cl) == 0){
    return(data.frame(Cluster = cl, Genes = &quot;  &quot;))
  }
  sorted_subset = obs_corr[row.names(obs_corr) %in% perm_cl,]
  sorted_subset &lt;- sorted_subset[order(-sorted_subset[,cl]), ]
  if (cl != &quot;NoSig&quot;){
    gene_list &lt;- paste(row.names(sorted_subset), &quot;(&quot;, round(sorted_subset[,cl], 2), &quot;)&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }else{
    gene_list &lt;- paste(row.names(sorted_subset), &quot;&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }
  
  return(data.frame(Cluster = cl, Genes = gene_list))
}))
knitr::kable(
  output_data,
  caption = &quot;Detected marker genes for each cluster by correlation approach in jazzPanda, in decreasing value of observed correlation&quot;
)</code></pre>
<table>
<caption>Detected marker genes for each cluster by correlation approach
in jazzPanda, in decreasing value of observed correlation</caption>
<colgroup>
<col width="0%" />
<col width="99%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="left">Genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">c1</td>
<td align="left">PTK2(0.98), SMARCA4(0.98), TP53(0.98), LRP6(0.98),
SHARPIN(0.98), ATR(0.98), DIABLO(0.98), TAPBP(0.98), CDKN1B(0.98),
CREBBP(0.98), KRAS(0.98), MYC(0.98), CDK4(0.98), ERBB2(0.98),
BRD4(0.98), NFE2L2(0.98), MSH6(0.97), SRPRB(0.97), BRAF(0.97),
STAT6(0.97), SMAD2(0.97), APC(0.97), STAT3(0.97), PPARD(0.97),
CHEK2(0.97), RAF1(0.97), NFKB2(0.97), SMO(0.97), MCM2(0.97),
EPHB4(0.97), DKK1(0.97), FZD7(0.97), BAK1(0.97), CASP8(0.97), SRC(0.97),
HDAC3(0.97), MSH3(0.97), CDH1(0.97), RELA(0.96), ICOSLG(0.96),
CCNE1(0.96), KITLG(0.96), DNMT3A(0.96), NFKB1(0.96), LAMB3(0.96),
EZH2(0.96), MSH2(0.96), ERBB3(0.96), CDK2(0.96), RGMB(0.96),
DNMT1(0.96), NF1(0.96), MLH1(0.96), IKBKB(0.96), E2F1(0.96),
FOXM1(0.96), TBX3(0.96), LRP5(0.96), AKT3(0.96), PMS2(0.96),
HDAC1(0.96), TAP2(0.95), IFNGR2(0.95), BMI1(0.95), PCNA(0.95),
TAP1(0.95), CDCA7(0.95), VEGFB(0.95), MET(0.95), FGFR2(0.95),
CCNB1(0.95), IGF1R(0.95), CD40(0.95), MAML1(0.95), SNAI1(0.95),
EPCAM(0.95), BRCA1(0.95), TCF7L2(0.95), PIK3CA(0.95), EPHB3(0.95),
BIRC5(0.95), WNT3(0.95), IFNGR1(0.95), MYBL2(0.95), CHEK1(0.95),
ARG1(0.94), AMOTL2(0.94), BAX(0.94), IL12A(0.94), VCAM1(0.94),
HRAS(0.94), EGFR(0.94), CD160(0.94), BCL2(0.94), AKT2(0.94), CSF2(0.94),
TSC1(0.94), CEACAM1(0.94), CD83(0.94), PKM(0.94), AURKA(0.94),
SOX9(0.94), CX3CL1(0.94), ARAF(0.94), PLK1(0.94), CIITA(0.94),
IFNAR2(0.94), AKT1(0.94), IRS1(0.93), IDH1(0.93), IRF3(0.93),
PTEN(0.93), MKI67(0.93), NRAS(0.93), AURKB(0.93), MTOR(0.93),
HLA.DMA(0.93), MCM6(0.93), BUB1(0.93), TEAD4(0.93), ATM(0.93),
CCND1(0.92), HLA.DRA(0.92), ITGB1(0.92), YAP1(0.92)</td>
</tr>
<tr class="even">
<td align="left">c2</td>
<td align="left">COL11A1(0.92), COL1A1(0.87), TGFBR2(0.85), TNC(0.85),
PDGFRB(0.85), CXCL12(0.84), COL6A3(0.83), ACTA2(0.83), ZEB1(0.82),
NRP1(0.8), PDGFRA(0.8), FGF1(0.79), CD14(0.78), CAV1(0.78), TGFB1(0.78),
MMP2(0.77), DUSP1(0.76), COL5A1(0.76), FN1(0.76), CMKLR1(0.76),
ELN(0.75), MMP11(0.75), C1QC(0.73), SFRP2(0.73), FAP(0.73), MFAP5(0.73),
BST2(0.72), EGR1(0.72), CSF1R(0.71), FCGR3A(0.71), CD248(0.71),
MRC1(0.71), PDPN(0.71), FOS(0.7), FCGR2A(0.7), CYBB(0.7), CD163(0.69),
NEDD4(0.68), TWIST1(0.68), MAFB(0.68), TGFBI(0.67), CD4(0.67),
CSF1(0.65), ITGB2(0.65), GPX3(0.65), PECAM1(0.64), SNAI2(0.64),
CD86(0.63), ITGA4(0.62), TLR2(0.61), LRP1(0.61), PTPRC(0.59),
MSR1(0.59), CCR1(0.59), VSIR(0.58), THBD(0.58), IFITM1(0.58),
CDKN1A(0.57), NCAM1(0.57), CD28(0.57), VWF(0.56), CD209(0.56),
FBLN1(0.55), POU2AF1(0.53), PLVAP(0.52), PLOD2(0.52), NDUFA4L2(0.52),
IL6(0.52), MMP9(0.47), SERPINE1(0.43)</td>
</tr>
<tr class="odd">
<td align="left">c3</td>
<td align="left">TAPBP(0.96), PTK2(0.96), CDH1(0.96), ERBB2(0.96),
SMARCA4(0.96), CDKN1B(0.96), TP53(0.96), SHARPIN(0.95), FZD7(0.95),
LRP6(0.95), MYC(0.95), DIABLO(0.95), SMAD2(0.95), LAMB3(0.95),
NFE2L2(0.95), KITLG(0.95), STAT3(0.95), ATR(0.95), SRPRB(0.94),
CREBBP(0.94), CDK4(0.94), KRAS(0.94), MSH6(0.94), CASP8(0.94),
STAT6(0.94), BRAF(0.94), BRD4(0.94), DKK1(0.94), EPHB4(0.94), MET(0.94),
MCM2(0.94), APC(0.94), CHEK2(0.94), PPARD(0.94), BAK1(0.94), RELA(0.94),
ERBB3(0.94), AKT3(0.93), SNAI1(0.93), SMO(0.93), SRC(0.93), NFKB2(0.93),
MSH2(0.93), CEACAM1(0.93), TBX3(0.93), DNMT3A(0.93), CD40(0.93),
EPCAM(0.93), PMS2(0.93), HDAC3(0.93), HDAC1(0.93), CCNE1(0.93),
RAF1(0.93), NFKB1(0.93), FGFR2(0.93), IGF1R(0.93), MSH3(0.93),
TAP1(0.92), ICOSLG(0.92), TAP2(0.92), EZH2(0.92), VEGFB(0.92),
RGMB(0.92), IFNGR2(0.92), FOXM1(0.92), IFNGR1(0.92), LRP5(0.92),
E2F1(0.92), CDK2(0.92), VCAM1(0.92), BMI1(0.92), DNMT1(0.92),
EPHB3(0.91), IKBKB(0.91), CIITA(0.91), BCL2(0.91), CCNB1(0.91),
MLH1(0.91), SOX9(0.91), PCNA(0.91), BRCA1(0.91), CD160(0.91),
BIRC5(0.91), EGFR(0.91), NF1(0.91), LAMC2(0.91), AMOTL2(0.91),
ARG1(0.91), IRS1(0.91), TCF7L2(0.91), PIK3CA(0.91), CDCA7(0.91),
IL12A(0.91), WNT3(0.9), MYBL2(0.9), HLA.DRA(0.9), HRAS(0.9), CHEK1(0.9),
MUC1(0.9), CD83(0.9), CX3CL1(0.9), BAX(0.9), SLC13A3(0.9), ARAF(0.9),
MAML1(0.9), CSF2(0.9), AURKA(0.9), IDH1(0.9), HLA.DMA(0.9), PTEN(0.9),
ICAM1(0.9), PKM(0.9), SOX2(0.9), AKT2(0.9), TSC1(0.89), IFNAR2(0.89),
TGFB2(0.89), PLK1(0.89), PKIB(0.89), NRAS(0.89), TEAD4(0.89),
MTOR(0.89), BUB1(0.89), MKI67(0.89), AKT1(0.89)</td>
</tr>
<tr class="even">
<td align="left">c4</td>
<td align="left">PTK2(0.89), RELA(0.89), BRD4(0.89), FZD7(0.89),
NFE2L2(0.89), DIABLO(0.88), LAMB3(0.88), PMS2(0.88), SHARPIN(0.88),
SMAD2(0.88), AKT3(0.88), LRP6(0.88), BAK1(0.88), STAT3(0.88),
TP53(0.88), NFKB2(0.88), MET(0.88), ERBB2(0.88), SNAI1(0.88),
SMARCA4(0.88), MSH6(0.88), CREBBP(0.88), TAPBP(0.88), SRPRB(0.88),
KRAS(0.88), MYC(0.88), EGFR(0.88), DKK1(0.88), KITLG(0.88), BRAF(0.88),
EPHB4(0.88), SRC(0.88), CDKN1B(0.88), TAP2(0.88), DNMT3A(0.88),
CDK4(0.87), MSH2(0.87), ATR(0.87), PPARD(0.87), TBX3(0.87), VCAM1(0.87),
IGF1R(0.87), VEGFB(0.87), RAF1(0.87), HDAC3(0.87), TAP1(0.87),
CDH1(0.87), SMO(0.87), LAMC2(0.87), TEAD4(0.87), LMNA(0.87), LRP5(0.87),
E2F1(0.87), PKM(0.87), IFNGR2(0.87), NFKB1(0.87), APC(0.87), HRAS(0.87),
ICOSLG(0.87), CCNE1(0.87), BCL2(0.87), EZH2(0.87), CASP8(0.87),
HDAC1(0.87), CHEK2(0.87), SOX9(0.87), IL12A(0.86), MCM2(0.86),
CDCA7(0.86), STAT6(0.86), MSH3(0.86), YAP1(0.86), ARG1(0.86),
EPHB3(0.86), CD40(0.86), NRAS(0.86), CD83(0.86), FOXM1(0.86),
CX3CL1(0.86), BIRC5(0.86), IFNGR1(0.86), RGMB(0.86), CCNB1(0.85),
TSC1(0.85), NF1(0.85), PIK3CA(0.85), TNFRSF18(0.85), PLK1(0.85),
CSF2(0.85), CHEK1(0.85), DNMT1(0.85), CDK6(0.85), BUB1(0.85),
MTOR(0.85), CDK2(0.85), WNT3(0.85), IRS1(0.85), AURKA(0.85),
AMOTL2(0.85), EPCAM(0.85), MYBL2(0.85), ERBB3(0.85), HLA.DQA1(0.85),
ICAM1(0.85), CD160(0.84), HLA.DRA(0.84), MCM6(0.84), PCNA(0.84),
FGFR2(0.84), SOX2(0.84), CIITA(0.84), CEACAM1(0.84), PDK1(0.84),
TCF7L2(0.84), BRCA1(0.84)</td>
</tr>
<tr class="odd">
<td align="left">c5</td>
<td align="left">CD14(0.94), C1QC(0.93), FCGR3A(0.93), CSF1R(0.93),
CYBB(0.93), ITGB2(0.92), CD163(0.91), MAFB(0.91), CD4(0.9), FCGR2A(0.9),
CD86(0.88), NRP1(0.88), CCR1(0.87), MSR1(0.86), TGFBR2(0.86),
PTPRC(0.86), DUSP1(0.85), MRC1(0.85), CSF3R(0.85), TLR2(0.84),
LYZ(0.83), HAVCR2(0.83), SERPINA1(0.82), FOS(0.81), VSIR(0.81),
TGFB1(0.8), PDGFRA(0.79), PECAM1(0.78), COL1A1(0.78), GPX3(0.77),
CMKLR1(0.77), ZEB1(0.75), CD209(0.75), CD28(0.74), ACTA2(0.74),
ITGA4(0.74), TREM2(0.74), BST2(0.74), PDGFRB(0.74), THBD(0.74),
COL11A1(0.73), CAV1(0.73), SELPLG(0.73), EGR1(0.72), SFRP2(0.72),
TNC(0.72), CXCL12(0.71), TGFBI(0.7), CD248(0.7), MMP2(0.7), PDK4(0.69),
LRP1(0.69), FAP(0.69), FGF1(0.69), SNAI2(0.68), COL6A3(0.68),
CSF1(0.68), NEDD4(0.67), ZAP70(0.66), FBLN1(0.66), ELN(0.66),
TWIST1(0.65), ITGAX(0.65), CCR2(0.65), LGR6(0.65), HGF(0.64), ITK(0.64),
PDPN(0.64), GZMA(0.63), CD2(0.63), VWF(0.63), CD3G(0.62), CDKN1A(0.62),
MFAP5(0.62), COL5A1(0.62), TRBC1(0.62), ITGA1(0.61), FN1(0.61),
MMP11(0.6), CLEC14A(0.59), IFITM1(0.59), CLDN5(0.58), CCL8(0.57),
ASCL2(0.57), POU2AF1(0.57), MMRN2(0.57), PLVAP(0.56), IL6(0.55),
MMP9(0.53)</td>
</tr>
<tr class="even">
<td align="left">c6</td>
<td align="left">SPP1(0.85), CXCL8(0.8), MMP12(0.75), VEGFA(0.72),
SERPINE1(0.56), MARCO(0.55), MMP9(0.5), PLOD2(0.48), TGFBI(0.44),
IL6(0.38)</td>
</tr>
<tr class="odd">
<td align="left">c7</td>
<td align="left">CD2(0.95), TRAC(0.95), CD3E(0.94), GZMA(0.92),
ZAP70(0.91), ITK(0.9), CD3D(0.89), IL2RB(0.89), CD3G(0.88), TRBC1(0.88),
CD5(0.86), TIGIT(0.85), PTPRC(0.84), NKG7(0.84), CD28(0.83), CCR4(0.83),
ITGA4(0.81), CTSW(0.81), CD8A(0.8), TBX21(0.8), PDCD1(0.8), FOXP3(0.79),
PDCD1LG2(0.79), CXCR3(0.79), SELPLG(0.78), CCR2(0.78), GZMH(0.78),
CD4(0.77), CYBB(0.76), ICOS(0.76), KLRK1(0.76), CCL5(0.75), GPX3(0.75),
CCR7(0.74), ENG(0.74), ITGA1(0.74), CCR5(0.73), PECAM1(0.73),
STAT5A(0.73), HGF(0.73), CD209(0.73), VSIR(0.73), PRF1(0.72),
CMKLR1(0.72), LRP1(0.72), ICAM3(0.72), CD274(0.72), LEF1(0.72),
TRAT1(0.71), MMRN2(0.71), CLEC14A(0.71), CSF3R(0.71), CXCL9(0.71),
C1QC(0.71), CSF1R(0.71), CLEC4C(0.7), GNLY(0.7), PDK4(0.7), TGFBR2(0.7),
CD248(0.7), CSF2RA(0.7), FOS(0.69), SNAI2(0.69), FCGR3A(0.69),
AXIN2(0.69), HAVCR2(0.69), PLVAP(0.69), VWF(0.69), THBD(0.68),
TNFRSF4(0.68), POU2AF1(0.68), CCL4(0.67), IKZF2(0.67), FAP(0.67),
ADAMTS4(0.67), FCGR2A(0.66), MRC1(0.66), MAFB(0.66), LYZ(0.66),
TGFBR1(0.66), BST2(0.66), ITGB2(0.66), CDH5(0.66), ANGPT1(0.66),
CSF1(0.65), MFAP5(0.65), DES(0.65), CD19(0.65), ZEB1(0.65), CD163(0.65),
CD14(0.64), PDGFRA(0.64), SERPINA1(0.64), PDGFB(0.64), CX3CR1(0.64),
NRP1(0.64), ACTA2(0.64), FLT4(0.64), SMOC2(0.63), CDKN1A(0.63),
TGFB3(0.63), CXCL12(0.63), MMP1(0.63), TREM2(0.63)</td>
</tr>
<tr class="even">
<td align="left">c8</td>
<td align="left">PLVAP(0.93), VWF(0.93), MMRN2(0.91), ITGA1(0.89),
PECAM1(0.89), ANGPT2(0.89), CLEC14A(0.88), CD248(0.86), PGF(0.84),
CDH5(0.83), NDUFA4L2(0.83), FLT4(0.83), SNAI2(0.79), ENG(0.79),
ZEB1(0.77), COL4A1(0.76), CLDN5(0.76), PDGFRB(0.75), KDR(0.75),
TNFRSF4(0.75), NRP1(0.73), TGFB3(0.72), ITGA4(0.72), PDGFB(0.72),
ADAMTS4(0.72), GPX3(0.71), HGF(0.71), FAP(0.7), TGFBR2(0.7),
CMKLR1(0.7), CAV1(0.7), FLT1(0.69), PREX2(0.69), CD28(0.68),
ACTA2(0.68), MMP11(0.66), ZAP70(0.66), POU2AF1(0.66), IL3RA(0.66),
MFAP5(0.65), ANGPT1(0.65), CXCL12(0.65), LEF1(0.65), TGFB1(0.64),
CCR4(0.64), SFRP2(0.63), COL1A1(0.63), NEDD4(0.63), VSIR(0.62),
LRP1(0.62), BST2(0.62), THBD(0.62), CCR2(0.62), CYBB(0.62),
TWIST1(0.61), CD4(0.61), PTPRC(0.61), AXIN2(0.61), CD3G(0.61),
CCR7(0.61), CD2(0.6), CSF1R(0.6), PDK4(0.6), TRAC(0.6), TRBC1(0.6),
GZMA(0.59), CD209(0.59), SMOC2(0.59), FOS(0.59), PDGFRA(0.59),
CSF3R(0.59), C1QC(0.57), FCGR3A(0.57), CD14(0.56), MAFB(0.56),
COL5A1(0.56), EGR1(0.56), FGF1(0.55), CSF1(0.55), IFITM1(0.54),
MRC1(0.53), DUSP1(0.53), CDKN1A(0.52), FCGR2A(0.52), FBLN1(0.52),
COL6A3(0.51)</td>
</tr>
<tr class="odd">
<td align="left">c9</td>
<td align="left">CD79A(0.67), POU2AF1(0.66), MZB1(0.66), MFAP5(0.44),
PECAM1(0.43), CD248(0.42), ZEB1(0.41), TGFBR2(0.37), ACTA2(0.37),
CXCL12(0.36)</td>
</tr>
<tr class="even">
<td align="left">c10</td>
<td align="left">CD207(0.86), FCER1A(0.76), CD1C(0.61), CD1E(0.61),
CD1B(0.56), CXCL5(0.52)</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="limma" class="section level2">
<h2>limma</h2>
<p>We applied the limma framework to identify differentially expressed
marker genes across clusters in the healthy liver sample. Count data
were normalized using log-transformed counts from the speckle package. A
linear model was fitted for each gene with cluster identity as the
design factor, and pairwise contrasts were constructed to compare each
cluster against all others. Empirical Bayes moderation was applied to
stabilize variance estimates, and significant genes were determined
using the moderated t-statistics from eBayes.</p>
<pre class="r"><code>cm &lt;- assay(se, &quot;counts&quot;)

y &lt;- DGEList(cm)
y$genes &lt;-row.names(cm)

logcounts &lt;- speckle::normCounts(y,log=TRUE,prior.count=0.1)
maxclust &lt;- length(unique(cluster_info$cluster))

grp &lt;- cluster_info$cluster

design &lt;- model.matrix(~0+grp)
colnames(design) &lt;- levels(grp)

mycont &lt;- matrix(NA,ncol=length(levels(grp)),nrow=length(levels(grp)))
rownames(mycont)&lt;-colnames(mycont)&lt;-levels(grp)
diag(mycont)&lt;-1
mycont[upper.tri(mycont)]&lt;- -1/(length(levels(factor(grp)))-1)
mycont[lower.tri(mycont)]&lt;- -1/(length(levels(factor(grp)))-1)

fit &lt;- lmFit(logcounts,design)
fit.cont &lt;- contrasts.fit(fit,contrasts=mycont)
fit.cont &lt;- eBayes(fit.cont,trend=TRUE,robust=TRUE)

limma_dt&lt;-decideTests(fit.cont)

saveRDS(fit.cont, here(gdata,paste0(data_nm, &quot;_fit_cont_obj.Rds&quot;)))</code></pre>
<pre class="r"><code>fit.cont = readRDS(here(gdata,paste0(data_nm, &quot;_fit_cont_obj.Rds&quot;)))
limma_dt &lt;- decideTests(fit.cont)
summary(limma_dt)</code></pre>
<pre><code>        c1  c2  c3  c4  c5  c6  c7  c8  c9 c10
Down   135 411 234 424 298 455 280 335 333 261
NotSig   3   2  12   4  13   3  19  20  43  65
Up     362  87 254  72 189  42 201 145 124 174</code></pre>
<pre class="r"><code>top_markers &lt;- lapply(cluster_names, function(cl) {
  tt &lt;- topTable(fit.cont,
                 coef = cl,
                 number = Inf,
                 adjust.method = &quot;BH&quot;,
                 sort.by = &quot;P&quot;)
  tt &lt;- tt[order(tt$adj.P.Val, -abs(tt$logFC)), ]  
  tt$contrast &lt;- cl
  tt$gene &lt;- rownames(tt)
  head(tt, 10)   # top 10
})


# Combine into one data.frame
top_markers_df &lt;- bind_rows(top_markers) %&gt;%
  select(contrast, gene, logFC, adj.P.Val)


p1&lt;- DotPlot(seu, features = unique(top_markers_df$gene)) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (limma)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/looad_limma_res-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-looad_limma_res-1">
Past versions of looad_limma_res-1.png
</button>
</p>
<div id="fig-looad_limma_res-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/looad_limma_res-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="wilcoxon-rank-sum-test" class="section level2">
<h2>Wilcoxon Rank Sum Test</h2>
<p>The <code>FindAllMarkers</code> function applies a Wilcoxon Rank Sum
test to detect genes with cluster-specific expression shifts. Only
positive markers (upregulated genes) were retained using a log
fold-change threshold of 0.1.</p>
<pre class="r"><code>cm &lt;- assay(se, &quot;counts&quot;)

hbm_seu&lt;-CreateSeuratObject(counts = cm,
                            project = &quot;hbreast&quot;)
Idents(hbm_seu) &lt;- cluster_info[match(colnames(hbm_seu), 
                                       cluster_info$cell_id),&quot;cluster&quot;]
hbm_seu &lt;- NormalizeData(hbm_seu, verbose = FALSE,
                         normalization.method = &quot;LogNormalize&quot;)
hbm_seu &lt;- FindVariableFeatures(hbm_seu, selection.method = &quot;vst&quot;, 
                                nfeatures = 1000, verbose = FALSE)
hbm_seu &lt;- ScaleData(hbm_seu, verbose = FALSE)

seu_markers &lt;- FindAllMarkers(hbm_seu, only.pos = TRUE,logfc.threshold = 0.1)


saveRDS(seu_markers, here(gdata,paste0(data_nm, &quot;_seu_markers.Rds&quot;)))

saveRDS(hlc_seu, here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))</code></pre>
<pre class="r"><code>FM_result= readRDS(here(gdata,paste0(data_nm, &quot;_seu_markers.Rds&quot;)))
# load generated data  (see code above for how they were created)

FM_result$cluster &lt;- factor(FM_result$cluster,
                            levels = cluster_names)

top_mg &lt;- FM_result %&gt;%
  group_by(cluster) %&gt;%
  arrange(desc(avg_log2FC), .by_group = TRUE) %&gt;% 
  slice_max(order_by = avg_log2FC, n = 10) %&gt;%
  select(cluster, gene, avg_log2FC)

p1&lt;-DotPlot(seu, features = unique(top_mg$gene)) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (WRS)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1


# # Combine into a new dataframe for LaTeX output
# output_data &lt;- do.call(rbind, lapply(cluster_names, function(cl) {
#   subset &lt;- top_mg[top_mg$cluster == cl, ]
#   sorted_subset &lt;- subset[order(-subset$avg_log2FC), ]
#   gene_list &lt;- paste(sorted_subset$gene, &quot;(&quot;, round(sorted_subset$avg_log2FC, 2), &quot;)&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
#  
#   return(data.frame(Cluster = cl, Genes = gene_list))
# }))
# latex_table &lt;- xtable(output_data, caption=&quot;Detected marker genes for each cluster by FindMarkers&quot;)
# print.xtable(latex_table, include.rownames = FALSE, hline.after = c(-1, 0, nrow(output_data)), comment = FALSE)</code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/load_FM_res-1.png" width="864" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-load_FM_res-1">
Past versions of load_FM_res-1.png
</button>
</p>
<div id="fig-load_FM_res-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/load_FM_res-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="comparison-of-marker-gene-detection-methods"
class="section level1">
<h1>Comparison of marker gene detection methods</h1>
<div id="cumulative-rank-plot" class="section level2">
<h2>Cumulative rank plot</h2>
<p>We compared the performance of different marker detection methods —
jazzPanda (linear modelling and correlation), limma, and Seurat’s
Wilcoxon Rank Sum test -by evaluating the cumulative average spatial
correlation between ranked marker genes and their corresponding target
clusters at the spatial vector level.</p>
<p>Spatially informed jazzPanda methods prioritize genes with stronger
spatial coherence in most clusters, especially c2, c5, c7, and c8, while
traditional non-spatial approaches (Wilcoxon, limma) tend to accumulate
high-correlation genes later or not at all.</p>
<pre class="r"><code>plot_lst=list()
cor_M = cor(sv_lst$gene_mt,
            sv_lst$cluster_mt[, cluster_names],
            method = &quot;spearman&quot;)
Idents(seu)=cluster_info$cluster[match(colnames(seu), cluster_info$cell_id)]
for (cl in cluster_names){
    
    fm_cl=FindMarkers(seu, ident.1 = cl, only.pos = TRUE,
            logfc.threshold = 0.1)
    fm_cl = fm_cl[fm_cl$p_val_adj&lt;0.05, ]
    fm_cl = fm_cl[order(fm_cl$avg_log2FC, decreasing = TRUE),]
    to_plot_fm =row.names(fm_cl)
    if (length(to_plot_fm)&gt;0){
        FM_pt =FM_pt = data.frame(&quot;name&quot;=to_plot_fm,&quot;rk&quot;= 1:length(to_plot_fm),
                       &quot;y&quot;= get_cmr_ma(to_plot_fm,cor_M = cor_M, cl = cl),
                       &quot;type&quot;=&quot;Wilcoxon Rank Sum Test&quot;)
    }else{
        FM_pt &lt;- data.frame(name = character(0), rk = integer(0), 
                               y= numeric(0), type = character(0))
    }

    limma_cl&lt;-topTable(fit.cont,coef=cl,p.value = 0.05, n=Inf, sort.by = &quot;p&quot;)
    limma_cl = limma_cl[limma_cl$logFC&gt;0, ]
    to_plot_lm = row.names(limma_cl)
    if (length(to_plot_lm)&gt;0){
         limma_pt = data.frame(&quot;name&quot;=to_plot_lm,&quot;rk&quot;= 1:length(to_plot_lm),
                                  &quot;y&quot;= get_cmr_ma(to_plot_lm,cor_M = cor_M, cl = cl),
                                  &quot;type&quot;=&quot;limma&quot;)
    }else{
        limma_pt &lt;- data.frame(name = character(0), rk = integer(0), 
                               y= numeric(0), type = character(0))
    }
    
    obs_cutoff = quantile(obs_corr[, cl], 0.75)
    perm_cl=intersect(row.names(perm_res[perm_res[,cl]&lt;0.05,]),
                 row.names(obs_corr[obs_corr[, cl]&gt;obs_cutoff,]))
    if (length(perm_cl) &gt; 0) {
        rounded_val=signif(as.numeric(obs_corr[perm_cl,cl]), digits = 3)
        roudned_pval=signif(as.numeric(perm_res[perm_cl,cl]), digits = 3)
        perm_sorted = as.data.frame(cbind(gene=perm_cl, value=rounded_val, pval=roudned_pval))
        perm_sorted$value = as.numeric(perm_sorted$value)
        perm_sorted$pval = as.numeric(perm_sorted$pval)
        perm_sorted=perm_sorted[order(-perm_sorted$pval,
                                  perm_sorted$value, 
                                  decreasing = TRUE),]
        corr_pt = data.frame(&quot;name&quot;=perm_sorted$gene,&quot;rk&quot;= 1:length(perm_sorted$gene),
                     &quot;y&quot;= get_cmr_ma(perm_sorted$gene,cor_M = cor_M, cl = cl),
                     &quot;type&quot;=&quot;jazzPanda-correlation&quot;)
    } else {
      corr_pt &lt;-data.frame(name = character(0), rk = integer(0), 
                               y= numeric(0), type = character(0))
    }
    
    lasso_sig = jazzPanda_res[jazzPanda_res$top_cluster==cl,]
    if (nrow(lasso_sig) &gt; 0) {
        lasso_sig &lt;- lasso_sig[order(lasso_sig$glm_coef, decreasing = TRUE), ]
        lasso_pt = data.frame(&quot;name&quot;=lasso_sig$gene,&quot;rk&quot;= 1:nrow(lasso_sig),
                          &quot;y&quot;= get_cmr_ma(lasso_sig$gene,cor_M = cor_M, cl = cl),
                          &quot;type&quot;=&quot;jazzPanda-glm&quot;)
    } else {
      lasso_pt &lt;- data.frame(name = character(0), rk = integer(0), 
                               y= numeric(0), type = character(0))
    }

    data_lst = rbind(limma_pt, FM_pt,corr_pt,lasso_pt)

    data_lst$type &lt;- factor(data_lst$type,
                            levels = c(&quot;jazzPanda-correlation&quot; ,
                                       &quot;jazzPanda-glm&quot;,
                                       &quot;limma&quot;, &quot;Wilcoxon Rank Sum Test&quot;))
    data_lst$rk = as.numeric(data_lst$rk)
    p &lt;-ggplot(data_lst, aes(x = rk, y = y, color = type)) +
        geom_step(size = 0.8) +  # type = &quot;s&quot;
        scale_color_manual(values = c(&quot;jazzPanda-correlation&quot; = &quot;orange&quot;,
                                    &quot;jazzPanda-glm&quot; = &quot;red&quot;,
                                    &quot;limma&quot; = &quot;black&quot;,
                                    &quot;Wilcoxon Rank Sum Test&quot; = &quot;blue&quot;)) +
        scale_x_continuous(limits = c(0, 50))+
        labs(title = paste0(&quot;Cluster &quot;,cl), x = &quot;Rank of marker genes&quot;,
             y = &quot;Cumulative average correlation&quot;,
             color = NULL) +
        theme_classic(base_size = 12) +
        theme(plot.title = element_text(hjust = 0.5),
            axis.line = element_blank(),  
            panel.border = element_rect(color = &quot;black&quot;, 
                                        fill = NA, linewidth = 1),
            legend.position = &quot;inside&quot;,
            legend.position.inside = c(0.98, 0.02),
            legend.justification = c(&quot;right&quot;, &quot;bottom&quot;),
            legend.background = element_rect(color = &quot;black&quot;, 
                                             fill = &quot;white&quot;, linewidth = 0.5),
            legend.box.background = element_rect(color = &quot;black&quot;,
                                                 linewidth = 0.5)
        )
    plot_lst[[cl]] &lt;- p
}


combined_plot &lt;- wrap_plots(plot_lst, ncol =3)

combined_plot </code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/cmr-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-cmr-1">
Past versions of cmr-1.png
</button>
</p>
<div id="fig-cmr-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/cmr-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="upset-plot" class="section level2">
<h2>Upset plot</h2>
<p>We further visualized the overlap between marker genes detected by
the four methods using UpSet plots for each cluster. Clusters c1, c5, c7
and c8 show the greatest overlap between methods, reflecting
well-defined expression domains, whereas c4, c9, and c10 exhibit smaller
intersections, consistent with weaker or more diffuse transcriptional
contrast.</p>
<pre class="r"><code>plot_lst=list()

for (cl in cluster_names){
    #ct_nm = anno_df[anno_df$cluster==cl, &quot;anno_name&quot;]
    findM_sig &lt;- FM_result[FM_result$cluster==cl &amp; FM_result$p_val_adj&lt;0.05,&quot;gene&quot;]
    limma_sig &lt;- row.names(limma_dt[limma_dt[,cl]==1,])
    lasso_cl &lt;- jazzPanda_res[jazzPanda_res$top_cluster==cl, &quot;gene&quot;]
    obs_cutoff = quantile(obs_corr[, cl], 0.75)
    perm_cl&lt;-intersect(row.names(perm_res[perm_res[,cl]&lt;0.05,]),
                     row.names(obs_corr[obs_corr[, cl]&gt;obs_cutoff,]))
    df_mt &lt;- as.data.frame(matrix(FALSE,nrow=nrow(jazzPanda_res),ncol=4))
    row.names(df_mt) &lt;- jazzPanda_res$gene
    colnames(df_mt) &lt;- c(&quot;jazzPanda-glm&quot;,
                      &quot;jazzPanda-correlation&quot;,
                      &quot;Wilcoxon Rank Sum Test&quot;,&quot;limma&quot;)
    df_mt[findM_sig,&quot;Wilcoxon Rank Sum Test&quot;] &lt;- TRUE
    df_mt[limma_sig,&quot;limma&quot;] &lt;- TRUE
    df_mt[lasso_cl,&quot;jazzPanda-glm&quot;] &lt;- TRUE
    df_mt[perm_cl,&quot;jazzPanda-correlation&quot;] &lt;- TRUE
    
    df_mt$gene_name &lt;- row.names(df_mt)
    
    p = plot(upset(df_mt,
               intersect=c(&quot;Wilcoxon Rank Sum Test&quot;, &quot;limma&quot;,
                           &quot;jazzPanda-correlation&quot;,&quot;jazzPanda-glm&quot;),
               wrap=TRUE, keep_empty_groups= FALSE, name=&quot;&quot;,
               stripes=&#39;white&#39;,
               sort_intersections_by =&quot;cardinality&quot;, sort_sets= FALSE,min_degree=1,
               set_sizes =( 
                   upset_set_size()
                   + theme(axis.title= element_blank(),
                           axis.ticks.y = element_blank(),
                           axis.text.y = element_blank())),
               sort_intersections= &quot;descending&quot;, warn_when_converting=FALSE,
               warn_when_dropping_groups=TRUE,encode_sets=TRUE,
               width_ratio=0.3, height_ratio=1/4)+
            ggtitle(paste0(&quot;Cluster &quot;,cl))+
                 theme(plot.title = element_text( size=15))

    )
           
        plot_lst[[cl]] &lt;- p 
}

combined_plot &lt;- wrap_plots(plot_lst, ncol =3)

combined_plot </code></pre>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-1">
Past versions of upset-mg-comparison-1.png
</button>
</p>
<div id="fig-upset-mg-comparison-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-1.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-2.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-2">
Past versions of upset-mg-comparison-2.png
</button>
</p>
<div id="fig-upset-mg-comparison-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-2.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-3.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-3">
Past versions of upset-mg-comparison-3.png
</button>
</p>
<div id="fig-upset-mg-comparison-3" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-3.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-4.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-4">
Past versions of upset-mg-comparison-4.png
</button>
</p>
<div id="fig-upset-mg-comparison-4" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-4.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-5.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-5">
Past versions of upset-mg-comparison-5.png
</button>
</p>
<div id="fig-upset-mg-comparison-5" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-5.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-6.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-6">
Past versions of upset-mg-comparison-6.png
</button>
</p>
<div id="fig-upset-mg-comparison-6" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-6.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-7.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-7">
Past versions of upset-mg-comparison-7.png
</button>
</p>
<div id="fig-upset-mg-comparison-7" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-7.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-8.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-8">
Past versions of upset-mg-comparison-8.png
</button>
</p>
<div id="fig-upset-mg-comparison-8" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-8.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-9.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-9">
Past versions of upset-mg-comparison-9.png
</button>
</p>
<div id="fig-upset-mg-comparison-9" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-9.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-10.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-10">
Past versions of upset-mg-comparison-10.png
</button>
</p>
<div id="fig-upset-mg-comparison-10" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-10.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><img src="figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-11.png" width="1152" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-upset-mg-comparison-11">
Past versions of upset-mg-comparison-11.png
</button>
</p>
<div id="fig-upset-mg-comparison-11" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://gh_personal_melody/phipsonlab/jazzPanda_workflowr/blob/ae17c0e531b852fdb283a825d1720a4c48f6e915/docs/figure/merscope-human-breast-cancer.Rmd/upset-mg-comparison-11.png" target="_blank">ae17c0e</a>
</td>
<td>
Melody
</td>
<td>
2026-01-22
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="annotate-cell-type-with-top-marker-genes"
class="section level1">
<h1>Annotate cell type with top marker genes</h1>
<pre class="r"><code>anno_df &lt;- data.frame(
    cluster = cluster_names,
  major_class = c(
    &quot;Epithelial tumor&quot;,&quot;Stromal&quot;,&quot;Epithelial tumor&quot;,&quot;Epithelial tumor&quot;,
    &quot;Myeloid&quot;,&quot;Myeloid&quot;,&quot;Lymphoid&quot;,&quot;Endothelial / Vascular&quot;,&quot;B lineage&quot;,&quot;Dendritic cell&quot;
  ),
  sub_class = c(
    &quot;Luminal-like / ERBB2-high&quot;,&quot;Cancer-associated fibroblast (CAF)&quot;,&quot;IFN-γ–licensed&quot;,&quot;Cycling (G2/M)&quot;,
    &quot;Tumor-associated macrophage (TAM)&quot;,&quot;Inflammatory/angiogenic TAM&quot;,&quot;T cells (activated)&quot;,&quot;Blood endothelial&quot;,&quot;Plasma cell / plasmablast&quot;,&quot;Langerhans / cDC2-like&quot;
  ),
  cell_type = c(
    &quot;Carcinoma—proliferative ERBB2/EGFR⁺&quot;,
    &quot;myCAF / iCAF hybrid&quot;,
    &quot;Carcinoma—MHC-II⁺ (APC-like)&quot;,
    &quot;Carcinoma—G2/M cycling&quot;,
    &quot;C1QC⁺ M2-like TAM&quot;,
    &quot;SPP1⁺ (osteopontin) TAM&quot;,
    &quot;Mixed CD4/CD8 with Treg features&quot;,
    &quot;Endothelium—PLVAP/KDR⁺&quot;,
    &quot;Plasma cells&quot;,
    &quot;CD207⁺ APC&quot;
  ),
  supporting_genes = c(
    &quot;EPCAM, ERBB2, ERBB3, EGFR, CDH1, MYC, FOXM1, MYBL2, MCM2, PCNA, CCNB1, LGR5, DKK1, FZD7, EPHB3&quot;,
    &quot;FN1, COL1A1, COL11A1, COL6A3, COL5A1, ACTA2, PDGFRA, TNC, FAP, CXCL12, IL6, SERPINE1, MFAP5, ELN, LOX&quot;,
    &quot;CDH1, EPCAM, HLA-DPB1/DRB1, CIITA, IDO1, TNFSF10, TGFB2, CEACAM1&quot;,
    &quot;BIRC5, PLK1, CCNB1, AURKA, AURKB, FOXM1, MYBL2, MCM2, TP53&quot;,
    &quot;C1QC, LYZ, FCGR3A, CD14, MRC1, CD163, MSR1, CSF1R, TREM2, LGALS9&quot;,
    &quot;SPP1, VEGFA, MMP9, MMP12, CXCL8, SERPINE1, TGFBI, ICAM1, ATF3&quot;,
    &quot;TRAC, CD3D/E/G, CD2, CD8A, GZMA/H/K, NKG7, CXCR3, TIGIT, ICOS, FOXP3, CCR7, TBX21, EOMES&quot;,
    &quot;PECAM1, VWF, CDH5, KDR, FLT1, CLDN5, PLVAP, MMRN2, ANGPT2, PGF, CLEC14A&quot;,
    &quot;XBP1, MZB1, IRF4, POU2AF1, CD79A/B, DERL3, FCRL5&quot;,
    &quot;CD207, FCER1A, CD1C, CD1B/E, CCL22, CSF1R&quot;
  ),
  notes = c(
    &quot;Luminal-like/HER2-high epithelial tumor with strong proliferation and WNT cues.&quot;,
    &quot;Mixed myCAF/iCAF phenotype; ECM deposition and inflammatory signaling coexist.&quot;,
    &quot;IFN-γ–induced antigen presentation by tumor cells; epithelial identity preserved.&quot;,
    &quot;Mitotic (G2/M) program dominates; classic cycling-tumor signature.&quot;,
    &quot;Immunoregulatory TAMs with M2-like polarization; chemokine-rich.&quot;,
    &quot;Angiogenic/remodeling TAMs; SPP1/VEGFA/MMP axis.&quot;,
    &quot;Activated T cells with cytotoxic and Treg subsets co-enriched.&quot;,
    &quot;Activated tumor endothelium; tip/stalk features (PLVAP/ANGPT2).&quot;,
    &quot;Canonical plasma cell program; minor contamination unlikely to change call.&quot;,
    &quot;Langerin+ dendritic cells bridging cDC2/Langerhans features.&quot;
  ),
    confidence = c(&quot;High&quot;,&quot;High&quot;,&quot;High&quot;,&quot;High&quot;,&quot;High&quot;,&quot;High&quot;,
                 &quot;Medium&quot;,&quot;High&quot;,&quot;High&quot;,&quot;High&quot;),
  
    anno_name = c(&quot;Tumor_Luminal_ERBB2&quot;, 
                &quot;CAF_Mixed&quot;,             
                &quot;Tumor_IFNg_APC&quot;,        
                &quot;Tumor_Cycling_G2M&quot;,     
                &quot;TAM_C1QC&quot;,             
                &quot;TAM_SPP1&quot;,              
                &quot;Tcell_Activated&quot;,       
                &quot;Endo_Blood&quot;,            
                &quot;Plasma&quot;,           
                &quot;DC_Langerhans_cDC2&quot;),
stringsAsFactors = FALSE
)

anno_df</code></pre>
<pre><code>   cluster            major_class                          sub_class
1       c1       Epithelial tumor          Luminal-like / ERBB2-high
2       c2                Stromal Cancer-associated fibroblast (CAF)
3       c3       Epithelial tumor                     IFN-γ–licensed
4       c4       Epithelial tumor                     Cycling (G2/M)
5       c5                Myeloid  Tumor-associated macrophage (TAM)
6       c6                Myeloid        Inflammatory/angiogenic TAM
7       c7               Lymphoid                T cells (activated)
8       c8 Endothelial / Vascular                  Blood endothelial
9       c9              B lineage          Plasma cell / plasmablast
10     c10         Dendritic cell             Langerhans / cDC2-like
                             cell_type
1  Carcinoma—proliferative ERBB2/EGFR⁺
2                  myCAF / iCAF hybrid
3         Carcinoma—MHC-II⁺ (APC-like)
4               Carcinoma—G2/M cycling
5                    C1QC⁺ M2-like TAM
6              SPP1⁺ (osteopontin) TAM
7     Mixed CD4/CD8 with Treg features
8               Endothelium—PLVAP/KDR⁺
9                         Plasma cells
10                          CD207⁺ APC
                                                                                        supporting_genes
1         EPCAM, ERBB2, ERBB3, EGFR, CDH1, MYC, FOXM1, MYBL2, MCM2, PCNA, CCNB1, LGR5, DKK1, FZD7, EPHB3
2  FN1, COL1A1, COL11A1, COL6A3, COL5A1, ACTA2, PDGFRA, TNC, FAP, CXCL12, IL6, SERPINE1, MFAP5, ELN, LOX
3                                       CDH1, EPCAM, HLA-DPB1/DRB1, CIITA, IDO1, TNFSF10, TGFB2, CEACAM1
4                                             BIRC5, PLK1, CCNB1, AURKA, AURKB, FOXM1, MYBL2, MCM2, TP53
5                                       C1QC, LYZ, FCGR3A, CD14, MRC1, CD163, MSR1, CSF1R, TREM2, LGALS9
6                                          SPP1, VEGFA, MMP9, MMP12, CXCL8, SERPINE1, TGFBI, ICAM1, ATF3
7               TRAC, CD3D/E/G, CD2, CD8A, GZMA/H/K, NKG7, CXCR3, TIGIT, ICOS, FOXP3, CCR7, TBX21, EOMES
8                                PECAM1, VWF, CDH5, KDR, FLT1, CLDN5, PLVAP, MMRN2, ANGPT2, PGF, CLEC14A
9                                                       XBP1, MZB1, IRF4, POU2AF1, CD79A/B, DERL3, FCRL5
10                                                             CD207, FCER1A, CD1C, CD1B/E, CCL22, CSF1R
                                                                               notes
1    Luminal-like/HER2-high epithelial tumor with strong proliferation and WNT cues.
2     Mixed myCAF/iCAF phenotype; ECM deposition and inflammatory signaling coexist.
3  IFN-γ–induced antigen presentation by tumor cells; epithelial identity preserved.
4                 Mitotic (G2/M) program dominates; classic cycling-tumor signature.
5                   Immunoregulatory TAMs with M2-like polarization; chemokine-rich.
6                                   Angiogenic/remodeling TAMs; SPP1/VEGFA/MMP axis.
7                     Activated T cells with cytotoxic and Treg subsets co-enriched.
8                    Activated tumor endothelium; tip/stalk features (PLVAP/ANGPT2).
9        Canonical plasma cell program; minor contamination unlikely to change call.
10                      Langerin+ dendritic cells bridging cDC2/Langerhans features.
   confidence           anno_name
1        High Tumor_Luminal_ERBB2
2        High           CAF_Mixed
3        High      Tumor_IFNg_APC
4        High   Tumor_Cycling_G2M
5        High            TAM_C1QC
6        High            TAM_SPP1
7      Medium     Tcell_Activated
8        High          Endo_Blood
9        High              Plasma
10       High  DC_Langerhans_cDC2</code></pre>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Red Hat Enterprise Linux 9.3 (Plow)

Matrix products: default
BLAS:   /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRblas.so 
LAPACK: /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRlapack.so;  LAPACK version 3.12.1

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Australia/Melbourne
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] harmony_1.2.3               Rcpp_1.1.0                 
 [3] Banksy_1.4.0                dotenv_1.0.3               
 [5] corrplot_0.95               xtable_1.8-4               
 [7] here_1.0.1                  limma_3.64.1               
 [9] RColorBrewer_1.1-3          patchwork_1.3.1            
[11] gridExtra_2.3               ggrepel_0.9.6              
[13] ComplexUpset_1.3.3          caret_7.0-1                
[15] lattice_0.22-7              glmnet_4.1-10              
[17] Matrix_1.7-3                tidyr_1.3.1                
[19] dplyr_1.1.4                 data.table_1.17.8          
[21] ggplot2_3.5.2               Seurat_5.3.0               
[23] SeuratObject_5.1.0          sp_2.2-0                   
[25] SpatialExperiment_1.18.1    SingleCellExperiment_1.30.1
[27] SummarizedExperiment_1.38.1 Biobase_2.68.0             
[29] GenomicRanges_1.60.0        GenomeInfoDb_1.44.2        
[31] IRanges_2.42.0              S4Vectors_0.46.0           
[33] BiocGenerics_0.54.0         generics_0.1.4             
[35] MatrixGenerics_1.20.0       matrixStats_1.5.0          
[37] jazzPanda_0.2.4            

loaded via a namespace (and not attached):
  [1] RcppHungarian_0.3       RcppAnnoy_0.0.22        splines_4.5.1          
  [4] later_1.4.2             tibble_3.3.0            polyclip_1.10-7        
  [7] hardhat_1.4.2           pROC_1.19.0.1           rpart_4.1.24           
 [10] fastDummies_1.7.5       lifecycle_1.0.4         aricode_1.0.3          
 [13] rprojroot_2.1.0         globals_0.18.0          MASS_7.3-65            
 [16] magrittr_2.0.3          plotly_4.11.0           sass_0.4.10            
 [19] rmarkdown_2.29          jquerylib_0.1.4         yaml_2.3.10            
 [22] httpuv_1.6.16           sctransform_0.4.2       spam_2.11-1            
 [25] spatstat.sparse_3.1-0   reticulate_1.43.0       cowplot_1.2.0          
 [28] pbapply_1.7-4           lubridate_1.9.4         abind_1.4-8            
 [31] Rtsne_0.17              presto_1.0.0            purrr_1.1.0            
 [34] BumpyMatrix_1.16.0      nnet_7.3-20             ipred_0.9-15           
 [37] git2r_0.36.2            lava_1.8.1              GenomeInfoDbData_1.2.14
 [40] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-5   
 [43] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.4-1  
 [46] fitdistrplus_1.2-4      parallelly_1.45.1       codetools_0.2-20       
 [49] DelayedArray_0.34.1     tidyselect_1.2.1        shape_1.4.6.1          
 [52] UCSC.utils_1.4.0        farver_2.1.2            spatstat.explore_3.5-2 
 [55] jsonlite_2.0.0          progressr_0.15.1        ggridges_0.5.6         
 [58] survival_3.8-3          iterators_1.0.14        dbscan_1.2.2           
 [61] foreach_1.5.2           tools_4.5.1             ica_1.0-3              
 [64] glue_1.8.0              prodlim_2025.04.28      SparseArray_1.8.1      
 [67] xfun_0.52               withr_3.0.2             fastmap_1.2.0          
 [70] digest_0.6.37           timechange_0.3.0        R6_2.6.1               
 [73] mime_0.13               colorspace_2.1-1        scattermore_1.2        
 [76] sccore_1.0.6            tensor_1.5.1            spatstat.data_3.1-8    
 [79] hexbin_1.28.5           recipes_1.3.1           class_7.3-23           
 [82] httr_1.4.7              htmlwidgets_1.6.4       S4Arrays_1.8.1         
 [85] whisker_0.4.1           uwot_0.2.3              ModelMetrics_1.2.2.2   
 [88] pkgconfig_2.0.3         gtable_0.3.6            timeDate_4041.110      
 [91] workflowr_1.7.1         lmtest_0.9-40           XVector_0.48.0         
 [94] htmltools_0.5.8.1       dotCall64_1.2           scales_1.4.0           
 [97] png_0.1-8               gower_1.0.2             spatstat.univar_3.1-4  
[100] knitr_1.50              reshape2_1.4.4          rjson_0.2.23           
[103] nlme_3.1-168            cachem_1.1.0            zoo_1.8-14             
[106] stringr_1.5.1           KernSmooth_2.23-26      parallel_4.5.1         
[109] miniUI_0.1.2            pillar_1.11.0           grid_4.5.1             
[112] vctrs_0.6.5             RANN_2.6.2              promises_1.3.3         
[115] cluster_2.1.8.1         evaluate_1.0.4          magick_2.8.7           
[118] cli_3.6.5               compiler_4.5.1          rlang_1.1.6            
[121] crayon_1.5.3            future.apply_1.20.0     labeling_0.4.3         
[124] mclust_6.1.1            plyr_1.8.9              fs_1.6.6               
[127] stringi_1.8.7           viridisLite_0.4.2       deldir_2.0-4           
[130] BiocParallel_1.42.1     lazyeval_0.2.2          spatstat.geom_3.5-0    
[133] RcppHNSW_0.6.0          future_1.67.0           statmod_1.5.0          
[136] shiny_1.11.1            ROCR_1.0-11             leidenAlg_1.1.5        
[139] igraph_2.1.4            bslib_0.9.0            </code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Red Hat Enterprise Linux 9.3 (Plow)

Matrix products: default
BLAS:   /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRblas.so 
LAPACK: /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRlapack.so;  LAPACK version 3.12.1

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Australia/Melbourne
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] harmony_1.2.3               Rcpp_1.1.0                 
 [3] Banksy_1.4.0                dotenv_1.0.3               
 [5] corrplot_0.95               xtable_1.8-4               
 [7] here_1.0.1                  limma_3.64.1               
 [9] RColorBrewer_1.1-3          patchwork_1.3.1            
[11] gridExtra_2.3               ggrepel_0.9.6              
[13] ComplexUpset_1.3.3          caret_7.0-1                
[15] lattice_0.22-7              glmnet_4.1-10              
[17] Matrix_1.7-3                tidyr_1.3.1                
[19] dplyr_1.1.4                 data.table_1.17.8          
[21] ggplot2_3.5.2               Seurat_5.3.0               
[23] SeuratObject_5.1.0          sp_2.2-0                   
[25] SpatialExperiment_1.18.1    SingleCellExperiment_1.30.1
[27] SummarizedExperiment_1.38.1 Biobase_2.68.0             
[29] GenomicRanges_1.60.0        GenomeInfoDb_1.44.2        
[31] IRanges_2.42.0              S4Vectors_0.46.0           
[33] BiocGenerics_0.54.0         generics_0.1.4             
[35] MatrixGenerics_1.20.0       matrixStats_1.5.0          
[37] jazzPanda_0.2.4            

loaded via a namespace (and not attached):
  [1] RcppHungarian_0.3       RcppAnnoy_0.0.22        splines_4.5.1          
  [4] later_1.4.2             tibble_3.3.0            polyclip_1.10-7        
  [7] hardhat_1.4.2           pROC_1.19.0.1           rpart_4.1.24           
 [10] fastDummies_1.7.5       lifecycle_1.0.4         aricode_1.0.3          
 [13] rprojroot_2.1.0         globals_0.18.0          MASS_7.3-65            
 [16] magrittr_2.0.3          plotly_4.11.0           sass_0.4.10            
 [19] rmarkdown_2.29          jquerylib_0.1.4         yaml_2.3.10            
 [22] httpuv_1.6.16           sctransform_0.4.2       spam_2.11-1            
 [25] spatstat.sparse_3.1-0   reticulate_1.43.0       cowplot_1.2.0          
 [28] pbapply_1.7-4           lubridate_1.9.4         abind_1.4-8            
 [31] Rtsne_0.17              presto_1.0.0            purrr_1.1.0            
 [34] BumpyMatrix_1.16.0      nnet_7.3-20             ipred_0.9-15           
 [37] git2r_0.36.2            lava_1.8.1              GenomeInfoDbData_1.2.14
 [40] irlba_2.3.5.1           listenv_0.9.1           spatstat.utils_3.1-5   
 [43] goftest_1.2-3           RSpectra_0.16-2         spatstat.random_3.4-1  
 [46] fitdistrplus_1.2-4      parallelly_1.45.1       codetools_0.2-20       
 [49] DelayedArray_0.34.1     tidyselect_1.2.1        shape_1.4.6.1          
 [52] UCSC.utils_1.4.0        farver_2.1.2            spatstat.explore_3.5-2 
 [55] jsonlite_2.0.0          progressr_0.15.1        ggridges_0.5.6         
 [58] survival_3.8-3          iterators_1.0.14        dbscan_1.2.2           
 [61] foreach_1.5.2           tools_4.5.1             ica_1.0-3              
 [64] glue_1.8.0              prodlim_2025.04.28      SparseArray_1.8.1      
 [67] xfun_0.52               withr_3.0.2             fastmap_1.2.0          
 [70] digest_0.6.37           timechange_0.3.0        R6_2.6.1               
 [73] mime_0.13               colorspace_2.1-1        scattermore_1.2        
 [76] sccore_1.0.6            tensor_1.5.1            spatstat.data_3.1-8    
 [79] hexbin_1.28.5           recipes_1.3.1           class_7.3-23           
 [82] httr_1.4.7              htmlwidgets_1.6.4       S4Arrays_1.8.1         
 [85] whisker_0.4.1           uwot_0.2.3              ModelMetrics_1.2.2.2   
 [88] pkgconfig_2.0.3         gtable_0.3.6            timeDate_4041.110      
 [91] workflowr_1.7.1         lmtest_0.9-40           XVector_0.48.0         
 [94] htmltools_0.5.8.1       dotCall64_1.2           scales_1.4.0           
 [97] png_0.1-8               gower_1.0.2             spatstat.univar_3.1-4  
[100] knitr_1.50              reshape2_1.4.4          rjson_0.2.23           
[103] nlme_3.1-168            cachem_1.1.0            zoo_1.8-14             
[106] stringr_1.5.1           KernSmooth_2.23-26      parallel_4.5.1         
[109] miniUI_0.1.2            pillar_1.11.0           grid_4.5.1             
[112] vctrs_0.6.5             RANN_2.6.2              promises_1.3.3         
[115] cluster_2.1.8.1         evaluate_1.0.4          magick_2.8.7           
[118] cli_3.6.5               compiler_4.5.1          rlang_1.1.6            
[121] crayon_1.5.3            future.apply_1.20.0     labeling_0.4.3         
[124] mclust_6.1.1            plyr_1.8.9              fs_1.6.6               
[127] stringi_1.8.7           viridisLite_0.4.2       deldir_2.0-4           
[130] BiocParallel_1.42.1     lazyeval_0.2.2          spatstat.geom_3.5-0    
[133] RcppHNSW_0.6.0          future_1.67.0           statmod_1.5.0          
[136] shiny_1.11.1            ROCR_1.0-11             leidenAlg_1.1.5        
[139] igraph_2.1.4            bslib_0.9.0            </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
