<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Melody Jin" />

<meta name="date" content="2026-02-18" />

<title>Xenium mouse brain - three replicates</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">jazzPanda_workflowr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="xenium-human-breast-cancer-analysis.html">Xenium breast cancer data</a>
    </li>
    <li>
      <a href="xenium-human-lung-cancer-analysis.html">Xenium human lung data</a>
    </li>
    <li>
      <a href="xenium-mouse-brain-analysis.html">Xenium mouse brain data</a>
    </li>
    <li>
      <a href="cosmx-human-healthy-liver.html">Cosmx healthy liver data</a>
    </li>
    <li>
      <a href="cosmx-human-liver-cancer.html">Cosmx liver cancer data</a>
    </li>
    <li>
      <a href="merscope-human-breast-cancer.html">Merscope breast cancer data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Xenium mouse brain - three replicates</h1>
<h4 class="author">Melody Jin</h4>
<h4 class="date">2026-02-18</h4>

</div>


<pre class="r"><code>library(spatstat)
library(jazzPanda)
library(SpatialExperiment)
library(Seurat)
library(ggplot2)
library(data.table)
library(dplyr)
library(glmnet)
library(caret)
library(ComplexUpset)
library(corrplot)
library(gridExtra)
library(patchwork)
library(RColorBrewer)
library(ggvenn)
library(limma)
library(edgeR)
library(speckle)
library(clustree)
library(xtable)

library(here)
source(here(&quot;code/utils.R&quot;))
library(Banksy)
library(harmony)</code></pre>
<div id="load-the-data" class="section level1">
<h1>Load the data</h1>
<pre class="r"><code>data_nm  &lt;- &quot;xenium_mbrain&quot;
rep1=get_xenium_data(path=rdata$xmb_r1, 
                     mtx_name = &quot;cell_feature_matrix/&quot;,
                     trans_name = &quot;transcripts.csv.gz&quot;,
                     cells_name=&quot;cells.csv.gz&quot;)

rep2=get_xenium_data(path=rdata$xmb_r2,
                     mtx_name = &quot;cell_feature_matrix/&quot;,
                     trans_name = &quot;transcripts.csv.gz&quot;,
                     cells_name=&quot;cells.csv.gz&quot;)

rep3=get_xenium_data(path=rdata$xmb_r3,
                     mtx_name = &quot;cell_feature_matrix/&quot;,
                     trans_name = &quot;transcripts.csv.gz&quot;,
                     cells_name=&quot;cells.csv.gz&quot;)

rep1$trans_info = rep1$trans_info[rep1$trans_info$qv &gt;=20 &amp; 
                                      rep1$trans_info$cell_id != -1 &amp; 
                                      !(rep1$trans_info$cell_id %in% rep1$zero_cells), ]
rep2$trans_info = rep2$trans_info[rep2$trans_info$qv &gt;=20 &amp;  
                                      rep2$trans_info$cell_id != -1 &amp; 
                                      !(rep2$trans_info$cell_id %in% rep2$zero_cells), ]
rep3$trans_info = rep3$trans_info[rep3$trans_info$qv &gt;=20 &amp; 
                                      rep3$trans_info$cell_id != -1 &amp; 
                                      !(rep3$trans_info$cell_id %in% rep3$zero_cells), ]</code></pre>
<pre class="r"><code>cm1 = rep1$cm
colnames(cm1) = paste(&quot;r1-&quot;, colnames(cm1), sep=&quot;&quot;)

cm2 = rep2$cm
colnames(cm2) = paste(&quot;r2-&quot;, colnames(cm2), sep=&quot;&quot;)

cm3 = rep3$cm
colnames(cm3) = paste(&quot;r3-&quot;, colnames(cm3), sep=&quot;&quot;)</code></pre>
</div>
<div id="single-cell-summary" class="section level1">
<h1>Single cell summary</h1>
<p>These summary plots characterise per-cell detection counts and
sparsity. We can calculate per-cell and per-gene summary statistics from
the counts matrix:</p>
<ul>
<li><p>Total detections per cell: sums all transcript counts in each
cell.</p></li>
<li><p>Proportion of zeroes per cell: fraction of genes not detected in
each cell.</p></li>
<li><p>Detected genes per cell: number of non-zero genes per
cell.</p></li>
<li><p>Average expression per gene: mean expression across cells,
ignoring zeros.</p></li>
</ul>
<p>Each distribution is visualised with histograms and density curves to
assess data quality and sparsity.</p>
<pre class="r"><code>td_r1 &lt;- colSums(cm1)
pz_r1 &lt;- colMeans(cm1==0)
numgene_r1 &lt;- colSums(cm1!=0)

td_r2 &lt;- colSums(cm2)
pz_r2 &lt;- colMeans(cm2==0)
numgene_r2 &lt;- colSums(cm2!=0)

td_r3 &lt;- colSums(cm3)
pz_r3 &lt;- colMeans(cm3==0)
numgene_r3 &lt;- colSums(cm3!=0)


# Build the entire summary as one string
output &lt;- paste0(
  &quot;\n================= Summary Statistics =================\n\n&quot;,
  &quot;--- Replicate 1 ---\n&quot;,
  make_sc_summary(td_r1, &quot;Total detections per cell:&quot;),
  make_sc_summary(pz_r1, &quot;Proportion of zeroes per cell:&quot;),
  make_sc_summary(numgene_r1, &quot;Detected genes per cell:&quot;),
  &quot;\n---Replicate 2 ---\n&quot;,
  make_sc_summary(td_r2, &quot;Total detections per cell:&quot;),
  make_sc_summary(pz_r2, &quot;Proportion of zeroes per cell:&quot;),
  make_sc_summary(numgene_r2, &quot;Detected genes per cell:&quot;),
  &quot;\n--- Replicate 3 ---\n&quot;,
  make_sc_summary(td_r3, &quot;Total detections per cell:&quot;),
  make_sc_summary(pz_r3, &quot;Proportion of zeroes per cell:&quot;),
  make_sc_summary(numgene_r3, &quot;Detected genes per cell:&quot;),
  &quot;\n========================================================\n&quot;
)

cat(output)</code></pre>
<pre><code>
================= Summary Statistics =================

--- Replicate 1 ---
Total detections per cell:
  Min: 2.0000 |  1Q: 153.0000 |  Median: 252.0000 |  Mean: 297.3727 |  3Q: 399.0000 |  Max: 2353.0000
Proportion of zeroes per cell:
  Min: 0.2863 |  1Q: 0.5927 |  Median: 0.6694 |  Mean: 0.6722 |  3Q: 0.7500 |  Max: 0.9919
Detected genes per cell:
  Min: 2.0000 |  1Q: 62.0000 |  Median: 82.0000 |  Mean: 81.3008 |  3Q: 101.0000 |  Max: 177.0000

---Replicate 2 ---
Total detections per cell:
  Min: 1.0000 |  1Q: 152.0000 |  Median: 250.0000 |  Mean: 294.2472 |  3Q: 396.0000 |  Max: 1817.0000
Proportion of zeroes per cell:
  Min: 0.3266 |  1Q: 0.6008 |  Median: 0.6774 |  Mean: 0.6788 |  3Q: 0.7540 |  Max: 0.9960
Detected genes per cell:
  Min: 1.0000 |  1Q: 61.0000 |  Median: 80.0000 |  Mean: 79.6632 |  3Q: 99.0000 |  Max: 167.0000

--- Replicate 3 ---
Total detections per cell:
  Min: 1.0000 |  1Q: 152.0000 |  Median: 253.0000 |  Mean: 297.7381 |  3Q: 400.0000 |  Max: 1974.0000
Proportion of zeroes per cell:
  Min: 0.2661 |  1Q: 0.6048 |  Median: 0.6774 |  Mean: 0.6809 |  3Q: 0.7581 |  Max: 0.9960
Detected genes per cell:
  Min: 1.0000 |  1Q: 60.0000 |  Median: 80.0000 |  Mean: 79.1436 |  3Q: 98.0000 |  Max: 182.0000

========================================================</code></pre>
<pre class="r"><code>td = as.data.frame(rbind(cbind(as.vector(td_r1),rep(&quot;replicate1&quot;, length(td_r1))),
                cbind(as.vector(td_r2),rep(&quot;replicate2&quot;, length(td_r2))), 
                cbind(as.vector(td_r3),rep(&quot;replicate3&quot;, length(td_r3)))))
colnames(td) = c(&quot;td&quot;,&quot;replicate&quot;)
td$td= as.numeric(td$td)

p1&lt;-ggplot(data = td, aes(x = td, color=replicate)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 100, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, size = 2) +
  facet_wrap(~replicate)+
  labs(title = &quot;Distribution of total detections per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  #xlim(0, 1000) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5), 
        strip.text = element_text(size=12)) 

pz = as.data.frame(rbind(cbind(as.vector(pz_r1),rep(&quot;replicate1&quot;, length(pz_r1))),
                cbind(as.vector(pz_r2),rep(&quot;replicate2&quot;, length(pz_r2))), 
                cbind(as.vector(pz_r3),rep(&quot;replicate3&quot;, length(pz_r3)))))
colnames(pz) = c(&quot;prop_ze&quot;,&quot;replicate&quot;)
pz$prop_ze= as.numeric(pz$prop_ze)

p2&lt;-ggplot(data = pz, aes(x = prop_ze, color=replicate)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 0.05, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, size = 2) +
  facet_wrap(~replicate)+
  labs(title = &quot;Distribution of proportion of zeroes per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  #xlim(0, 1) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))


numgens = as.data.frame(rbind(cbind(as.vector(numgene_r1),rep(&quot;replicate1&quot;, length(numgene_r1))),
                cbind(as.vector(numgene_r2),rep(&quot;replicate2&quot;, length(numgene_r2))), 
                cbind(as.vector(numgene_r3),rep(&quot;replicate3&quot;, length(numgene_r3)))))
colnames(numgens) = c(&quot;numgen&quot;,&quot;replicate&quot;)
numgens$numgen= as.numeric(numgens$numgen)

p3&lt;-ggplot(data = numgens, aes(x = numgen, color=replicate)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 10, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;steelblue&quot;, size = 2) +
  facet_wrap(~replicate)+
  labs(title = &quot;Distribution of detected genes per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  #xlim(0, 1) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))




layout_design &lt;- p1/p2/p3

layout_design &lt;- layout_design + 
                 patchwork::plot_layout(widths = c(1, 1,1), heights = c(1, 1 ,1)) 

layout_design</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/sc_summary-1.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code>cm_new1=cm1
cm_new1[cm_new1==0] = NA
cm_new1 = as.data.frame(cm_new1)
cm_new1$avg2 = rowMeans(cm_new1,na.rm = TRUE)
summary(cm_new1$avg2)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.21    2.00    2.59    3.09    3.69   15.73 </code></pre>
<pre class="r"><code>cm_new2=cm2
cm_new2[cm_new2==0] = NA
cm_new2 = as.data.frame(cm_new2)
cm_new2$avg2 = rowMeans(cm_new2,na.rm = TRUE)
summary(cm_new2$avg2)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.22    1.99    2.63    3.13    3.75   15.89 </code></pre>
<pre class="r"><code>cm_new3=cm3
cm_new3[cm_new3==0] = NA
cm_new3 = as.data.frame(cm_new3)
cm_new3$avg2 = rowMeans(cm_new3,na.rm = TRUE)
summary(cm_new3$avg2)</code></pre>
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   1.24    2.01    2.66    3.18    3.81   16.16 </code></pre>
<pre class="r"><code>avg_exp = as.data.frame(cbind(&quot;avg&quot;=c(cm_new1$avg2, cm_new2$avg2, cm_new3$avg2),
                             &quot;sample&quot;=c(rep(&quot;replicate1&quot;, nrow(cm1)), 
                               rep(&quot;replicate2&quot;, nrow(cm2)), 
                               rep(&quot;replicate3&quot;, nrow(cm3)) )))

avg_exp$avg=as.numeric(avg_exp$avg)
avg_exp$sample = factor(avg_exp$sample,
                        levels = c(&quot;replicate1&quot;, &quot;replicate2&quot;,&quot;replicate3&quot;))

ggplot(data = avg_exp, aes(x = avg, color=sample)) +
  geom_histogram(aes(y = after_stat(density)), 
                 binwidth = 0.5, fill = &quot;gray&quot;, color = &quot;black&quot;) +
  geom_density(color = &quot;orange&quot;, linewidth = 1) +
  facet_wrap(~sample)+
  labs(title = &quot;Distribution of average gene expression per cell&quot;,
       x = &quot; &quot;, y = &quot;Density&quot;) +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        strip.text = element_text(size=12))</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/ave-expression-1.png" width="576" style="display: block; margin: auto;" /></p>
<p>The three replicates show highly similar distributions. The median
total detections per cell range from 250–253, and the median number of
detected genes is around 80. The median proportion of zero counts is
about 0.67–0.68. The average gene expression is about 2.59, 2.63, and
2.66 for replciate 1, 2 and 3.</p>
</div>
<div id="quality-assessment-of-negative-controls"
class="section level1">
<h1>Quality assessment of negative controls</h1>
<div id="negative-control-probes" class="section level2">
<h2>Negative control probes</h2>
<p>Negative control probes are non-targeting probes designed not to
hybridize to any transcript. A total of 27 negative control probe
targets were included in this dataset, with median total detections of
approximately 1078, 831, and 779 for replicates 1, 2, and 3,
respectively. There are no particular localised patterns for these
probes.</p>
<pre class="r"><code>probe_coords &lt;- as.data.frame(rbind(cbind(sample=&quot;replicate1&quot;,rep1$trans_info[rep1$trans_info$feature_name %in% rep1$probe, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate2&quot;,rep2$trans_info[rep2$trans_info$feature_name %in% rep2$probe, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate3&quot;,rep3$trans_info[rep3$trans_info$feature_name %in% rep3$probe, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)])
                                        ))

codeword_coords &lt;- as.data.frame(rbind(cbind(sample=&quot;replicate1&quot;,rep1$trans_info[rep1$trans_info$feature_name %in% rep1$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate2&quot;,rep2$trans_info[rep2$trans_info$feature_name %in% rep2$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate3&quot;,rep3$trans_info[rep3$trans_info$feature_name %in% rep3$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)])
                                        ))
ordered_feature = probe_coords %&gt;% group_by(feature_name) %&gt;% count() %&gt;% arrange(desc(n))%&gt;% pull(feature_name) 
probe_tb = as.data.frame(probe_coords %&gt;% group_by(sample, feature_name) %&gt;% count())
colnames(probe_tb) = c(&quot;sample&quot;,&quot;feature_name&quot;,&quot;value_count&quot;)
probe_tb$feature_name = factor(probe_tb$feature_name, levels= ordered_feature)

probe_tb = probe_tb[order(probe_tb$feature_name), ]

ggplot(probe_tb, aes(x = feature_name, y = value_count)) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge()) + 
    facet_wrap(~sample, ncol=1, scales = &quot;free_y&quot;)+
    labs(title = &quot; &quot;, x = &quot; &quot;, y = &quot;Number of total detections&quot;) +
    theme_bw()+
   theme(legend.position = &quot;none&quot;,
          axis.text.x= element_text(size=8, angle=45,vjust=1,hjust = 1),
          axis.text.y=element_text(size=10),
          axis.ticks=element_blank(),
          strip.text = element_text(size=12),
          axis.title.y = element_text(size=12),
          axis.title.x=element_blank(),
          strip.background = element_rect(fill=&quot;white&quot;, colour =&quot;black&quot;),
          panel.background=element_blank(),
          panel.grid.minor=element_blank(),
          plot.background=element_blank())</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/nc_probe-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p1&lt;-ggplot(probe_coords, aes(x = x, y = y)) +
            geom_hex(bins = auto_hex_bin(min(table(probe_coords$sample)), 
                                         target_points_per_bin = 1)) +
            theme_bw() +
            scale_fill_gradient(low = &quot;white&quot;, high = &quot;black&quot;) +
            guides(fill = guide_colorbar(barheight = unit(0.06, &quot;npc&quot;),
                                         barwidth  = unit(0.4, &quot;npc&quot;),
                                         )) +
            scale_x_reverse()+
            facet_grid(~sample) +
            defined_theme +
            theme(legend.title = element_text(size = 10,
                                              hjust = 1, vjust=0.8),
                legend.position = &quot;bottom&quot;,
                aspect.ratio = 7/10,
                legend.text = element_text(size = 9),
                strip.text = element_blank(),
                strip.background = element_blank())

p1</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/neg_probe_xy-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
<div id="negative-control-codewords" class="section level2">
<h2>Negative control codewords</h2>
<p>Negative control codewords are unused barcode sequences included to
detect random or erroneous decoding events, providing a measure of
decoding noise. There are 41 codeword targets used in this dataset, and
the overall detection profiles were consistent across the three
replicates. The median total detections per codeword ranged from
approximately 40 to 65, with replicate 1 showing slightly lower counts
and replicate 3 showing the highest overall detection levels. The
codewords appear randomly distributed across the tissue sections.</p>
<pre class="r"><code>codeword_coords &lt;- as.data.frame(rbind(cbind(sample=&quot;replicate1&quot;,
                                             rep1$trans_info[rep1$trans_info$feature_name %in% rep1$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate2&quot;,rep2$trans_info[rep2$trans_info$feature_name %in% rep2$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)]),
                                        cbind(sample=&quot;replicate3&quot;,rep3$trans_info[rep3$trans_info$feature_name %in% rep3$codeword, c(&quot;feature_name&quot;,&quot;x&quot;,&quot;y&quot;)])
                                        ))
ordered_feature = codeword_coords %&gt;% group_by(feature_name) %&gt;% count() %&gt;% arrange(desc(n))%&gt;% pull(feature_name) 
codeword_tb = as.data.frame(codeword_coords %&gt;% group_by(sample, feature_name) %&gt;% count())
colnames(codeword_tb) = c(&quot;sample&quot;,&quot;feature_name&quot;,&quot;value_count&quot;)
codeword_tb$feature_name = factor(codeword_tb$feature_name, levels= ordered_feature)

codeword_tb = codeword_tb[order(codeword_tb$feature_name), ]

ggplot(codeword_tb, aes(x = feature_name, y = value_count)) +
    geom_bar(stat = &quot;identity&quot;, position = position_dodge()) + 
    facet_wrap(~sample, ncol=1, scales = &quot;free_y&quot;)+
    labs(title = &quot; &quot;, x = &quot; &quot;, y = &quot;Number of total detections&quot;) +
    theme_bw()+
    theme(legend.position = &quot;none&quot;,
          axis.text.x= element_text(size=8, angle=45,vjust=1,hjust = 1),
          axis.text.y=element_text(size=10),
          axis.ticks=element_blank(),
          strip.text = element_text(size=12),
          axis.title.y = element_text(size=12),
          axis.title.x=element_blank(),
          strip.background = element_rect(fill=&quot;white&quot;, colour =&quot;black&quot;),
          panel.background=element_blank(),
          panel.grid.minor=element_blank(),
          plot.background=element_blank())</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/nc_codeword-1.png" width="1152" style="display: block; margin: auto;" /></p>
<pre class="r"><code>p2&lt;- ggplot(codeword_coords, aes(x = x, y = y)) +
            geom_hex(bins = auto_hex_bin(min(table(probe_coords$sample)), 
                                         target_points_per_bin = 1)) +
            theme_bw() +
            scale_fill_gradient(low = &quot;white&quot;, high = &quot;black&quot;) +
            guides(fill = guide_colorbar(barheight = unit(0.06, &quot;npc&quot;),
                                         barwidth  = unit(0.4, &quot;npc&quot;),
                                         )) +
            scale_x_reverse()+
            facet_grid(~sample) +
            defined_theme +
            theme(legend.title = element_text(size = 10, hjust = 1, vjust=0.8),
                legend.position = &quot;bottom&quot;,
                aspect.ratio = 7/10,
                legend.text = element_text(size = 9),
                strip.text = element_blank(),
                strip.background = element_blank())

p2</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/nc_codeword_xy-1.png" width="864" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="banksy-clustering" class="section level1">
<h1>Banksy clustering</h1>
<p>We combined the three replicates into a single SpatialExperiment
object after assigning unique cell IDs and log-normalizing the data
through Seurat. BANKSY was then run with λ = 0.2 and k_geom = 15, using
adaptive geometric features to capture local spatial context. We
performed dimensional reduction (PCA and UMAP) and clustered the data at
resolutions 0.1. The resulting clusters were merged across samples to
obtain consistent labels for marker detection and annotation.</p>
<pre class="r"><code># List of source directories
sps &lt;- c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;)

spe_list &lt;- lapply(sps, function(id) {
    rp =  get(paste0(&quot;rep&quot;, id))
    cell_info &lt;-rp$cell_info
    cell_info$cells = paste0(&quot;r&quot;, id, &quot;-&quot;,cell_info$cell_id)
    cm &lt;-get(paste0(&quot;cm&quot;, id))
    # Subset and rename cells
    sub_info &lt;- cell_info[, c(&quot;x_centroid&quot;, &quot;y_centroid&quot;, &quot;cells&quot;)]
    colnames(sub_info) = c(&quot;x&quot;, &quot;y&quot;, &quot;cells&quot;)
    rownames(sub_info) &lt;- sub_info$cells
    sub_info = sub_info[colnames(cm),]
    coords &lt;- as.matrix(sub_info[, c(&quot;x&quot;, &quot;y&quot;)])
    
    # Construct SpatialExperiment
    SpatialExperiment(
        assays = list(counts = cm),
        spatialCoords = coords,
        sample_id = paste0(&quot;replicate&quot;,as.character(id))
    )
})

se &lt;- do.call(cbind, spe_list)

# Convert to Seurat and normalize
seu &lt;- as.Seurat(se, data = NULL)
seu &lt;- NormalizeData(seu, normalization.method = &quot;LogNormalize&quot;)
seu &lt;- FindVariableFeatures(seu, nfeatures = nrow(seu))

# Back to SpatialExperiment
aname &lt;- &quot;logcounts&quot;
# assay(se, aname) &lt;- GetAssayData(seu)

logcounts_mat &lt;- GetAssayData(seu, slot = &quot;data&quot;)[, colnames(se)]
assay(se, &quot;logcounts&quot;, withDimnames = FALSE) &lt;- logcounts_mat


# Re-split by sample
spe_list &lt;- split(seq_len(ncol(se)), se$sample_id) |&gt;
    lapply(function(cols) se[, cols])

lambda &lt;- c(0.2)
k_geom &lt;- 15
use_agf &lt;- TRUE
compute_agf &lt;- TRUE

spe_list &lt;- lapply(spe_list, computeBanksy, assay_name = aname, 
                   compute_agf = compute_agf, k_geom = k_geom)

spe_joint &lt;- do.call(cbind, spe_list)

spe_joint &lt;- runBanksyPCA(spe_joint, use_agf = use_agf, 
                          lambda = lambda, group = &quot;sample_id&quot;, seed = 1000)

spe_joint &lt;- runBanksyUMAP(spe_joint, use_agf = use_agf, 
                           lambda = lambda, seed = 1000)

cat(&quot;Clustering starts\n&quot;)
spe_joint &lt;- clusterBanksy(spe_joint, use_agf = use_agf, lambda = lambda,
                           resolution = c(0.1, 0.5), seed = 1000)

spe_joint &lt;- connectClusters(spe_joint)

cluster_info &lt;- data.frame(
    x = spatialCoords(spe_joint)[, 1],
    y = spatialCoords(spe_joint)[, 2],
    cell_id = colnames(spe_joint), 
    cluster = paste0(&quot;c&quot;, colData(spe_joint)$clust_M1_lam0.2_k50_res0.1),  
    sample = colData(spe_joint)$sample_id
)

saveRDS(seu, here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))
saveRDS(spe_joint, here(gdata,paste0(data_nm, &quot;_se.Rds&quot;)))
saveRDS(cluster_info, here(gdata,paste0(data_nm, &quot;_clusters.Rds&quot;)))</code></pre>
<pre class="r"><code># load generated data  (see code above for how they were created)
cluster_info = readRDS(here(gdata,paste0(data_nm, &quot;_clusters.Rds&quot;)))
cluster_names = paste0(&quot;c&quot;,1:9)

fig_ratio = cluster_info %&gt;%
  group_by(sample) %&gt;%
  summarise(
    x_range = diff(range(x, na.rm = TRUE)),
    y_range = diff(range(y, na.rm = TRUE)),
    ratio   = y_range / x_range,
    .groups = &quot;drop&quot;
  ) %&gt;% summarise(max_ratio = max(ratio)) %&gt;%  pull(max_ratio)

seu = readRDS(here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))
se = readRDS(here(gdata,paste0(data_nm, &quot;_se.Rds&quot;)))

Idents(seu)=cluster_info$cluster[match(colnames(seu), cluster_info$cell_id)]
Idents(seu) = factor(Idents(seu), levels = cluster_names)
seu$sample = cluster_info$sample[match(colnames(seu), cluster_info$cell_id)]</code></pre>
<p>There are 162,033, 154,654, and 158,047 cells in replicates 1, 2, and
3 respectively. The relative cluster proportions are consistent across
all three replicates. Clusters c1, c2, and c3 are the largest, together
accounting for over 70% of all cells in each replicate. In contrast,
smaller clusters such as c8 and c9 contian only around 1.2%-1.5% of
total cells.</p>
<pre class="r"><code>sum_df = cluster_info[,c(&quot;cluster&quot;, &quot;cell_id&quot;,&quot;sample&quot;),drop=FALSE] %&gt;% 
    group_by(cluster, sample) %&gt;% dplyr::count()

ggplot(sum_df, aes(x = cluster, y = n, fill=sample)) +
  geom_bar(stat=&quot;identity&quot;, position=position_dodge()) +
  scale_y_continuous(expand = c(0,10), limits = c(0, 1.1*max(sum_df$n)))+
  labs(title = &quot; &quot;, x = &quot;cluster&quot;, y = &quot;number of cells&quot;, fill=&quot; &quot;) +
  theme_bw()+
   theme(legend.position = &quot;top&quot;,
         axis.text= element_text(size=12),
         axis.title.y = element_text(size=12),
         axis.title.x = element_blank(),
         legend.text = element_text(size=12),
         strip.text = element_text(size = rel(1))
         )</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/ct_prop-1.png" width="960" style="display: block; margin: auto;" /></p>
<div id="cluster-visualisation" class="section level2">
<h2>Cluster visualisation</h2>
<p>Whole-sample view of every mouse brain replicate, colored by cluster.
The tissue architecture is well preserved, with major brain structures
clearly visible.</p>
<pre class="r"><code>p1&lt;- ggplot(data = cluster_info,aes(x = x, y = y, color=cluster))+
        geom_point(size=0.001, alpha=0.4)+
        facet_wrap(~sample, ncol=3)+
        scale_x_reverse()+
        theme_classic()+
        guides(color=guide_legend(title=&quot;&quot;, nrow = 1,
        override.aes=list(alpha=1, size=4)))+
        defined_theme + theme(aspect.ratio = fig_ratio,,
                              legend.position = &quot;top&quot;, 
                              legend.text = element_text(size=15),
                              strip.background = element_blank(),
                              strip.text = element_text(size = rel(1.2)))
p1</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/clusters-vis-xy-whole-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>Cluster-specific views showing that most clusters are spatially
separable.</p>
<pre class="r"><code>p1&lt;- ggplot(data = cluster_info[cluster_info$sample==&quot;replicate1&quot;, ],
        aes(x = x, y = y, color=cluster))+
        geom_point(size=0.001)+
        facet_wrap(~cluster, ncol=5)+
        scale_x_reverse()+
        scale_color_manual(values = my_colors)+
        theme_minimal()+
        defined_theme+
        labs(title = &quot;Mouse brain replicate1&quot;)+
        theme(aspect.ratio = fig_ratio,
               plot.title = element_text(size = rel(1.6),hjust = 0),
              strip.text = element_text(size = rel(1.2)))


p2&lt;- ggplot(data = cluster_info[cluster_info$sample==&quot;replicate2&quot;, ],
        aes(x = x, y = y, color=cluster))+
        geom_point(size=0.001)+
        facet_wrap(~cluster, ncol=5)+
        scale_x_reverse()+
        scale_color_manual(values = my_colors)+
        theme_minimal()+
        defined_theme+
        labs(title = &quot;Mouse brain replicate2&quot;)+
        theme(aspect.ratio =fig_ratio,
              plot.title = element_text(size = rel(1.6),hjust = 0),
              strip.text = element_text(size = rel(1.2)))

p3&lt;- ggplot(data = cluster_info[cluster_info$sample==&quot;replicate3&quot;, ],
        aes(x = x, y = y, color=cluster))+
        geom_point(size=0.001)+
        facet_wrap(~cluster, ncol=5)+
        scale_x_reverse()+
        scale_color_manual(values = my_colors)+
        theme_minimal()+
        defined_theme+
        labs(title = &quot;Mouse brain replicate3&quot;)+
        theme(aspect.ratio =fig_ratio,
              plot.title = element_text(size = rel(1.6), hjust = 0),
              strip.text = element_text(size = rel(1.2)))

layout_design &lt;- p1 / p2 / p3
layout_design</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/clusters-vis-xy-percluster-1.png" width="1440" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="detect-marker-genes" class="section level1">
<h1>Detect marker genes</h1>
<div id="jazzpanda" class="section level2">
<h2>jazzPanda</h2>
<p>We applied jazzPanda to this dataset using the linear modelling
approach to detect spatially enriched marker genes across clusters. The
dataset was converted into spatial gene and cluster vectors using a 50 ×
50 square binning strategy. Transcript detections for the negative
control probes and codewords targets were included as the background
control. For each gene, a linear model was fitted to estimate its
spatial association with relevant cell type patterns.</p>
<pre class="r"><code>all_genes =row.names(rep1$cm)
grid_length=50

# get spatial vectors
all_vectors = get_vectors(x= list(&quot;replicate1&quot; = rep1$trans_info,
                                  &quot;replicate2&quot; = rep2$trans_info,
                                  &quot;replicate3&quot; = rep3$trans_info), 
                          sample_names=c(&quot;replicate1&quot;,&quot;replicate2&quot;,&quot;replicate3&quot;),
                          cluster_info = cluster_info, bin_type=&quot;square&quot;,
                          bin_param=c(grid_length,grid_length), 
                          test_genes = all_genes , 
                          n_cores = 5)

nc_vectors = create_genesets(x=list(&quot;replicate1&quot; = rep1$trans_info,
                                    &quot;replicate2&quot; = rep2$trans_info,
                                    &quot;replicate3&quot; = rep3$trans_info), 
                             sample_names=c(&quot;replicate1&quot;,&quot;replicate2&quot;,&quot;replicate3&quot;),
                             name_lst=list(probe=rep1$probe, 
                                           codeword=rep1$codeword),
                             bin_type=&quot;square&quot;,
                             bin_param=c(grid_length,grid_length), 
                             cluster_info = NULL)
})


set.seed(188)
jazzPanda_res_lst = lasso_markers(gene_mt=all_vectors$gene_mt,
                                  cluster_mt = all_vectors$cluster_mt,
                                  sample_names=c(&quot;replicate1&quot;,&quot;replicate2&quot;,&quot;replicate3&quot;),
                                  keep_positive=TRUE, 
                                  background=nc_vectors)

saveRDS(all_vectors, here(gdata,paste0(data_nm, &quot;_sq50_vector_lst.Rds&quot;)))
saveRDS(jazzPanda_res_lst, here(gdata,paste0(data_nm, &quot;_jazzPanda_res_lst.Rds&quot;)))</code></pre>
<pre class="r"><code># load generated data  (see code above for how they were created)
jazzPanda_res_lst = readRDS(here(gdata,paste0(data_nm, &quot;_jazzPanda_res_lst.Rds&quot;)))
nbins = 2500
sv_lst = readRDS(here(gdata,paste0(data_nm, &quot;_sq50_vector_lst.Rds&quot;)))

jazzPanda_res = get_top_mg(jazzPanda_res_lst, coef_cutoff=0.1)  
jazzPanda_summary = as.data.frame(table(jazzPanda_res$top_cluster))
colnames(jazzPanda_summary) = c(&quot;cluster&quot;,&quot;jazzPanda_glm&quot;)</code></pre>
<p>This dotplot shows the top 10 marker genes per cluster identified
using the linear modelling approach. Some clusters (such as c7 and c9)
display relatively distinct marker profiles with strong and specific
expression signals. In contrast, clusters like c3 show some overlap in
gene expression, with markers moderately expressed across multiple
clusters rather than being cluster-restricted.</p>
<pre class="r"><code>glm_mg = jazzPanda_res[jazzPanda_res$top_cluster != &quot;NoSig&quot;, ]
glm_mg$top_cluster = factor(glm_mg$top_cluster, levels = cluster_names)
top_mg_glm &lt;- glm_mg %&gt;%
  group_by(top_cluster) %&gt;%
  arrange(desc(glm_coef), .by_group = TRUE) %&gt;% 
  slice_max(order_by = glm_coef, n = 10) %&gt;%
  select(top_cluster, gene, glm_coef)

p1&lt;-DotPlot(seu, features = top_mg_glm$gene) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (jazzPanda-glm)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/top10_glm-1.png" width="768" style="display: block; margin: auto;" /></p>
<div id="cluster-to-cluster-correlation" class="section level3">
<h3>Cluster to cluster correlation</h3>
<p>We computed pairwise Spearman correlations between spatial cluster
vectors to assess relationships among clusters for every replicate.
Overall, the pairwise correlations are generally low, indicating that
most clusters occupy distinct spatial regions. The highest correlations
(≈0.45–0.47) are observed between clusters c1 and c3, which show
partially overlapping spatial patterns.</p>
<pre class="r"><code>for (curr_sp in c(&quot;replicate1&quot;, &quot;replicate2&quot;, &quot;replicate3&quot;)){
    
    cluster_mtt &lt;- sv_lst$cluster_mt[sv_lst$cluster_mt[, curr_sp] == 1, ]
    cluster_mtt &lt;- cluster_mtt[, cluster_names]
    cor_M_cluster &lt;- cor(cluster_mtt, cluster_mtt, method = &quot;spearman&quot;)
    
    col &lt;- colorRampPalette(c(&quot;#4477AA&quot;, &quot;#77AADD&quot;, &quot;#FFFFFF&quot;,
                              &quot;#EE9988&quot;, &quot;#BB4444&quot;))
    
    corrplot(cor_M_cluster,
            method = &quot;color&quot;,
            col = col(200),
            diag = TRUE,
            addCoef.col = &quot;black&quot;,
            type = &quot;upper&quot;,
            tl.col = &quot;black&quot;,
            tl.cex = 1,
            number.cex = 0.8,
            tl.srt = 0,
            mar = c(0, 0, 3, 0),   
            sig.level = 0.05,
            insig = &quot;blank&quot;)
    
    title(paste(curr_sp, &quot;cluster–cluster correlation&quot;),
            line = 1,               
            cex.main = 1.2)

}</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/cc-corr-1.png" width="576" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/cc-corr-2.png" width="576" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/cc-corr-3.png" width="576" style="display: block; margin: auto;" /></p>
<pre class="r"><code># Combine into a new dataframe for LaTeX output
output_data &lt;- do.call(rbind, lapply(cluster_names, function(cl) {
  subset &lt;- jazzPanda_res[jazzPanda_res$top_cluster == cl, ]
  sorted_subset &lt;- subset[order(-subset$glm_coef), ]
  gene_list &lt;- paste(sorted_subset$gene, &quot;(&quot;, round(sorted_subset$glm_coef, 2), &quot;)&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
 
  return(data.frame(Cluster = cl, Genes = gene_list))
}))

latex_table &lt;- xtable(output_data, caption=&quot;Detected marker genes for each cluster by jazzPanda, in decreasing value of glm coefficients&quot;)
print.xtable(latex_table, include.rownames = FALSE, hline.after = c(-1, 0, nrow(output_data)), comment = FALSE)</code></pre>
<pre><code>\begin{table}[ht]
\centering
\begin{tabular}{ll}
  \hline
Cluster &amp; Genes \\ 
  \hline
c1 &amp; Id2(7.78), Garnl3(7.63), Dpy19l1(7.18), Lamp5(6.47), Nrep(5.23), Plcxd2(4.59), Fhod3(4.5), Gfra2(4.31), Satb2(4.28), Bhlhe40(4.14), Kcnh5(4.12), Rasgrf2(3.51), Cux2(3.31), Tmem132d(3.18), Tle4(3.08), Pvalb(2.98), Hs3st2(2.35), Grik3(2.34), Neto2(2.29), Nxph3(2.24), Gng12(2.1), Slit2(2.07), Igfbp6(1.99), Rab3b(1.91), Pdzrn3(1.75), Myl4(1.7), Fezf2(1.69), Syndig1(1.64), Igsf21(1.51), Ndst3(1.47), Sema3a(1.38), Cdh6(1.37), Myo16(1.27), Gsg1l(1.26), Cntnap4(1.22), Lypd6(1.11), Gadd45a(1.06), Syt2(1.06), Cplx3(0.98), Galnt14(0.97), Rspo1(0.95), Ccn2(0.94), Prr16(0.92), Cbln4(0.87), Arhgap25(0.86), Rxfp1(0.73), Col19a1(0.73), Rbp4(0.62), Rspo2(0.55), Laptm5(0.54), Fos(0.46), Sema5b(0.43), Vip(0.43), Trem2(0.42), Siglech(0.39), Cwh43(0.35), Kcnmb2(0.33), Trbc2(0.31), Gm19410(0.29), Sla(0.28), Cort(0.27), Gli3(0.26), Sst(0.26), Crh(0.21) \\ 
  c2 &amp; Rmst(5.92), Calb2(4.49), Gjc3(4.06), Opalin(2.34), Tmem255a(2.12), Zfp536(1.89), Sox10(1.88), Chrm2(1.23), Nts(1.15), Ebf3(1.07), Adamts2(0.61), Gpr17(0.43), Stard5(0.36), Sema3d(0.31), Chodl(0.22) \\ 
  c3 &amp; Slc17a7(21.79), Dner(13.69), Nr2f2(7.96), Bhlhe22(7.25), Hpcal1(5.23), Rprm(4.15), Cdh13(4.12), Kctd12(3.99), Arhgef28(2.83), Plcxd3(2.79), Cdh4(2.28), Bdnf(2), Cdh9(1.66), Parm1(1.52), Dpyd(1.33), Npy2r(1.29), Syt17(1.28), Zfpm2(1.26), Pde11a(0.98), Ndst4(0.92), Hapln1(0.64), Pcsk5(0.58), Angpt1(0.56), Cspg4(0.48), Prss35(0.21) \\ 
  c4 &amp; Rims3(17.98), Slc17a6(17.67), Clmn(15.71), Nrn1(13.7), Rorb(6.91), Nell1(6.48), Necab1(6.42), Acsbg1(5.48), Rnf152(5.36), Car4(5.13), Tanc1(4.81), Tmem163(4.78), Sema6a(4.4), Ntsr2(3.85), Foxp2(3.76), Inpp4b(2.77), Unc13c(2.55), Opn3(2.25), Btbd11(2.13), Cntn6(1.95), Sdk2(1.75), Tox(1.7), Cdh20(1.67), Ly6a(1.51), Fign(1.42), Plch1(1.29), Cntnap5b(1.07), Vwc2l(0.99), Rfx4(0.97), Cbln1(0.94), Cldn5(0.87), Igf1(0.74), Adgrl4(0.66), Deptor(0.59), Pkib(0.39), Pglyrp1(0.33), Nostrin(0.19), Zfp366(0.13) \\ 
  c5 &amp; Igf2(20.4), Slc13a4(9.35), Vat1l(8.41), Gfap(6.1), Mapk4(5.54), Dcn(5.36), Slc39a12(4.66), Strip2(4.13), Pdgfra(3.45), Aqp4(2.87), Fn1(2.72), Cobll1(2.63), Cd24a(2.55), Aldh1a2(2.54), Fmod(2.54), Col1a1(2.41), Acta2(2.28), Spag16(2.18), Lyz2(2.08), Spp1(1.93), Kdr(1.86), Hat1(1.77), Gjb2(1.61), Sntb1(1.31), Paqr5(1.2), Cyp1b1(1.19), Pecam1(1.08), Adamtsl1(1.04), Mecom(0.96), Emcn(0.86), Acvrl1(0.72), Trp73(0.62), Cd93(0.62), Carmn(0.6), Prph(0.59), Eya4(0.43), Pln(0.37), Cd68(0.34), Cd53(0.33), Cd44(0.32), Fgd5(0.3), Sox17(0.23), Slfn5(0.23), Pthlh(0.21) \\ 
  c6 &amp; Meis2(12.44), Penk(11.51), Ppp1r1b(10.21), Gad1(7.38), Pde7b(5.67), Gad2(5.64), Sox11(3.4), Col6a1(3.36), Cacna2d2(2.31), Htr1f(1.77), Pdyn(1.69), Arhgap6(1.05) \\ 
  c7 &amp; Prox1(10.76), Plekha2(4.82), Calb1(4.73), Prdm8(4.02), Igfbp5(3.58), Rasl10a(3.05), Pdzd2(2.44), Slc44a5(2.37), Orai2(1.52), Npnt(1.45) \\ 
  c8 &amp; Neurod6(20.59), Epha4(17.96), 2010300C02Rik(14.76), Dkk3(12.47), Cabp7(10.94), Gm2115(9.87), Arc(8.95), Wfs1(8.56), Bcl11b(8.27), Cpne6(8.26), Sipa1l3(7.74), Pou3f1(7.56), Igfbp4(6.83), Sema3e(6.53), Arhgap12(6.25), Pip5k1b(6), Trpc4(4), Fibcd1(3.98), Sorcs3(3.89), Mdga1(3.49), Shisa6(3.35), Cpne8(3.07), Ror1(0.83) \\ 
  c9 &amp; Nwd2(19.13), Necab2(10.2), Cpne4(8.47), Gucy1a1(7.32), Syt6(6.99), Nrp2(6.24), Thsd7a(4.39), Sncg(2.87), Kctd8(2.57), Chat(1.59), Ano1(1.22), Tacr1(1.05) \\ 
   \hline
\end{tabular}
\caption{Detected marker genes for each cluster by jazzPanda, in decreasing value of glm coefficients} 
\end{table}</code></pre>
</div>
<div id="top3-marker-genes-by-jazzpanda_glm" class="section level3">
<h3>Top3 marker genes by jazzPanda_glm</h3>
<p>We visualized the top three marker genes identified by the linear
modelling approach for each cluster. For each selected gene, its spatial
transcript locations were plotted alongside the corresponding cluster
map. Across clusters, the spatial expression patterns of marker genes
closely match the distribution of their associated cluster.</p>
<pre class="r"><code>x_range &lt;- c(0, 11000)
y_range &lt;- c(0, 7500)

plot_lst=list()
for (cl in cluster_names){
    inters=jazzPanda_res[jazzPanda_res$top_cluster==cl,&quot;gene&quot;]
    rounded_val=signif(as.numeric(jazzPanda_res[inters,&quot;glm_coef&quot;]),
                          digits = 3)
    inters_df = as.data.frame(cbind(gene=inters, value=rounded_val))
    inters_df$value = as.numeric(inters_df$value)
    inters_df=inters_df[order(inters_df$value, decreasing = TRUE),]
    inters_df$text= paste(inters_df$gene,inters_df$value,sep=&quot;: &quot;)
    
    if (length(inters &gt; 0)){
        inters_df = inters_df[1:min(3, nrow(inters_df)),]
        inters = inters_df$gene
        all_trans = as.data.frame(matrix(0, ncol=6))
        colnames(all_trans) = c(&quot;x&quot;,&quot;y&quot;, &quot;feature_name&quot;,&quot;value&quot;,&quot;text_label&quot;,&quot;sample&quot;)
        for (rp in c(&quot;rep1&quot;,&quot;rep2&quot;,&quot;rep3&quot;)){
          rp_lst = get(rp)
          iters_sp1= rp_lst$trans_info$feature_name %in% inters
          vis_r1 =rp_lst$trans_info[iters_sp1,
                            c(&quot;x&quot;,&quot;y&quot;,&quot;feature_name&quot;)]
          vis_r1$value = inters_df[match(vis_r1$feature_name,inters_df$gene),
                            &quot;value&quot;]
          vis_r1=vis_r1[order(vis_r1$value,decreasing = TRUE),]
          vis_r1$text_label= paste(vis_r1$feature_name,
                                vis_r1$value,sep=&quot;: &quot;)
          vis_r1$text_label=factor(vis_r1$text_label, levels=inters_df$text)
          vis_r1$sample=paste(&quot;replicate&quot;, substr(rp,start = 4, stop=4),sep=&quot;&quot;)
          all_trans = rbind(all_trans,vis_r1)
        }
        all_trans = all_trans[2:nrow(all_trans),]
        all_trans$text_label = factor(all_trans$text_label, 
                                      levels = inters_df$text)
        all_trans = all_trans[order(all_trans$text_label),]

        genes_plt&lt;- ggplot(data = all_trans,
                    aes(x = x, y = y))+ 
                    geom_hex(bins = auto_hex_bin(max(table(all_trans$sample, 
                                                            all_trans$feature_name))))+
                    facet_grid(sample~text_label)+
                    scale_x_reverse() +
                    scale_fill_gradient(low=&quot;grey90&quot;, high=&quot;maroon4&quot;) + 
                    # scale_fill_viridis_c(option = &quot;turbo&quot;)+
                    guides(fill = guide_colorbar(barheight = unit(0.2, &quot;npc&quot;), 
                                                 barwidth  = unit(0.02, &quot;npc&quot;)))+
                    defined_theme+
            theme(legend.position = &quot;right&quot;, 
              strip.background = element_rect(fill =NA,colour = NA),
              strip.text = element_text(size = rel(1.3)), 
              strip.text.y.right = element_blank(),
              legend.key.width = unit(2, &quot;cm&quot;),
              aspect.ratio = fig_ratio,
              legend.title = element_text(size = 10),
              legend.text = element_text(size = 10),
               legend.key.height = unit(2, &quot;cm&quot;))
        
        cluster_data =  cluster_info[cluster_info$cluster==cl, ]
        cluster_data$label = paste(cluster_data$sample, cluster_data$cluster, sep=&quot;-&quot;)
        cluster_data$label =factor(cluster_data$label, 
                                   levels=paste(paste(&quot;replicate&quot;,1:3,sep=&quot;&quot;),
                                                cl,sep=&quot;-&quot;))
        p_cl&lt;- ggplot(data =cluster_data,
                    aes(x = x, y = y, color=sample))+ 
                    geom_point(size=0.01)+
                    facet_wrap(~label, nrow=3,strip.position = &quot;left&quot;)+
                    scale_x_reverse(limits = rev(x_range)) +  # enforce global x
                    scale_y_continuous(limits = y_range) +  # enforce global y

                    scale_color_manual(values = c(&quot;black&quot;,&quot;black&quot;,&quot;black&quot;))+
                    defined_theme + 
            theme(legend.position = &quot;none&quot;, 
                  strip.background = element_blank(), 
                  aspect.ratio = fig_ratio,
                  legend.key.width = unit(15, &quot;cm&quot;),
                  strip.text = element_text(size = rel(1.2)))

        lyt &lt;-p_cl | genes_plt
        layout_design &lt;- lyt + patchwork::plot_layout(heights = c(1,1),
                                                      widths = c(1, 3)) 
        print(layout_design)

    }
  }</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-1.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-2.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-3.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-4.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-5.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-6.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-7.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-8.png" width="960" style="display: block; margin: auto;" /><img src="xenium-mouse-brain-analysis_files/figure-html/top3-gene-vis-xy-9.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="cluster-vector-versus-marker-gene-vector"
class="section level3">
<h3>Cluster vector versus marker gene vector</h3>
<p>We visualized the top three marker genes for each cluster by plotting
their spatial gene vectors against the corresponding cluster vector.
Each scatter plot shows how strongly each marker gene’s spatial pattern
aligns with the overall spatial signal of its cluster, allowing
assessment of both marker specificity and coherence with spatial
domains.</p>
<p>For all clusters, marker genes show a strong positive trend with the
corresponding cluster vector, confirming that their expression is
spatially aligned with the inferred cluster domain.</p>
<pre class="r"><code>plot_lst=list()
for (cl in cluster_names){
    inters_df=jazzPanda_res[jazzPanda_res$top_cluster==cl,]

    inters_df=inters_df[order(inters_df$glm_coef, decreasing = TRUE),]
    
    inters_df = inters_df[1:min(3, nrow(inters_df)),]
    mk_genes = inters_df$gene

    dff = as.data.frame(cbind(sv_lst$cluster_mt[,cl],
                              sv_lst$gene_mt[,mk_genes]))
    colnames(dff) = c(&quot;cluster&quot;, mk_genes)
    dff$sample= &quot;replicate1&quot;
    dff[nbins:(nbins*2),&quot;sample&quot;] = &quot;replicate2&quot;
    dff[(nbins*2):(nbins*3),&quot;sample&quot;] = &quot;replicate3&quot;
    dff$vector_id = c(1:nbins, 1:nbins, 1:nbins)
    long_df &lt;- dff %&gt;% 
    tidyr::pivot_longer(cols = -c(cluster, sample, vector_id), 
                        names_to = &quot;gene&quot;, 
               values_to = &quot;vector_count&quot;)
    long_df$gene = factor(long_df$gene, levels=mk_genes)
    long_df$sample = factor(long_df$sample , 
                        levels=c(&quot;replicate1&quot;,&quot;replicate2&quot;, &quot;replicate3&quot;))
    p&lt;-ggplot(long_df, aes(x = cluster, y = vector_count, color=gene )) +
    geom_point(size=0.01) +
    facet_grid(sample~gene, scales=&quot;free_y&quot;) +
    labs(x = paste(cl, &quot; cluster vector &quot;, sep=&quot;&quot;), 
         y = &quot;Top3 marker gene vector&quot;) +
    theme_minimal()+
        scale_y_continuous(expand = c(0.01,0.01))+ 
        scale_x_continuous(expand =  c(0.01,0.01))+ 
        theme(panel.grid = element_blank(),legend.position = &quot;none&quot;,
        strip.text = element_text(size = 12),
        axis.line=element_blank(),
        axis.text=element_text(size = 9),
        axis.ticks=element_line(color=&quot;black&quot;),
        axis.title=element_text(size = 13),
        panel.border  =element_rect(colour = &quot;black&quot;, 
                                    fill=NA, linewidth=0.5)
        )
    plot_lst[[cl]] = p
}

combined_plot &lt;- wrap_plots(plot_lst, ncol = 3)

combined_plot </code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/vvplot-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="marker-gene-summary-per-cluster" class="section level3">
<h3>Marker gene summary per cluster</h3>
<p>For each cluster, we summarized the detected marker genes and their
corresponding strengths. We list genes with model coefficients above the
defined cutoff (e.g. 0.1) for each cluster.</p>
<p>Cluster c1 shows enrichment for Id2, Lamp5, and Cplx3, consistent
with VIP/LAMP5 interneurons, though some mixed excitatory markers
suggest partial transcriptional overlap. Cluster c2 expresses Opalin,
Gjc3, and Sox10, clearly corresponding to myelinating oligodendrocytes
(MOL). Cluster c3 (Bhlhe22, Hpcal1, Nr2f2) matches IT-L2/3 excitatory
neurons, while cluster c4 (Slc17a6, Foxp2, Ntsr2) represents layer 6
corticothalamic (L6 CT-like) neurons. Cluster c5 is dominated by Igf2,
Gfap, and Aqp4, marking choroid plexus and glial-enriched regions with
broad non-neuronal signatures. Cluster c6 (Ppp1r1b, Gad1, Penk) fits
medium spiny neurons (MSNs) of GABAergic lineage. Cluster c7 (Prox1,
Calb1) corresponds to dentate gyrus granule cells, whereas cluster c8
(Neurod6, Wfs1, Shisa6) identifies deep-layer IT-L6/CC projection
neurons. Finally, c9 shows Chat, Sncg, and Syt6, consistent with
ChAT⁺/SNCG⁺ interneurons.</p>
<pre class="r"><code>## linear modelling 
output_data &lt;- do.call(rbind, lapply(cluster_names, function(cl) {
  subset &lt;- jazzPanda_res[jazzPanda_res$top_cluster == cl, ]
  sorted_subset &lt;- subset[order(-subset$glm_coef), ]
  if (length(sorted_subset) == 0){
    return(data.frame(Cluster = cl, Genes = &quot;  &quot;))
  }
  if (cl != &quot;NoSig&quot;){
    gene_list &lt;- paste(sorted_subset$gene, &quot;(&quot;, round(sorted_subset$glm_coef, 2), &quot;)&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }else{
    gene_list &lt;- paste(sorted_subset$gene, &quot;&quot;, sep=&quot;&quot;, collapse=&quot;, &quot;)
  }
  
  return(data.frame(Cluster = cl, Genes = gene_list))
}))

knitr::kable(
  output_data,
  caption = &quot;Detected marker genes for each cluster by linear modelling approach in jazzPanda, in decreasing value of glm coefficients&quot;
)</code></pre>
<table>
<caption>Detected marker genes for each cluster by linear modelling
approach in jazzPanda, in decreasing value of glm coefficients</caption>
<colgroup>
<col width="0%" />
<col width="99%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Cluster</th>
<th align="left">Genes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">c1</td>
<td align="left">Id2(7.78), Garnl3(7.63), Dpy19l1(7.18), Lamp5(6.47),
Nrep(5.23), Plcxd2(4.59), Fhod3(4.5), Gfra2(4.31), Satb2(4.28),
Bhlhe40(4.14), Kcnh5(4.12), Rasgrf2(3.51), Cux2(3.31), Tmem132d(3.18),
Tle4(3.08), Pvalb(2.98), Hs3st2(2.35), Grik3(2.34), Neto2(2.29),
Nxph3(2.24), Gng12(2.1), Slit2(2.07), Igfbp6(1.99), Rab3b(1.91),
Pdzrn3(1.75), Myl4(1.7), Fezf2(1.69), Syndig1(1.64), Igsf21(1.51),
Ndst3(1.47), Sema3a(1.38), Cdh6(1.37), Myo16(1.27), Gsg1l(1.26),
Cntnap4(1.22), Lypd6(1.11), Gadd45a(1.06), Syt2(1.06), Cplx3(0.98),
Galnt14(0.97), Rspo1(0.95), Ccn2(0.94), Prr16(0.92), Cbln4(0.87),
Arhgap25(0.86), Rxfp1(0.73), Col19a1(0.73), Rbp4(0.62), Rspo2(0.55),
Laptm5(0.54), Fos(0.46), Sema5b(0.43), Vip(0.43), Trem2(0.42),
Siglech(0.39), Cwh43(0.35), Kcnmb2(0.33), Trbc2(0.31), Gm19410(0.29),
Sla(0.28), Cort(0.27), Gli3(0.26), Sst(0.26), Crh(0.21)</td>
</tr>
<tr class="even">
<td align="left">c2</td>
<td align="left">Rmst(5.92), Calb2(4.49), Gjc3(4.06), Opalin(2.34),
Tmem255a(2.12), Zfp536(1.89), Sox10(1.88), Chrm2(1.23), Nts(1.15),
Ebf3(1.07), Adamts2(0.61), Gpr17(0.43), Stard5(0.36), Sema3d(0.31),
Chodl(0.22)</td>
</tr>
<tr class="odd">
<td align="left">c3</td>
<td align="left">Slc17a7(21.79), Dner(13.69), Nr2f2(7.96),
Bhlhe22(7.25), Hpcal1(5.23), Rprm(4.15), Cdh13(4.12), Kctd12(3.99),
Arhgef28(2.83), Plcxd3(2.79), Cdh4(2.28), Bdnf(2), Cdh9(1.66),
Parm1(1.52), Dpyd(1.33), Npy2r(1.29), Syt17(1.28), Zfpm2(1.26),
Pde11a(0.98), Ndst4(0.92), Hapln1(0.64), Pcsk5(0.58), Angpt1(0.56),
Cspg4(0.48), Prss35(0.21)</td>
</tr>
<tr class="even">
<td align="left">c4</td>
<td align="left">Rims3(17.98), Slc17a6(17.67), Clmn(15.71), Nrn1(13.7),
Rorb(6.91), Nell1(6.48), Necab1(6.42), Acsbg1(5.48), Rnf152(5.36),
Car4(5.13), Tanc1(4.81), Tmem163(4.78), Sema6a(4.4), Ntsr2(3.85),
Foxp2(3.76), Inpp4b(2.77), Unc13c(2.55), Opn3(2.25), Btbd11(2.13),
Cntn6(1.95), Sdk2(1.75), Tox(1.7), Cdh20(1.67), Ly6a(1.51), Fign(1.42),
Plch1(1.29), Cntnap5b(1.07), Vwc2l(0.99), Rfx4(0.97), Cbln1(0.94),
Cldn5(0.87), Igf1(0.74), Adgrl4(0.66), Deptor(0.59), Pkib(0.39),
Pglyrp1(0.33), Nostrin(0.19), Zfp366(0.13)</td>
</tr>
<tr class="odd">
<td align="left">c5</td>
<td align="left">Igf2(20.4), Slc13a4(9.35), Vat1l(8.41), Gfap(6.1),
Mapk4(5.54), Dcn(5.36), Slc39a12(4.66), Strip2(4.13), Pdgfra(3.45),
Aqp4(2.87), Fn1(2.72), Cobll1(2.63), Cd24a(2.55), Aldh1a2(2.54),
Fmod(2.54), Col1a1(2.41), Acta2(2.28), Spag16(2.18), Lyz2(2.08),
Spp1(1.93), Kdr(1.86), Hat1(1.77), Gjb2(1.61), Sntb1(1.31), Paqr5(1.2),
Cyp1b1(1.19), Pecam1(1.08), Adamtsl1(1.04), Mecom(0.96), Emcn(0.86),
Acvrl1(0.72), Trp73(0.62), Cd93(0.62), Carmn(0.6), Prph(0.59),
Eya4(0.43), Pln(0.37), Cd68(0.34), Cd53(0.33), Cd44(0.32), Fgd5(0.3),
Sox17(0.23), Slfn5(0.23), Pthlh(0.21)</td>
</tr>
<tr class="even">
<td align="left">c6</td>
<td align="left">Meis2(12.44), Penk(11.51), Ppp1r1b(10.21), Gad1(7.38),
Pde7b(5.67), Gad2(5.64), Sox11(3.4), Col6a1(3.36), Cacna2d2(2.31),
Htr1f(1.77), Pdyn(1.69), Arhgap6(1.05)</td>
</tr>
<tr class="odd">
<td align="left">c7</td>
<td align="left">Prox1(10.76), Plekha2(4.82), Calb1(4.73), Prdm8(4.02),
Igfbp5(3.58), Rasl10a(3.05), Pdzd2(2.44), Slc44a5(2.37), Orai2(1.52),
Npnt(1.45)</td>
</tr>
<tr class="even">
<td align="left">c8</td>
<td align="left">Neurod6(20.59), Epha4(17.96), 2010300C02Rik(14.76),
Dkk3(12.47), Cabp7(10.94), Gm2115(9.87), Arc(8.95), Wfs1(8.56),
Bcl11b(8.27), Cpne6(8.26), Sipa1l3(7.74), Pou3f1(7.56), Igfbp4(6.83),
Sema3e(6.53), Arhgap12(6.25), Pip5k1b(6), Trpc4(4), Fibcd1(3.98),
Sorcs3(3.89), Mdga1(3.49), Shisa6(3.35), Cpne8(3.07), Ror1(0.83)</td>
</tr>
<tr class="odd">
<td align="left">c9</td>
<td align="left">Nwd2(19.13), Necab2(10.2), Cpne4(8.47), Gucy1a1(7.32),
Syt6(6.99), Nrp2(6.24), Thsd7a(4.39), Sncg(2.87), Kctd8(2.57),
Chat(1.59), Ano1(1.22), Tacr1(1.05)</td>
</tr>
</tbody>
</table>
</div>
<div id="nosig-genes" class="section level3">
<h3>Nosig genes</h3>
<p>These five genes (Cd300c2, Ikzf1, Slc6a3, Spi1, and Th) show low or
spatially restricted expression across replicates. Their known biology
suggests that most are associated with immune or dopaminergic lineages,
which are not well represented among the nine annotated cortical and
hippocampal cell types. The weak and patchy expression patterns observed
here are therefore expected, and it is reasonable that our method did
not identify them as cluster-defining marker genes.</p>
<pre class="r"><code>cl=&quot;NoSig&quot;
inters=jazzPanda_res[jazzPanda_res$top_cluster==cl,&quot;gene&quot;]
all_trans = as.data.frame(matrix(0, ncol=4))
colnames(all_trans) = c(&quot;x&quot;,&quot;y&quot;, &quot;feature_name&quot;,&quot;sample&quot;)
for (rp in c(&quot;rep1&quot;,&quot;rep2&quot;,&quot;rep3&quot;)){
  rp_lst = get(rp)
  iters_sp1= rp_lst$trans_info$feature_name %in% inters
  vis_r1 =rp_lst$trans_info[iters_sp1,
                    c(&quot;x&quot;,&quot;y&quot;,&quot;feature_name&quot;)]
  vis_r1$sample=paste(&quot;replicate&quot;, substr(rp,start = 4, stop=4),sep=&quot;&quot;)
  all_trans = rbind(all_trans,vis_r1)
}
all_trans = all_trans[2:nrow(all_trans),]


genes_plt&lt;- ggplot(data = all_trans,
            aes(x = x, y = y))+ 
            geom_point(size=0.001, alpha=0.5)+
            facet_grid(sample~feature_name)+
            scale_x_reverse()+
            guides(fill = guide_colorbar(barwidth = unit(15, &quot;cm&quot;), 
                                         height= unit(15, &quot;cm&quot;)))+
            defined_theme+ theme(legend.position = &quot;bottom&quot;, 
                      strip.background = element_rect(fill = &quot;white&quot;,
                                                      colour = &quot;black&quot;),
                      strip.text = element_text(size = rel(1.2)), 
                      legend.key.width = unit(2, &quot;cm&quot;),
                      legend.title = element_text(size = 10),
                      legend.text = element_text(size = 10),
                       legend.key.height = unit(2, &quot;cm&quot;))
        
genes_plt</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/Nosig-genes-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="wilcoxon-rank-sum-test" class="section level2">
<h2>Wilcoxon Rank Sum Test</h2>
<p>The <code>FindAllMarkers</code> function applies a Wilcoxon Rank Sum
test to detect genes with cluster-specific expression shifts. Only
positive markers (upregulated genes) were retained using a log
fold-change threshold of 0.1</p>
<pre class="r"><code>seu = readRDS(here(gdata,paste0(data_nm, &quot;_seu.Rds&quot;)))
rownames(cluster_info) &lt;- cluster_info$cell_id

# Reorder clusters$cluster to match Seurat cells
Idents(seu) &lt;- cluster_info$cluster[match(colnames(seu), rownames(cluster_info))]
find_markers_result &lt;- FindAllMarkers(seu, only.pos = TRUE,
                                      logfc.threshold = 0.1)
table(find_markers_result$cluster)
saveRDS(find_markers_result, here(gdata,paste0(data_nm, &quot;_seu_markers.Rds&quot;)))</code></pre>
<pre class="r"><code>FM_result= readRDS(here(gdata,paste0(data_nm, &quot;_seu_markers.Rds&quot;)))
FM_result = FM_result[FM_result$avg_log2FC&gt;0.1, ]

table(FM_result$cluster)</code></pre>
<pre><code>
c3 c6 c1 c2 c5 c4 c8 c9 c7 
96 64 99 96 79 71 49 50 58 </code></pre>
<pre class="r"><code>FM_result$cluster &lt;- factor(FM_result$cluster,
                            levels = cluster_names)

top_mg &lt;- FM_result %&gt;%
  group_by(cluster) %&gt;%
  arrange(desc(avg_log2FC), .by_group = TRUE) %&gt;% 
  slice_max(order_by = avg_log2FC, n = 10) %&gt;%
  select(cluster, gene, avg_log2FC)

p1&lt;-DotPlot(seu, features = unique(top_mg$gene)) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (WRS)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/load_FM_res-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="limma" class="section level2">
<h2>limma</h2>
<p>We applied the limma framework to identify differentially expressed
marker genes across clusters in the healthy liver sample. Count data
were normalized using log-transformed counts from the speckle package. A
linear model was fitted for each gene with cluster identity as the
design factor, and pairwise contrasts were constructed to compare each
cluster against all others. Empirical Bayes moderation was applied to
stabilize variance estimates, and significant genes were determined
using the moderated t-statistics from eBayes.</p>
<pre class="r"><code>all.bct &lt;- factor(cluster_info$cluster)
sample &lt;- spe_joint$sample_id
y &lt;- DGEList(cbind(cm1, cm2, cm3))
y$genes &lt;-row.names(rep1$cm)

logcounts.all &lt;- normCounts(y,log=TRUE,prior.count=0.1)

design &lt;- model.matrix(~0+all.bct+sample)
colnames(design)[1:(length(levels(all.bct)))] &lt;- levels(all.bct)

mycont &lt;- matrix(0,ncol=length(levels(all.bct)),nrow=length(levels(all.bct)))
colnames(mycont)&lt;-levels(all.bct)
diag(mycont)&lt;-1
mycont[upper.tri(mycont)]&lt;- -1/(length(levels(all.bct))-1)
mycont[lower.tri(mycont)]&lt;- -1/(length(levels(all.bct))-1)

# Fill out remaining rows with 0s
zero.rows &lt;- matrix(0,ncol=length(levels(all.bct)),nrow=(ncol(design)-length(levels(all.bct))))
test &lt;- rbind(mycont,zero.rows)
fit &lt;- lmFit(logcounts.all,design)
fit.cont &lt;- contrasts.fit(fit,contrasts=test)
fit.cont &lt;- eBayes(fit.cont,trend=TRUE,robust=TRUE)

fit.cont$genes &lt;-row.names(rep1$cm)
limma_dt &lt;- decideTests(fit.cont)
saveRDS(fit.cont, here(gdata,paste0(data_nm, &quot;_fit_cont_obj.Rds&quot;)))</code></pre>
<pre class="r"><code>fit.cont = readRDS(here(gdata,paste0(data_nm, &quot;_fit_cont_obj.Rds&quot;)))
limma_dt &lt;- decideTests(fit.cont)
summary(limma_dt)</code></pre>
<pre><code>        c1  c2  c3  c4  c5  c6  c7  c8  c9
Down   117 120  85 119 157 145 189 181 187
NotSig   3  11  10   6   5   8   1   6   3
Up     128 117 153 123  86  95  58  61  58</code></pre>
<pre class="r"><code>top_markers &lt;- lapply(cluster_names, function(cl) {
  tt &lt;- topTable(fit.cont,
                 coef = cl,
                 number = Inf,
                 adjust.method = &quot;BH&quot;,
                 sort.by = &quot;P&quot;)
  tt &lt;- tt[order(tt$adj.P.Val, -abs(tt$logFC)), ]   # tie-break by logFC
  tt$contrast &lt;- cl
  tt$gene &lt;- rownames(tt)
  head(tt, 10)   # top 10
})


# Combine into one data.frame
top_markers_df &lt;- bind_rows(top_markers) %&gt;%
  select(contrast, gene, logFC, adj.P.Val)


p1&lt;- DotPlot(seu, features = unique(top_markers_df$gene)) +
     scale_colour_gradient(low = &quot;white&quot;, high = &quot;red&quot;) +
    coord_flip()+
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))+
    labs(title = &quot;Top 10 marker genes (limma)&quot;, x=&quot;&quot;, y=&quot;&quot;)

p1</code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/looad_limma_res-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="comparison-of-marker-gene-detection-methods"
class="section level1">
<h1>Comparison of marker gene detection methods</h1>
<div id="cumulative-rank-plot" class="section level2">
<h2>Cumulative rank plot</h2>
<p>We compared the performance of different marker detection methods —
jazzPanda (linear modelling), limma, and Seurat’s Wilcoxon Rank Sum test
— by evaluating the cumulative average spatial correlation between
ranked marker genes and their corresponding target clusters at the
spatial vector level.</p>
<p>We can see that the linear modelling approach prioritizes highly
correlated marker genes more effectively than the other two methods.</p>
<pre class="r"><code>plot_lst=list()

cor1 &lt;- cor(sv_lst$gene_mt[1:2500, ],
            sv_lst$cluster_mt[1:2500, cluster_names], method = &quot;spearman&quot;)

cor2 &lt;- cor(sv_lst$gene_mt[2500:5000, ],
            sv_lst$cluster_mt[2500:5000, cluster_names], method = &quot;spearman&quot;)

cor3 &lt;- cor(sv_lst$gene_mt[5000:7500, ],
            sv_lst$cluster_mt[5000:7500, cluster_names], method = &quot;spearman&quot;)

cor_M &lt;-(cor1 + cor2 + cor3) / 3


Idents(seu)=cluster_info$cluster[match(colnames(seu), cluster_info$cell_id)]
for (cl in cluster_names){
    fm_cl=FindMarkers(seu, ident.1 = cl, only.pos = TRUE,
            logfc.threshold = 0.1)
    fm_cl = fm_cl[fm_cl$p_val_adj&lt;0.05, ]
    fm_cl = fm_cl[order(fm_cl$avg_log2FC, decreasing = TRUE),]
    to_plot_fm =row.names(fm_cl)

    
    limma_cl&lt;-topTable(fit.cont,coef=cl,p.value = 0.05, n=Inf, sort.by = &quot;p&quot;)
    limma_cl = limma_cl[limma_cl$logFC&gt;0, ]
    to_plot_lm = row.names(limma_cl)
    
    
    lasso_sig = jazzPanda_res[jazzPanda_res$top_cluster==cl,]
    lasso_sig = lasso_sig[order(lasso_sig$glm_coef, decreasing = TRUE),]
    
    FM_pt = data.frame(&quot;name&quot;=to_plot_fm,&quot;rk&quot;= 1:length(to_plot_fm),
                       &quot;y&quot;= get_cmr_ma(to_plot_fm,cor_M = cor_M, cl = cl),
                       &quot;type&quot;=&quot;Wilcoxon Rank Sum Test&quot;)
    
    
    lasso_pt = data.frame(&quot;name&quot;=lasso_sig$gene,&quot;rk&quot;= 1:nrow(lasso_sig),
                          &quot;y&quot;= get_cmr_ma(lasso_sig$gene,cor_M = cor_M, cl = cl),
                          &quot;type&quot;=&quot;jazzPanda-glm&quot;)
    
    limma_pt = data.frame(&quot;name&quot;=to_plot_lm,&quot;rk&quot;= 1:length(to_plot_lm),
                          &quot;y&quot;= get_cmr_ma(to_plot_lm,cor_M = cor_M, cl = cl),
                          &quot;type&quot;=&quot;limma&quot;)
  
    data_lst = rbind(limma_pt, lasso_pt,FM_pt)

    data_lst$type &lt;- factor(
        data_lst$type,
        levels = c(&quot;jazzPanda-glm&quot;, &quot;limma&quot;, &quot;Wilcoxon Rank Sum Test&quot;)
    )
    data_lst$rk = as.numeric(data_lst$rk)
    p &lt;-ggplot(data_lst, aes(x = rk, y = y, color = type)) +
        geom_step(size = 0.8) +  # type = &quot;s&quot;
        scale_color_manual(values = c(&quot;jazzPanda-glm&quot; = &quot;red&quot;,
                                      &quot;limma&quot; = &quot;black&quot;,
                                      &quot;Wilcoxon Rank Sum Test&quot; = &quot;blue&quot;)) +
        labs(title = paste0(&quot;Cluster: &quot;, cl), x = &quot;Rank of marker genes&quot;,
             y = &quot;Cumulative average correlation&quot;,
             color = NULL) +
        scale_x_continuous(limits = c(0, 50))+
        theme_classic(base_size = 12) +
        theme(plot.title = element_text(hjust = 0.5),
            axis.line = element_blank(),  
            panel.border = element_rect(color = &quot;black&quot;, 
                                        fill = NA, linewidth = 1),
            legend.position = &quot;inside&quot;,
            legend.position.inside = c(0.98, 0.02),
            legend.justification = c(&quot;right&quot;, &quot;bottom&quot;),
            legend.background = element_rect(color = &quot;black&quot;, 
                                             fill = &quot;white&quot;, linewidth = 0.5),
            legend.box.background = element_rect(color = &quot;black&quot;,
                                                 linewidth = 0.5)
        )
    plot_lst[[cl]] &lt;- p
}

combined_plot &lt;- wrap_plots(plot_lst, ncol = 3)
combined_plot </code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/cmr%20plot-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="upset-plot" class="section level2">
<h2>Upset plot</h2>
<p>We further visualized the overlap between marker genes detected by
the four methods using UpSet plots for each cluster. While limma and the
Wilcoxon rank-sum test detect a larger number of marker genes, many of
these lack clear spatial relevance. In contrast, our method identifies a
more refined and spatially coherent set of markers that align closely
with cluster boundaries. The moderate overlaps with non-spatial methods
suggest that our approach captures the most spatially informative
gene.</p>
<pre class="r"><code>plot_lst=list()
for (cl in cluster_names){
    findM_sig &lt;- FM_result[FM_result$cluster==cl &amp; FM_result$p_val_adj&lt;0.05,&quot;gene&quot;]
    limma_sig &lt;- row.names(limma_dt[limma_dt[,cl]==1,])
    lasso_cl &lt;- jazzPanda_res[jazzPanda_res$top_cluster==cl, &quot;gene&quot;]
    df_mt &lt;- as.data.frame(matrix(FALSE,nrow=nrow(jazzPanda_res),ncol=3))
    row.names(df_mt) &lt;- jazzPanda_res$gene
    colnames(df_mt) &lt;- c(&quot;jazzPanda-glm&quot;,
                      &quot;Wilcoxon Rank Sum Test&quot;,&quot;limma&quot;)
    df_mt[findM_sig,&quot;Wilcoxon Rank Sum Test&quot;] &lt;- TRUE
    df_mt[limma_sig,&quot;limma&quot;] &lt;- TRUE
    df_mt[lasso_cl,&quot;jazzPanda-glm&quot;] &lt;- TRUE
    
    df_mt$gene_name &lt;- row.names(df_mt)
    
    p = upset(df_mt,
               intersect=c(&quot;Wilcoxon Rank Sum Test&quot;, &quot;limma&quot;,
                           &quot;jazzPanda-glm&quot;),
               wrap=TRUE, keep_empty_groups= FALSE, name=&quot;&quot;,
               #themes=theme_grey(),
               stripes=&#39;white&#39;,
               sort_intersections_by =&quot;cardinality&quot;, sort_sets= FALSE,min_degree=1,
               set_sizes =( 
                   upset_set_size()
                   + theme(axis.title= element_blank(),
                           axis.ticks.y = element_blank(),
                           axis.text.y = element_blank())),
               sort_intersections= &quot;descending&quot;, warn_when_converting=FALSE,
               warn_when_dropping_groups=TRUE,encode_sets=TRUE,

               width_ratio=0.3, height_ratio=1/4)+
              ggtitle(paste0(&quot;Cluster: &quot;, cl))+
                 theme(plot.title = element_text( size=15))
         
    plot_lst[[cl]] &lt;- p 
}


combined_plot &lt;- wrap_plots(plot_lst, ncol = 3)

combined_plot </code></pre>
<p><img src="xenium-mouse-brain-analysis_files/figure-html/upset-mg-comparison-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="annotate-cell-type-with-top-marker-genes"
class="section level1">
<h1>Annotate cell type with top marker genes</h1>
<pre class="r"><code>anno_df &lt;- data.frame(
    cluster = cluster_names,
    major_class = c(
    &quot;Inhibitory neuron&quot;,
    &quot;Oligodendrocyte lineage&quot;,
    &quot;Excitatory neuron&quot;,
    &quot;Excitatory neuron&quot;,
    &quot;Non-neuronal (barrier/stromal mix)&quot;,
    &quot;Inhibitory projection neuron&quot;,
    &quot;Excitatory neuron&quot;,
    &quot;Excitatory neuron&quot;,
    &quot;Cholinergic neuron&quot;
  ),
  sub_class = c(
    &quot;L2/3 LAMP5/VIP interneuron&quot;,
    &quot;Mature oligodendrocyte (myelinating)&quot;,
    &quot;IT L2/3&quot;,
    &quot;Corticothalamic-like&quot;,
    &quot;Choroid plexus &amp; vascular/stromal admixture&quot;,
    &quot;Striatal medium spiny neuron (MSN)&quot;,
    &quot;Dentate gyrus granule&quot;,
    &quot;IT L6 corticocortical (Neurod6+)&quot;,
    &quot;Cortical/striatal cholinergic interneuron&quot;
  ),
  cell_type = c(
    &quot;VIP/LAMP5 (CPLX3+, ID2+)&quot;,
    &quot;MOL (Opalin+/Gjc3+/Sox10+)&quot;,
    &quot;IT-L2/3 (BHLHE22+/HPcal1+)&quot;,
    &quot;L6 CT-like (FOXP2+/NTSR2+/VGlut2+)&quot;,
    &quot;Choroid plexus epithelium–enriched&quot;,
    &quot;MSN (DARPP-32+, GABAergic)&quot;,
    &quot;DG granule (Prox1+/Calb1+)&quot;,
    &quot;IT-L6/CC (Neurod6+/Wfs1+/Shisa6+)&quot;,
    &quot;ChAT+/SNCG+ interneuron&quot;
  ),
  supporting_genes = c(
    &quot;JP: Id2, Lamp5, Vip, Cplx3, Crh, Cort; FM: Lamp5, Vip, Cplx3, Satb2&quot;,
    &quot;JP: Opalin, Gjc3, Sox10; FM: Opalin, Sox10, Gjc3, Spag16&quot;,
    &quot;JP: Slc17a7, Bhlhe22, Hpcal1, Cdh13; FM: Cabp7, Bhlhe22, Syt17&quot;,
    &quot;JP: Slc17a6, Foxp2, Ntsr2, Rims3; FM: Slc17a6, Foxp2, Rims3&quot;,
    &quot;JP: Igf2, Slc13a4, Kdr, Pecam1, Emcn, Col1a1, Acta2; FM: Fmod, Aldh1a2, Igf2, Col1a1, Spp1&quot;,
    &quot;JP: Ppp1r1b, Gad1/2, Penk, Pdyn, Meis2; FM: Penk, Pdyn, Ppp1r1b, Gad2&quot;,
    &quot;JP: Prox1, Calb1, Prdm8; FM: Prox1, Calb1&quot;,
    &quot;JP: Neurod6, Wfs1, Shisa6, Igfbp4; FM: Neurod6, Pou3f1, Shisa6, Wfs1&quot;,
    &quot;JP: Chat, Sncg, Syt6; FM: Chat, Sncg, Tacr1, Kctd8&quot;
  ),
  notes = c(
    &quot;Robust LAMP5/VIP signal (Id2/Lamp5/Cplx3). SATB2/CUX2 likely ambient; minor immune genes likely contamination.&quot;,
    &quot;Opalin+Gjc3+Sox10 define mature oligos; low Gpr17 argues against OPC.&quot;,
    &quot;Slc17a7+ with IT adhesion/signaling genes; endothelial/immune hits in FM likely contaminants.&quot;,
    &quot;Foxp2+Ntsr2+ with Slc17a6 (VGlut2) suggest L6 corticothalamic-like; some L4/5 IT (Rorb) admixture.&quot;,
    &quot;Strong CP markers (Igf2, Slc13a4) with endothelial/smc/fibro mix → perivascular/meningeal CP neighborhood; ependymal Trp73 low.&quot;,
    &quot;Classic MSN signature (DARPP-32 + Gad1/2 + Penk/Pdyn); indicates striatal tissue present.&quot;,
    &quot;Prox1 hallmark of DG granule; Calb1 supportive.&quot;,
    &quot;Canonical Neurod6 IT-L6 corticocortical program; small PT marker bleed-through possible.&quot;,
    &quot;Clear ChAT/SNCG cholinergic identity; Syt6 hints at small L6 CT admixture.&quot;
  ),
  anno_name=c(
  &quot;Inh_L23_LAMP5_VIP&quot;,        
  &quot;Oligo_Mature_MOL&quot;,         
  &quot;Exc_L23_IT&quot;,               
  &quot;Exc_L6_CTlike&quot;,            
  &quot;NonNeur_CP_VascStromal&quot;,   
  &quot;Inh_Striatal_MSN&quot;,         
  &quot;Exc_DG_Granule&quot;,           
  &quot;Exc_L6_IT_CC&quot;,            
  &quot;Cholin_ChAT_SNCG&quot;),
  confidence = c(
    &quot;Medium-High&quot;,
    &quot;High&quot;,
    &quot;Medium&quot;,
    &quot;Medium&quot;,
    &quot;Medium-Low&quot;,
    &quot;High&quot;,
    &quot;High&quot;,
    &quot;Medium-High&quot;,
    &quot;High&quot;
  ),
  stringsAsFactors = FALSE
)

anno_df</code></pre>
<pre><code>  cluster                        major_class
1      c1                  Inhibitory neuron
2      c2            Oligodendrocyte lineage
3      c3                  Excitatory neuron
4      c4                  Excitatory neuron
5      c5 Non-neuronal (barrier/stromal mix)
6      c6       Inhibitory projection neuron
7      c7                  Excitatory neuron
8      c8                  Excitatory neuron
9      c9                 Cholinergic neuron
                                    sub_class
1                  L2/3 LAMP5/VIP interneuron
2        Mature oligodendrocyte (myelinating)
3                                     IT L2/3
4                        Corticothalamic-like
5 Choroid plexus &amp; vascular/stromal admixture
6          Striatal medium spiny neuron (MSN)
7                       Dentate gyrus granule
8            IT L6 corticocortical (Neurod6+)
9   Cortical/striatal cholinergic interneuron
                           cell_type
1           VIP/LAMP5 (CPLX3+, ID2+)
2         MOL (Opalin+/Gjc3+/Sox10+)
3         IT-L2/3 (BHLHE22+/HPcal1+)
4 L6 CT-like (FOXP2+/NTSR2+/VGlut2+)
5 Choroid plexus epithelium–enriched
6         MSN (DARPP-32+, GABAergic)
7         DG granule (Prox1+/Calb1+)
8  IT-L6/CC (Neurod6+/Wfs1+/Shisa6+)
9            ChAT+/SNCG+ interneuron
                                                                            supporting_genes
1                        JP: Id2, Lamp5, Vip, Cplx3, Crh, Cort; FM: Lamp5, Vip, Cplx3, Satb2
2                                   JP: Opalin, Gjc3, Sox10; FM: Opalin, Sox10, Gjc3, Spag16
3                             JP: Slc17a7, Bhlhe22, Hpcal1, Cdh13; FM: Cabp7, Bhlhe22, Syt17
4                                JP: Slc17a6, Foxp2, Ntsr2, Rims3; FM: Slc17a6, Foxp2, Rims3
5 JP: Igf2, Slc13a4, Kdr, Pecam1, Emcn, Col1a1, Acta2; FM: Fmod, Aldh1a2, Igf2, Col1a1, Spp1
6                      JP: Ppp1r1b, Gad1/2, Penk, Pdyn, Meis2; FM: Penk, Pdyn, Ppp1r1b, Gad2
7                                                  JP: Prox1, Calb1, Prdm8; FM: Prox1, Calb1
8                       JP: Neurod6, Wfs1, Shisa6, Igfbp4; FM: Neurod6, Pou3f1, Shisa6, Wfs1
9                                         JP: Chat, Sncg, Syt6; FM: Chat, Sncg, Tacr1, Kctd8
                                                                                                                            notes
1                  Robust LAMP5/VIP signal (Id2/Lamp5/Cplx3). SATB2/CUX2 likely ambient; minor immune genes likely contamination.
2                                                           Opalin+Gjc3+Sox10 define mature oligos; low Gpr17 argues against OPC.
3                                   Slc17a7+ with IT adhesion/signaling genes; endothelial/immune hits in FM likely contaminants.
4                              Foxp2+Ntsr2+ with Slc17a6 (VGlut2) suggest L6 corticothalamic-like; some L4/5 IT (Rorb) admixture.
5 Strong CP markers (Igf2, Slc13a4) with endothelial/smc/fibro mix → perivascular/meningeal CP neighborhood; ependymal Trp73 low.
6                                       Classic MSN signature (DARPP-32 + Gad1/2 + Penk/Pdyn); indicates striatal tissue present.
7                                                                                 Prox1 hallmark of DG granule; Calb1 supportive.
8                                        Canonical Neurod6 IT-L6 corticocortical program; small PT marker bleed-through possible.
9                                                      Clear ChAT/SNCG cholinergic identity; Syt6 hints at small L6 CT admixture.
               anno_name  confidence
1      Inh_L23_LAMP5_VIP Medium-High
2       Oligo_Mature_MOL        High
3             Exc_L23_IT      Medium
4          Exc_L6_CTlike      Medium
5 NonNeur_CP_VascStromal  Medium-Low
6       Inh_Striatal_MSN        High
7         Exc_DG_Granule        High
8           Exc_L6_IT_CC Medium-High
9       Cholin_ChAT_SNCG        High</code></pre>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Red Hat Enterprise Linux 9.3 (Plow)

Matrix products: default
BLAS:   /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRblas.so 
LAPACK: /stornext/System/data/software/rhel/9/base/tools/R/4.5.1/lib64/R/lib/libRlapack.so;  LAPACK version 3.12.1

locale:
 [1] LC_CTYPE=en_AU.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_AU.UTF-8        LC_COLLATE=en_AU.UTF-8    
 [5] LC_MONETARY=en_AU.UTF-8    LC_MESSAGES=en_AU.UTF-8   
 [7] LC_PAPER=en_AU.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_AU.UTF-8 LC_IDENTIFICATION=C       

time zone: Australia/Melbourne
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] harmony_1.2.3               Rcpp_1.1.0                 
 [3] Banksy_1.4.0                dotenv_1.0.3               
 [5] here_1.0.1                  xtable_1.8-4               
 [7] clustree_0.5.1              ggraph_2.2.1               
 [9] speckle_1.8.0               edgeR_4.6.3                
[11] limma_3.64.1                ggvenn_0.1.10              
[13] RColorBrewer_1.1-3          patchwork_1.3.1            
[15] gridExtra_2.3               corrplot_0.95              
[17] ComplexUpset_1.3.3          caret_7.0-1                
[19] lattice_0.22-7              glmnet_4.1-10              
[21] Matrix_1.7-3                dplyr_1.1.4                
[23] data.table_1.17.8           ggplot2_3.5.2              
[25] Seurat_5.3.0                SeuratObject_5.1.0         
[27] sp_2.2-0                    SpatialExperiment_1.18.1   
[29] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1
[31] Biobase_2.68.0              GenomicRanges_1.60.0       
[33] GenomeInfoDb_1.44.2         IRanges_2.42.0             
[35] S4Vectors_0.46.0            BiocGenerics_0.54.0        
[37] generics_0.1.4              MatrixGenerics_1.20.0      
[39] matrixStats_1.5.0           jazzPanda_0.2.4            
[41] spatstat_3.4-0              spatstat.linnet_3.3-1      
[43] spatstat.model_3.4-0        rpart_4.1.24               
[45] spatstat.explore_3.5-2      nlme_3.1-168               
[47] spatstat.random_3.4-1       spatstat.geom_3.5-0        
[49] spatstat.univar_3.1-4       spatstat.data_3.1-8        

loaded via a namespace (and not attached):
  [1] spatstat.sparse_3.1-0   lubridate_1.9.4         httr_1.4.7             
  [4] tools_4.5.1             sctransform_0.4.2       R6_2.6.1               
  [7] lazyeval_0.2.2          uwot_0.2.3              mgcv_1.9-3             
 [10] withr_3.0.2             progressr_0.15.1        cli_3.6.5              
 [13] textshaping_1.0.1       fastDummies_1.7.5       labeling_0.4.3         
 [16] sass_0.4.10             ggridges_0.5.6          pbapply_1.7-4          
 [19] systemfonts_1.3.1       dbscan_1.2.2            R.utils_2.13.0         
 [22] aricode_1.0.3           parallelly_1.45.1       shape_1.4.6.1          
 [25] ica_1.0-3               abind_1.4-8             R.methodsS3_1.8.2      
 [28] lifecycle_1.0.4         yaml_2.3.10             recipes_1.3.1          
 [31] SparseArray_1.8.1       Rtsne_0.17              promises_1.3.3         
 [34] crayon_1.5.3            miniUI_0.1.2            cowplot_1.2.0          
 [37] magick_2.8.7            pillar_1.11.0           knitr_1.50             
 [40] rjson_0.2.23            future.apply_1.20.0     codetools_0.2-20       
 [43] glue_1.8.0              vctrs_0.6.5             png_0.1-8              
 [46] spam_2.11-1             gtable_0.3.6            cachem_1.1.0           
 [49] gower_1.0.2             xfun_0.52               S4Arrays_1.8.1         
 [52] mime_0.13               prodlim_2025.04.28      tidygraph_1.3.1        
 [55] survival_3.8-3          timeDate_4041.110       RcppHungarian_0.3      
 [58] iterators_1.0.14        hardhat_1.4.2           lava_1.8.1             
 [61] statmod_1.5.0           fitdistrplus_1.2-4      ROCR_1.0-11            
 [64] ipred_0.9-15            bit64_4.6.0-1           RcppAnnoy_0.0.22       
 [67] BumpyMatrix_1.16.0      rprojroot_2.1.0         bslib_0.9.0            
 [70] irlba_2.3.5.1           KernSmooth_2.23-26      colorspace_2.1-1       
 [73] nnet_7.3-20             tidyselect_1.2.1        bit_4.6.0              
 [76] compiler_4.5.1          DelayedArray_0.34.1     plotly_4.11.0          
 [79] scales_1.4.0            hexbin_1.28.5           lmtest_0.9-40          
 [82] stringr_1.5.1           digest_0.6.37           goftest_1.2-3          
 [85] presto_1.0.0            spatstat.utils_3.1-5    rmarkdown_2.29         
 [88] XVector_0.48.0          htmltools_0.5.8.1       pkgconfig_2.0.3        
 [91] fastmap_1.2.0           rlang_1.1.6             htmlwidgets_1.6.4      
 [94] UCSC.utils_1.4.0        shiny_1.11.1            farver_2.1.2           
 [97] jquerylib_0.1.4         zoo_1.8-14              jsonlite_2.0.0         
[100] BiocParallel_1.42.1     mclust_6.1.1            ModelMetrics_1.2.2.2   
[103] R.oo_1.27.1             magrittr_2.0.3          GenomeInfoDbData_1.2.14
[106] dotCall64_1.2           viridis_0.6.5           reticulate_1.43.0      
[109] leidenAlg_1.1.5         stringi_1.8.7           pROC_1.19.0.1          
[112] MASS_7.3-65             plyr_1.8.9              parallel_4.5.1         
[115] listenv_0.9.1           ggrepel_0.9.6           deldir_2.0-4           
[118] graphlayouts_1.2.2      sccore_1.0.6            splines_4.5.1          
[121] tensor_1.5.1            locfit_1.5-9.12         igraph_2.1.4           
[124] RcppHNSW_0.6.0          reshape2_1.4.4          evaluate_1.0.4         
[127] foreach_1.5.2           tweenr_2.0.3            httpuv_1.6.16          
[130] RANN_2.6.2              tidyr_1.3.1             purrr_1.1.0            
[133] polyclip_1.10-7         future_1.67.0           scattermore_1.2        
[136] ggforce_0.5.0           RSpectra_0.16-2         later_1.4.2            
[139] viridisLite_0.4.2       class_7.3-23            ragg_1.5.0             
[142] tibble_3.3.0            memoise_2.0.1           cluster_2.1.8.1        
[145] timechange_0.3.0        globals_0.18.0         </code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
